<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="By ChenShanshan" />


<title>Introduction to StatComp18017</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Introduction to StatComp18017</h1>
<h4 class="author"><em>By ChenShanshan</em></h4>
<h4 class="date"><em>2019/1/19</em></h4>



<p><font color="#FF7F50"></p>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p><font color="#000000"></p>
<p><strong>StatComp18017</strong> is a simple R package including my homework.</p>
<p>The R package ‘microbenchmark’ can be used to benchmark the above R and C++ functions.</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-1" class="section level2">
<h2>Question 1</h2>
<p><font color="#000000"></p>
<p>Most otfen, it is necessary to split the picture or know how the device splits the windows that we need.Apparently??the function split.screen()could fits our needs,but with the shortcomings of incompatibility and unability to be performed on several devices.Thus,how to know that the route to split??and how to code?</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-1" class="section level2">
<h2>Answer 1</h2>
<p><font color="#000000"></p>
<p>First,we can generate a matrix with specified construction,then utilize the funcyion of layout() to display the details of the split process.The examples are follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat&lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">2</span>)
mat</code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]    1    3
## [2,]    2    4</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">layout</span> (mat)
<span class="kw">layout.show</span>(<span class="dv">4</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAV1BMVEUAAAAAADoAAGYAOjoAOpA6AAA6ADo6AGY6OpA6ZmY6kNtmAABmADpmOpBmtv+QOgCQOmaQ2/+2ZgC225C2///bkDrbtrbb////tmb/25D//7b//9v///8SIqCzAAAACXBIWXMAAA7DAAAOwwHHb6hkAAADE0lEQVR4nO3cwVITURBA0QFBVBQFiSRh/v87BXcu9C7sKl9R535A16uTnpfJJtumaF+qxY6zA4oARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARVNAp5vHkTkzQMdtu7ifGDQGdL69XAjo9P5+Pw4daAbosH1caoNeOn+eWaEZoB9Paz1iLx2unkbmvM07aD9dL3YHrQa0L3YHLQi01h0EKFsK6PXxOn1wSf+5w7aNvZf5qfH3AEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKNmX/+zP6vcWO4xGrAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAohmg0/W23Q3M2ceATjePM4NGgM63d/vx4v7fB+1jQA+XKwEdr572568zKzQDdHy31Aa99rpFE40c5/nb9+WADi9bNNHIcQ6f1rqDXjqMPfMDM85fnlYDOgxd0TNAD3eLfYvN7c8I0Pn211+wDt2JE0Cn91P780bfgx4mP7GRKYsBTbbYcQBVgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBijZFPwE0za13Amcg8QAAAABJRU5ErkJggg==" /><!-- --></p>
<p>It is also feasible that fuse the function layout() and matrix() to simplify the code and get to our targets. As the examples show below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">layout</span> (<span class="kw">matrix</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>,<span class="dv">3</span>,<span class="dv">2</span>))
<span class="kw">layout.show</span>(<span class="dv">6</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAXVBMVEUAAAAAAGYAOjoAOpAAZrY6AAA6ADo6AGY6OpA6kNtmAABmADpmAGZmZjpmkJBmtv+QOgCQOjqQOmaQ2/+2ZgC2/7a2///bkDrb////tmb/25D/29v//7b//9v///97PZtAAAAACXBIWXMAAA7DAAAOwwHHb6hkAAADQUlEQVR4nO3dy1ITURRA0aig4BuNdoiQ//9MyZCJm8EpuIVrf8CtWyt90qOus9spOi3VYtc5AYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYCiKaC7z79HzhkD2q5nzhkCuvv4Zi2g4+VSQPc/fq31BN3ffFkKaLkR264XG7HFgI5XB0D/ajt/KDgj9DqBTsu9xQA9W4tdB1AFKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFD0hO1R/30v/Rs9brHrGLEKUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhTNAN1/3+0uBs45TQHtd7uhfV4zQNvFg9G3gYOGgO5vfk4cc25uxIZ2xYxc5/hhaJHOINDUcp+R69y+vRwb+anlR5+GHuqZ6/w5nLahkZ8BOr6fGvqxt9jUyI8ATc78xCG3D/O11BO0P3/7utCf9PlC9os9T4AiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYCi3q6ll/6NHrfYdYxYBSgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKhoD2u93MIpQhoG1u39kI0O3FWtuhjleHsX1nr3J91tDio3NTQNtSI7b/utx+sYdLLbTCb//uMHafV7nC74yz1H6x8z6vlZ6gyfuMvean9nmNnLLZL/ZcAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYqesl/r/+4vYZME64txSzsAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m&lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">3</span>),<span class="dv">2</span>,<span class="dv">2</span>)
<span class="kw">layout</span> (m)
<span class="kw">layout.show</span>(<span class="dv">3</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAASFBMVEUAAAAAADoAAGYAOpA6AAA6ADo6AGY6OpA6ZmZmAABmADpmtv+QOgCQ2/+2ZgC2///bkDrbtrbb////tmb/25D//7b//9v///+vIR/xAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAC20lEQVR4nO3cS25TQRBA0YaETwghxnZ4+98pMnNyJ6XIap2zgFL7ql97VmsRjrtyZ8c5BAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoChalA16+vI3N2DfT29Fmgd5zWdzfoPb/PPrEgUBAoCBQECgIFgT7KnR1HoCJQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKewa6rPXpZWLQpoGuX16Oy9Bixy0D3bz9mLlC6+6M/KzjOD2cR+bMnWjIzHGuj96g4A0Kg2/QyJwxAoWJ49w+r+s3j/T/ndYa22+9ZaBBAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQmAl0fVzreWDOsWmgt6fn4zK0VW3LQJeH8/Hn58wV2jLQze0WTdg20Nhix5Epc6YCncZ2hs2MGTMU6DS1+HLTQGP3Z9NAt/XEU7YM9OvfGl//Yh/izo4jUBEoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaCwCH8B232KufdIPmwAAAAASUVORK5CYII=" /><!-- --></p>
<p>In addition, options in function of layout(), such as widths and heights, give the possibilities to alter the split width and weight.So, the screen could be separated into plentiful shapes even be some complicated or weird.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m&lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>),<span class="dv">2</span>,<span class="dv">2</span>)
<span class="kw">layout</span> (m,<span class="dt">widths =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>),<span class="dt">heights =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">layout.show</span>(<span class="dv">2</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAQlBMVEUAAAAAADoAAGYAOpA6AAA6ADo6AGY6ZmZmAABmtv+QOgCQ2/+2ZgC2///bkDrbtrbb////tmb/25D//7b//9v///8bolO5AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAC30lEQVR4nO3d2W0CQRRFwfEOXjCLyT9VLwH4qCUjD+o6AVxB6QFfaJZF0fmqu/TLB1T7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGJ/KqDj/bJsB/dnAjpttufDzcvY/kxAh7v9+eN57ISmAvru+4qG9mcD2n1d0dD+ZEC727fB/bmAdoNf0bMBDd/PZEDHh9H7mQzo9ec/un7F/nYfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGL/6oEu33+/xXUHKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYrWBnR8HH40yGVbGdBpM/7slMu2LqDd8uSCfut97yMWAYoARYAiQBGgCNC1BSgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQtCj6BOSsJIjMeJ3DAAAAAElFTkSuQmCC" /><!-- --></p>
<p>These are about the split of screen with the help of function of layout and matrix.</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-2" class="section level2">
<h2>Question 2</h2>
<p><font color="#000000"></p>
<p>In the analysis process,to make a intuitive impression, the images are indispensable. So knowing how to code to display the relevant images is essential.Thus??here gives some simple examples.</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-2" class="section level2">
<h2>Answer 2</h2>
<p><font color="#000000"></p>
<p>The simplist example of the plot is followed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x&lt;-<span class="kw">rnorm</span>(<span class="dv">10</span>)
y&lt;-<span class="kw">rnorm</span>(<span class="dv">10</span>)
<span class="kw">plot</span>(x,y)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAUVBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6Ojo6kNtmAABmADpmZmZmtv+QOgCQ2/+2ZgC2/7a2///bkDrb/9vb////tmb/25D//7b//9v///8QDhghAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAFxUlEQVR4nO3djXaaSABAYdqk2627Td3aGOP7P+iKqBEF7gxhYMB7z9k9TTuO+lUBy4/F3jorpn4AuScQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQFAe0KYri5/EXX383TzebkgBtvvza71bPXUB4f+EPLWlJgN5ffhz///TaG6iIvMtkJQHarY5vr/366fUGKPSVW0TfZ6oSvoIOrZ97voIWDnR5Y+1WhUCNbap12OG15DIo0f0uei02/nTpEggSCBIIEggSCBIIEggSCBIIEggSCBIIEggSCBIIEggSCBIIEghKtOPwYwehu56bat/bE3pMwNJ3+7y/PH9qusXvONxvT3sOe063fKBPTifQoy+DPj/dstdio0+XLoEggSCBoLkCjbYMnynQeFsB8wQacTtSoKC7GnLgGNM9ElC/pe3jLIP6PtNHWYvl86m9LYEggaC5LoNGa2qglqVtLv8alAFQ659mIpRsr0bnXjGYLqclUzzQblXALp3ybKjTGXXb8y+i7nfeQNUpzS3Pu+pyxuFh7NNr/P3OHWhPRudzVg9te52zuoRl0KZjCfPZV9D812LborxuwPG878bOZ2T2XAblVA+g8tiNSmbbupI6H9/RJrhkoN3qy68R73fq8txQzKjHBgpYFzwc0LVJyNbEnIH6bAtcmwRtj84YqM/WZM1k4UC9Po8IFHWjZS+D+n2irZssey3W7xNt7JJ9xkDjfKKdM9AoCQQJBAkECQQJBAnU1mkjQqCWiqv/gm8w8P3n3OWDjEDNCQQJRE20DIq/0PdkuRYLSyBIIEggSCBIIEggSCBIIEggSCBIIEggSCBIIEggSCBIIEggKBEQf0tm1HQTlgZogG/JzKUkQKHfkjmLnRuDD9x3fUtmbbp+R2GOXMJX0B6+JfOy8zLrEi2DQr4l85GBWr8l83rP6kMDBU33uMugU1cnPzdP97BrsVMINIcmA5pNEwHFTx00asCpYp+xQMMOH35qgYYYJdB4Uwk0zKi+w4efWqAhRgk03lQZAS0jgSCBIIEggSCBIIEggSCBIIEggSCBIIGgFEDborhcF7a6PO79tc6vx9R/iJyo7O3776abtI7qmuu2BEDbwyPcnh/l21+ND7c2pvZD5ERlu9XlSIG2qeqjOua6a3igaq/Q+vT303zp5dqY+g3iJjr+ycclwtumqo/qmOu+4YHevpVHgJyPlNk0vpBrY+o3iJtof7z49+UJt01VH9U+V0MJgI6v3/OjWf9dNFwmvzamfoO4iao+gFqmqo/qnOu24YGqRcBpQbBblRcvX98+mNqY2g+RE51Gnp9621T1UZ1z3ZYY6OahNY4JAmqZ6Oa3w4A657ot9Vus+q1vPzvGhL3Fmieqin2Ldc1126BAx6/laFhQ3q1WeyykmyeqClhI7++Bwtb1iVfz1SO++yuNX823TFR1+e2O1fwd42RvsfrG2vHB3i8P4zcUWyaqBoZsKF6vxTrmui3FR41Ntbn//nJ8HEVRNLzZa2M2LZ8PQiYqOz717qnqozrmus0Pq5BAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkEBQ7kDl0WK7VeARqSnKHag8ZmxTHpQ6VbkD7bdf//sn/MSB4cseaL8OPOskUfkDbUMPlktT9kDvL/9GnHoyfNkDbZ7+BF5ULk25A5WXaI45e2nwcgdaH1bx1aG7E5U70OQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkECQQJBAkE/Q+OH2Z7VB2UuwAAAABJRU5ErkJggg==" /><!-- --></p>
<p>In fact, changing the image characters and making the iamges more artistic and analyzable, through adjusting some parameter values, are practicible.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(x,y,<span class="dt">xlab =</span> <span class="st">&quot;Ten random values&quot;</span>,<span class="dt">ylab =</span> <span class="st">&quot;Ten other values&quot;</span>, <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">pch=</span><span class="dv">22</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">bg=</span><span class="st">&quot;yellow&quot;</span>, <span class="dt">bty=</span><span class="st">&quot;7&quot;</span>,<span class="dt">tcl=</span><span class="fl">0.4</span>, <span class="dt">main=</span><span class="st">&quot;How to customize a plot with R&quot;</span>, <span class="dt">las=</span><span class="dv">1</span>,<span class="dt">cex=</span><span class="fl">1.5</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAbFBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6kNtmAABmADpmZmZmtrZmtv+QOgCQOjqQkGaQtpCQ2/+2ZgC2Zma225C2/7a2/9u2///bkDrb/7bb/9vb////AAD/tmb/25D//wD//7b//9v///9ulbq9AAAACXBIWXMAAA7DAAAOwwHHb6hkAAALCklEQVR4nO2di5qjKAJGra7UbjI9qZ0td2fb3W5rUnn/d1wuguDtBwU15j/fN+maCAJHQKKgxZ1MUmydgb1DQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFAcYEfb0Xrz/1v6eQ/fzrx8wMoIgmIzMYijqSnAkqN1fFy0e7JY2g2+XbPEE4YlJBo8npoHpzDkFlMVPQ7Igh9AWB5PTmeYI+34qiOJvvm61lk4FSbFP7NqEstdxytWnfLmprJb/VO+hGFCFOtYojAsmMqqTEtwqVcxnJOWgqalskkdCfF71TUwa7c5tcrxj/k381m6Wg0u4yUFClMyi+qmQKIsMiflPcNuE2lM2t4uwLaoKJL/oRjQobxBPUBrA1oXbUOUnKL5oytDtvBXWL8eUJKpwkJgQZTuoYKDeieJ9vYp/6GItvr/aw/fBC2cyfZfpKhRUkK0mzrR9Rlv98t3Xo7LSTUlkom5KdbT51/kyapSpZLbOto7q5aptYpxg6qGlibjHCBDU7lv+IDWcR+7v6NMnp7U4o9+uq3aYKJjJkj38vogihq4kumiNIlbltpG5nUjmNrmybtFtq/U+btU4xOoKuNqEwQSaTqvsqxRflyz/fTuovp5xeqGYftrU5gppdnwcj6uqlPz1BuiLq/sRtUyajXlb0vlTUTt6dQ+MWwxck962zENgHmcC1dFvLLk10hN/+825amK33baj73T8D2rZ0vtt+5joQcVSQPq6myymcg2BjuYLkvkz31ebKEeQXY6Eg7yjcLsK76NHEp9MzhtUg21eY81hwDRJ/m7Pp1c2nVR5bg/xiLBTktWPxzfem/XaKP9IH1TI9narumb/+8XH3OgW/DxoSJD71Th0Vd1sxOp10SB/UKcZSQd75qSp0S3NGPHqfQ2cxVf2cM5LupG2WehFHBFVFe8IUf+lOvKlBqpa4nfT4WcwdA3rFcHMzR5AZS5haLvagP1sTznCmPcSlO1CwY6Kq7UV6EYcF2dgn04F1+2hXkDeE8vJeOyMorxg6aF2YgWK0IF1A2/fq1J0zbTPydEK15nSFkpte/6vPnSqYjt2NOCzIds2nRkk7FpX/+/LhjThEt6sCeCPpq5vPXjHsmFrtK0TQw5L61x0FASgIQEGAwwlKDQUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgQCvodjk3EzyJQyuolNNwTvcyaGnG82AF3S5yupX4D9/aVjN0zoObiodhlqBSTvaCgr7eXz5EY8xa035ZMiUQL0i0rdtFzpqFBf98k7O1sMgl/PqrYUeC1MKFr/fQilG7S2KSs0dBkZSDNSi6iY/w+ILq4V461bBql4LU5ONy+OzUpR7po48sSHQqleykQwyN1J9DC/p6FwNpOds84OxUjfk5siA5DpKCAgaKlb+eZFa6gB0KMjWohD/GPt/GW+GBBTV90ETtMDhLLeanC9jjSFovE1k4/HuYqyf5B4qr7C4fFASYcxZrfiQs+hF6YEENAWf5JOluzfwmtuyK4hMIWlaFnkDQsgthxxcUcEUxSbpbM/8stuy2z4EFrZzu1lAQgIIAkYLc5z7xLDYvoM/n34c1UpBm9HF7R74eZJpZQBOrR0Md+Ypi+fqzOjX3laepi/PYD5IDC5L3e2p5VyNopDgq6Lh3VuVdDdn1jnW/Pk9Yg+RdjdtvHxQ0GlD+jC/PS5tYaLqAPQqSl8qahxVinlJQDBQEeEJBcVfKsgva4Ui6XH5fNd9vseTC5jUx9TS+XV5yTd7kZvdBX++7vNyxE0GsQRMB1fT5/fZB2wtaeL8nNt0BpjriHQhaOd0BpiRQ0J2CIBQEmBS0/UBRXg9aMd0Bsv/+cpl3RXHFdAdwBGX/KTZrGnCKBWCpBGWvTHnnKNajQ8oDC4pArqUbWU+XaKD42IJ0dz48lzHRSXGfgkLXi02tWT2yoOD1Yp9/k61ruE8/sKDw9WK6+xnuhBLdWd2joPD1YpOCQtOdZo+CItaLPWcTC18vtkYnvceRdPB6sRVO83H86oHjPOJAcTa2Qf71K7hh5r3cUWX5qTHNRP1YSZB5U9cub/t4FiY25RQ0/DCOSA4sKOyRC8nSjWYHgra+YDbN5oLCnxyUJN1oNhcUNAE4XbrRbCvoAdZqbF+DkrCNoFVG0nIKsGSvi3qjio+ZL4iLegcDlm0ftGg4dFhBbQ1aKd38TDfJ43XS0UxfdJsjSL0S+TjPD0ouqG7elR40XHyAJZmpBZnZHUGLWbIvyUxAakHmx2rIOCj/kswEbFiDVliSmYBt+6BnFBR1Fsu9ZjUB6QXF8JQ1KIZHELTFSLoyv9UeQNA0rEEACgJQEOB4d1YTc7w7q4k53p3VxBzvzmpijndnNTGzfqzu+s5qYuY0sX3fWU0ML9oDKAgwc5Zr8Lt9Fqe7NbPnSS8cDh1YUMy7fRKkuzXzBoqB7/ZJke7WzK9BeK1GknR75F994DG7Dwq5qzH1GtHQdHusuih89lkscK3G+GtEi6jU3XATggJ3FxUs5zgIr/Z5ckGaqQV1xxM041pHiteIHlnQ6GsgY1I/sKCx14i6Ow1NXHMsQZP155iCwmbaN3dWx18j6u40NHHN3gXF1CA4mCx2TXZBU68RPR4zBE29RvR4zB4HPQsPc31iKygIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQYAcgibmD3XBLy0df35+9K6icmbIIGhq/lCH0Sc4WCYeixq7q6icWTIImpo/5DP+BAfD1IN1I3cVlbOWbH1QwGGfeIKDIbRIAbuKyZlDNkFh6/OgoImHe0fuyhK3cjCXIDD/wwYDmZ16PHzkrmy4uF46k6Dp+UNOuNUFhebMkEdQ8FFavYlF1p/UggLnD+EnOBgizjtBgmDOeuSoQYGP2JGgUgWf5sMEReTMkGUcFH6UYKmCB4ohgubMbMogKGb+EC7V+PPzo3c1a2YTf6wCKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQYDsghY8hjlqMmGup2usUoP0NKhoKAjwfIKaiWW3y++i3eny3H77QzY++Z7Xq7tFznr5QwqqdZzPN7nlLF/xpSLqqWdSoY16VY7Uh5nBpt4ItkzcmoIqURw5h+l2kRPedfW4XeSzY+UUMvmF3SKn6tZyIo980v7tchLx1CR58bWej6ffUVC+/myjtoJMQsriwmf1ryhIP6q6/vZD/dG8vV39rV6CK78wW/TG8uVDT8ETFUVNDms+VETpSQR3olpBNqFlTzbWrChIdymiMG1TcLqOulB1RW/RJVNervdWWfuhIzadlIlqY7cJLTe0pqDmfD8gSPQZ3/791gqqfEHim54g2Szlo52dqFaQSUiv7nmYPsielHqCbMEjapBoQn+K0G5UvwYZymUPW1u1D2qOZU+Q8lE7TayZHu31QV1Bt8t32ZM7UU0P5p/xF57/1z6LyQM6WIPEiPvcqqvUCcw7i3UFifP7yY/69f76UzSqq01IVaW4xT09Vh8HqZPPQB/08uGp64+DeoLqohtV/qz53YyDdNU60uu99wkFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFAT4P2dIjFNu3Tl0AAAAAElFTkSuQmCC" /><!-- --></p>
<p>As we all know, the function of plot is one of the simple praphing methods.Thus more advanced functions are needed to modify the images pet by function plot as function par.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">opar&lt;-<span class="kw">par</span>()
<span class="kw">par</span>(<span class="dt">bg=</span><span class="st">&quot;lightyellow&quot;</span>, <span class="dt">col.axis=</span><span class="st">&quot;blue&quot;</span>, <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">4</span>,<span class="fl">2.5</span>,<span class="fl">0.25</span>))
<span class="kw">plot</span>(x,y,<span class="dt">xlab =</span> <span class="st">&quot;Ten random values&quot;</span>,<span class="dt">ylab =</span> <span class="st">&quot;Ten other values&quot;</span>, <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">pch=</span><span class="dv">22</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">bg=</span><span class="st">&quot;yellow&quot;</span>,<span class="dt">bty=</span><span class="st">&quot;7&quot;</span>, <span class="dt">tcl=</span><span class="op">-</span><span class="fl">0.25</span>, <span class="dt">las=</span><span class="dv">1</span>, <span class="dt">cex=</span><span class="fl">1.5</span>)
<span class="kw">title</span>(<span class="st">&quot;How to customize a plot with R(bias)&quot;</span>,<span class="dt">font.main=</span><span class="dv">3</span>, <span class="dt">adj=</span><span class="dv">1</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAqFBMVEUAAAAAADIAAFoAAPUAAPoAAP8AOjIAOn4AZloAZp8AZuo6AAA6ADI6AFo6APU6APo6AP86OgA6Zlo6kJ86kME6kOVmAABmADJmAP9mZjJmZlpmtp9mtuCQOgCQOv+QkDKQkFqQ2+C2ZgC2Zlq2Zv+22362/5+2/+DbkDLbkPrb2+Db/5/b/+D/AAD/tlr/tvX/237/2/D//wD//5///8H//+D//+X//+oyZBY1AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAMYElEQVR4nO2dC3vbthWGaS+qy6ntuo1W2kZ2Yy5bqC2hlMl0//8/Kw7u4O0AvJM63/PEkUTgAHgJHIAkCEQFqVXR3BlYuggQIgKEiAAhIkCICBAiAoSIACEiQIgIECIChIgAISJAiAgQIgKEiAAhIkCIyoD+9TncBhYnj47hRrmyuxevpEQKcLAcI4+kYhMyM/lpzFr28EV8cAG9Pqnf/YXHqRTTW+m9A6QxKUhBHCzFYDAcQtfD0QnTmLVcHXABXQ9xYBG6xfFUGUhrUvxgJUQqC/r6BB9en2LPWnA9JOKDC+iyTzzy7apLHG/Tcel7S1L8YDnE65OqLbyu8D9eZ5SjBEUCrDAD9ZF9en0ybVYEZt+BujAtzgBv3ElNHJbHyz5ibRv+QkjIND8sQ/DPSY15IVa5M25SeQhtWialSqsqxf1nSEEcZDFy2/r1oAxDVRL5Z2EymaDgaWVAlcpUMyiJdlUpDwgFs7NsvotwPJlcFbgSJ7v7FT7efYxkYhDretCAZFB9AuR37RfS+9/Vd87BmE6tcvBcq7zDP3FQuhztWXLFKjX5lxkUZ/noZECXyjiqiHkt0yjhV1aWhFM9qpPAE4EzIFwat5qKysrCl+OkqmLFMn/aEeaRFVSXQXgJHYgZgaylEKpiWlG87CHX36u888iprMfHwuqmjI/mbpzHT2U2jjJVKwOmVA6g3LCIwWYi8hCrNFSFlSBFa1ZnqRxHkhW/iILIpDJRK1RQafX1vxamQmVPmoLCG9PGeUDsPPog885TkD7apc3rlarJskgyATh3PKKdgczUPQMoMRXILo/KjXFzjgvitTGuiSPjCfuZbviQOMRTzsj4CVkGUzkSZUo4L2MakjI5SR/+J/POU+AH5VlNVYFUnRHFFhVBJiAaflzKgCqV5YMOiRks5VZ5VOs1bk4nEKvUebsuxbH8lMAkM5QK96GcUWScA+if+iTlqjCJVQb5XZ9e9isDcpV55ynwg07iJq+Sk2h50krGu5GkkgFRKlPs6CnRvDl5eUQ7Cfkd7ImaC95BJgw/luOIQIKKdOdHHivW596WjJprpy3OtjZlmzYZZZ0w+/Iq8y6cIh8mWonbZ1mYz6z2xxuaOAM6A1apjI3o6TtdceUwCoyyE23qyVG2D+kGebbhR8h1JY5oW7nJacrdo9WihIuKFSAIzzqSRCcXq78yPWVaD03Ayne8+Ym8ixTMENDUNOWtRRWSg4OIn9IoUV2/zoApleXGotT051D9j0W5l891x8kr493fhauVgcpx5AlJ9ThGelrdrMrmlRvV/cT3B8eUbVqHEj2RbLbSD7GDjv8z1bGQXYMcJv5jL7tyeXFiMqBLZXXsUWpdrqWRMGHyWxSFHvBxBxvL86hHVKU4auilMclezlAom4cMxnq8wl2L6nSSwgmfRlbVSHRtVK727sVJvLD9J68nclilh5KKp8mALpXpD5Z2u6P7hS2ugLsKxt0tDVD5anxI6QtQVMaLLQ1Ql/st/ko9jbPuUX9eGKAR750w5X7V0+4ulwZoeSJAiAgQIgKEiAAhIkCICBAiAoSIACEiQIgIECIChGhQQNHG1A3Q2/Nut3sckutS1Q3Q2/O7l+K0++vw+VmcugH69jPctzz9ZbwbfzX6qjVlqh2bGNf53Xi3jmv09f9SqwH0aeIatDZA56qX1m5/DK0N0LnBRxMgoZr6Y1kcQ+sCdGocBREg0GnX+HySABUwDnpstziG1gTotOOqHQcRIB+LY2h9I+kWi9sRAUJEgBARIEQECBEBQkSAEBEgRAQIkQMIZoBm9nttnS1uRw6gFF6Ki4u01xzKDQO6ylevPOeBtlncjsqAYA53RoCM3CYWw4sNPacpbxnQ9RDJd5v6WtyOqJtHRIAQuYBYG3v4kvZbRGHLgGC9A3DSvQhtGBC8AAQrwlA3b6k8DgJANFC0VFeDfF85a7G4HdX4oMz/rc5Gi9tRpRezVk3pYXE7onEQou6Avv1U78o3DEivOeLTi/3xS8MEvA0DkvLq5c+73dSAlvNs3uOO4nn3eJ4c0GJmd/gNFOsA3cgkTr9LjdutQZ53FG8QkOrF/K40bhBQmAgQIgKE6NYAmaX7/EbSMwBazEBxAIvbEQFCVH5w6N/EWi1uR+XZHZm9Wmp3i9uRO1BMihyeatA9aaPyU43Lj5/5v54Wt6PyU43r+5cVAfo6ftfv+CC4jE+TFTUxPXYcb/DodvMprJXdrxPbNqDBLE4kAoRoYkDDLBG6YUB8UeXeKxVvGVAhV5teyCROvBOfxwdZe0R0tjiI8OLfeg3CAU07UOT7gSzIB01QP3AtuRdbHKABLQ4iAoSIACEiQIgWB8jeKKGnxUHkAJqgR69V+Y7iQBYHkcNkrurkTgMeYkeUka7FlgAoaI5iu8XhtQRAITo3LK9EgIRgEc6GhTg3Dcj3fbG350f291PtQopbBuT9vljbYrcbBuT/vti3v0Hrqk6AGXGW6xIA+b8vJtxPvRPaMCD/98XmALSAkbT/+2JNTcw2thFVejGve4ozOOm51G0cNEM3P5dWM1CcSw4gtYu3x7XYaepLjbnkPlld8NV8f3XrBstT8PprcYAMmC4DqSXfMBtIZoZ+X0B9Vw6yLE6ngAf4vQH1nQBsWZxOAQ/w+wEKflej1eJ0mgzQYNowIJgCDFrV6i9zAFrV+kHTAUqND1rTClRz1KB+WjCgviPpYbQ8QL3utLmALvvVrR809p1G944i9z60ApWtutkd63mZZQLVXayuahw0tqgGISIfhGj1vdjYWv04aGwRIEQ3DchnjHnbgLCrlIIAEaA2BQMKeLK6iaVKgwGFPFndwlKloYBCnqzOsFTp8AoH5H2NMcdSpcMrFFDYk9UbBMS3hvLW5Gu5Dq/wJrboVfCG17gj6Q0A8lEwINgo/BE+3CSgkL19bhFQ0N4+NwgobG+fGwREe/vUqK4G0d4+lmp8ED3VsFXpxeiphqubvmHmIwKEyJrlOsQs8oIA+VncjggQIgKEaPUz7ccW1SBEBAgRAUJEgBDRSBoRAUJEgBARIEQECBEBQkSAEBEgRAQIEQFCtFhAc61ZVlZHQG/PapJHvcX+WsRi0kVXQG/P716K027MJbrWDWiCRd7WDUho1HUUtwDoE9WgVp2rXnrAWa7rB3Su99FDafWAaurPoFotIDnL9TQyn/UCEjrtBrrF36h1j6S//fw4eE4Wqm6AoJ3tGhcrlZZXrl6AfBRqNDD8yObd8AQICU+AkPAECAm/sZuAw4sAISJAiAgQIgKEiAAhIkCICBAiAoSIACEiQIgIECIChIgAISJAiAgQIgKEiAAhGgNQyxy0JjWtzFhV81bc/W0XNXkfAVDbHLQGNa7MWFHLDou9bRd1eR8BUNsctHo1r8xYVtsenX1tg6p5H80HBZzmlpUZywqGH2DbimTlfTRA9XPQmuQNqGWf4L62jT5NUIMC5xD5FqJtp+m+tk0EO+8jAQqdg7YkQG7exwEUPAdtQU2slPdhAYXOQUNXZiwrvIcMBVTO+xg1qMMcNN9ChHfzgYAqeR9lHPQYHMe7EOEDxSBA1byPAMhnDlpZ/oVo3oq7v+26vNPFKiIChIgAISJAiAgQIgKEiAAhIkCICBAiAoSIACEiQIgIECIChIgAISJAiAgQIgKEiAAhIkCICBAiAoSIACEiQIgIECIChIgAIZocUNiWt47ykO3PhtrJYJYadPmh00ZvBAjR7QHKWEtLoCwfWLsT5bm+/wiNL434D+YIC3r3EQDlIs5lD0eSy15GvOzhP0Coox45I/5HJlRA8MBtHOcElLHiXPYJKwP7kInqcT3APp1pLI7qIymDlsMWjLCr+fUQs3hwBFiKrWHFfvDpwxcT1QBSCXGKQfuizwpIbAuc33/mH0QVED9e37+IH9QRcTC9e4HtTnlFgfIW8g+PCJxYcCuqBqQT6rKL7IyAhEthhTFNwXIdecTrijgiSsa5HAuDzPwREaWTUlF1bJNQOKE5Acn+vgYQ8xn3/94bQJkLiP1SAQTNErbRtaJqQCoh1hDX5IN0p1QBpAseUINYE/oPC21HdWuQUhq2meysPkieywogziO3mpiAkDk+qAzoevgNPLkVVXkwt8cP7P/n7sXghNbWIDbiTgy6jHdgTi9WBsT699iN+vr08IU1qqNOiFeloNHUAsZBvPOp8UF3Lw666jioAiiPylHhsuaDGgeJqhW8XTNdrCIiQIgIEKI/AR6tspNBeIBSAAAAAElFTkSuQmCC" /><!-- --></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(opar)</code></pre></div>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;cin&quot;</code></pre>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;cra&quot;</code></pre>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;csi&quot;</code></pre>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;cxy&quot;</code></pre>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;din&quot;</code></pre>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;page&quot;</code></pre>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-3" class="section level2">
<h2>Question 3</h2>
<p><font color="#000000"></p>
<p>In this informative age, processing big data is common, thus we generally need some function with the cycle structure to find the potential rules. And in the code of R, some of the cycle functions are different and unique. Those we have to study this.So how the procedure of the cycle does?</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-3" class="section level2">
<h2>Answer 3</h2>
<p><font color="#000000"></p>
<p>In the R software, the function with the form of multi function apply can avoid the ues of the direct cycle structure.as the example follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x&lt;-<span class="kw">rnorm</span>(<span class="dv">10</span>,<span class="op">-</span><span class="dv">5</span>,<span class="fl">0.1</span>)
y&lt;-<span class="kw">rnorm</span>(<span class="dv">10</span>,<span class="dv">5</span>,<span class="dv">2</span>)
X&lt;-<span class="kw">cbind</span>(x,y)
<span class="kw">apply</span>(X, <span class="dv">2</span>,mean)</code></pre></div>
<pre><code>##         x         y 
## -4.982899  5.934897</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">apply</span>(X, <span class="dv">2</span>, sd)</code></pre></div>
<pre><code>##        x        y 
## 0.105836 1.817203</code></pre>
<p>More information, the function of lappy() which has the silimar syntax with the function apply can work on the list objects and eventually return a list object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">forms&lt;-<span class="kw">list</span>(y<span class="op">~</span>x,y<span class="op">~</span><span class="kw">poly</span>(x,<span class="dv">2</span>))
<span class="kw">lapply</span>(forms,lm)</code></pre></div>
<pre><code>## [[1]]
## 
## Call:
## FUN(formula = X[[i]])
## 
## Coefficients:
## (Intercept)            x  
##      -18.93        -4.99  
## 
## 
## [[2]]
## 
## Call:
## FUN(formula = X[[i]])
## 
## Coefficients:
## (Intercept)  poly(x, 2)1  poly(x, 2)2  
##       5.935       -1.584       -1.574</code></pre>
<p>In more flexible situations, to accept the vectors or the matrixes as the main parameters, function sapply can be a good alternate with the final return of the form of table. Here we do not talk these.</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-1-1" class="section level2">
<h2>Question 1</h2>
<p><font color="#000000"></p>
<p>Most otfen, it is necessary to split the picture or know how the device splits the windows that we need.Apparently??the function split.screen()could fits our needs,but with the shortcomings of incompatibility and unability to be performed on several devices.Thus,how to know that the route to split??and how to code?</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-1-1" class="section level2">
<h2>Answer 1</h2>
<p><font color="#000000"></p>
<p>First,we can generate a matrix with specified construction,then utilize the funcyion of layout() to display the details of the split process.The examples are follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat&lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">2</span>)
mat</code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]    1    3
## [2,]    2    4</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">layout</span> (mat)
<span class="kw">layout.show</span>(<span class="dv">4</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAV1BMVEUAAAAAADoAAGYAOjoAOpA6AAA6ADo6AGY6OpA6ZmY6kNtmAABmADpmOpBmtv+QOgCQOmaQ2/+2ZgC225C2///bkDrbtrbb////tmb/25D//7b//9v///8SIqCzAAAACXBIWXMAAA7DAAAOwwHHb6hkAAADE0lEQVR4nO3cwVITURBA0QFBVBQFiSRh/v87BXcu9C7sKl9R535A16uTnpfJJtumaF+qxY6zA4oARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARVNAp5vHkTkzQMdtu7ifGDQGdL69XAjo9P5+Pw4daAbosH1caoNeOn+eWaEZoB9Paz1iLx2unkbmvM07aD9dL3YHrQa0L3YHLQi01h0EKFsK6PXxOn1wSf+5w7aNvZf5qfH3AEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKNmX/+zP6vcWO4xGrAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAohmg0/W23Q3M2ceATjePM4NGgM63d/vx4v7fB+1jQA+XKwEdr572568zKzQDdHy31Aa99rpFE40c5/nb9+WADi9bNNHIcQ6f1rqDXjqMPfMDM85fnlYDOgxd0TNAD3eLfYvN7c8I0Pn211+wDt2JE0Cn91P780bfgx4mP7GRKYsBTbbYcQBVgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBijZFPwE0za13Amcg8QAAAABJRU5ErkJggg==" /><!-- --></p>
<p>It is also feasible that fuse the function layout() and matrix() to simplify the code and get to our targets. As the examples show below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">layout</span> (<span class="kw">matrix</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>,<span class="dv">3</span>,<span class="dv">2</span>))
<span class="kw">layout.show</span>(<span class="dv">6</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAXVBMVEUAAAAAAGYAOjoAOpAAZrY6AAA6ADo6AGY6OpA6kNtmAABmADpmAGZmZjpmkJBmtv+QOgCQOjqQOmaQ2/+2ZgC2/7a2///bkDrb////tmb/25D/29v//7b//9v///97PZtAAAAACXBIWXMAAA7DAAAOwwHHb6hkAAADQUlEQVR4nO3dy1ITURRA0aig4BuNdoiQ//9MyZCJm8EpuIVrf8CtWyt90qOus9spOi3VYtc5AYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYCiKaC7z79HzhkD2q5nzhkCuvv4Zi2g4+VSQPc/fq31BN3ffFkKaLkR264XG7HFgI5XB0D/ajt/KDgj9DqBTsu9xQA9W4tdB1AFKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFD0hO1R/30v/Rs9brHrGLEKUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhQBigBFgCJAEaAIUAQoAhTNAN1/3+0uBs45TQHtd7uhfV4zQNvFg9G3gYOGgO5vfk4cc25uxIZ2xYxc5/hhaJHOINDUcp+R69y+vRwb+anlR5+GHuqZ6/w5nLahkZ8BOr6fGvqxt9jUyI8ATc78xCG3D/O11BO0P3/7utCf9PlC9os9T4AiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYCi3q6ll/6NHrfYdYxYBSgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKhoD2u93MIpQhoG1u39kI0O3FWtuhjleHsX1nr3J91tDio3NTQNtSI7b/utx+sYdLLbTCb//uMHafV7nC74yz1H6x8z6vlZ6gyfuMvean9nmNnLLZL/ZcAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYqesl/r/+4vYZME64txSzsAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m&lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">3</span>),<span class="dv">2</span>,<span class="dv">2</span>)
<span class="kw">layout</span> (m)
<span class="kw">layout.show</span>(<span class="dv">3</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAASFBMVEUAAAAAADoAAGYAOpA6AAA6ADo6AGY6OpA6ZmZmAABmADpmtv+QOgCQ2/+2ZgC2///bkDrbtrbb////tmb/25D//7b//9v///+vIR/xAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAC20lEQVR4nO3cS25TQRBA0YaETwghxnZ4+98pMnNyJ6XIap2zgFL7ql97VmsRjrtyZ8c5BAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoCBYGCQEGgIFAQKAgUBAoChalA16+vI3N2DfT29Fmgd5zWdzfoPb/PPrEgUBAoCBQECgIFgT7KnR1HoCJQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKewa6rPXpZWLQpoGuX16Oy9Bixy0D3bz9mLlC6+6M/KzjOD2cR+bMnWjIzHGuj96g4A0Kg2/QyJwxAoWJ49w+r+s3j/T/ndYa22+9ZaBBAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQECgIFAQKAgWBgkBBoCBQmAl0fVzreWDOsWmgt6fn4zK0VW3LQJeH8/Hn58wV2jLQze0WTdg20Nhix5Epc6YCncZ2hs2MGTMU6DS1+HLTQGP3Z9NAt/XEU7YM9OvfGl//Yh/izo4jUBEoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaAgUBAoCBQECgIFgYJAQaCwCH8B232KufdIPmwAAAAASUVORK5CYII=" /><!-- --></p>
<p>In addition, options in function of layout(), such as widths and heights, give the possibilities to alter the split width and weight.So, the screen could be separated into plentiful shapes even be some complicated or weird.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m&lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>),<span class="dv">2</span>,<span class="dv">2</span>)
<span class="kw">layout</span> (m,<span class="dt">widths =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>),<span class="dt">heights =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">layout.show</span>(<span class="dv">2</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAQlBMVEUAAAAAADoAAGYAOpA6AAA6ADo6AGY6ZmZmAABmtv+QOgCQ2/+2ZgC2///bkDrbtrbb////tmb/25D//7b//9v///8bolO5AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAC30lEQVR4nO3d2W0CQRRFwfEOXjCLyT9VLwH4qCUjD+o6AVxB6QFfaJZF0fmqu/TLB1T7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGJ/KqDj/bJsB/dnAjpttufDzcvY/kxAh7v9+eN57ISmAvru+4qG9mcD2n1d0dD+ZEC727fB/bmAdoNf0bMBDd/PZEDHh9H7mQzo9ec/un7F/nYfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGIfUOwDin1AsQ8o9gHFPqDYBxT7gGL/6oEu33+/xXUHKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYoARYAiQBGgCFAEKAIUAYrWBnR8HH40yGVbGdBpM/7slMu2LqDd8uSCfut97yMWAYoARYAiQBGgCNC1BSgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQBCgCFAGKAEWAIkARoAhQtCj6BOSsJIjMeJ3DAAAAAElFTkSuQmCC" /><!-- --></p>
<p>These are about the split of screen with the help of function of layout and matrix.</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-2-1" class="section level2">
<h2>Question 2</h2>
<p><font color="#000000"></p>
<p>In the analysis process,to make a intuitive impression, the images are indispensable. So knowing how to code to display the relevant images is essential.Thus??here gives some simple examples.</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-2-1" class="section level2">
<h2>Answer 2</h2>
<p><font color="#000000"></p>
<p>The simplist example of the plot is followed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x&lt;-<span class="kw">rnorm</span>(<span class="dv">10</span>)
y&lt;-<span class="kw">rnorm</span>(<span class="dv">10</span>)
<span class="kw">plot</span>(x,y)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAATlBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6kNtmAABmADpmZmZmtv+QOgCQ2/+2ZgC2/7a2///bkDrb/9vb////tmb/25D//7b//9v///9g34SVAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAF00lEQVR4nO2d63biNhgAvZu03bQldN1lA+//osUQ0u0WaeTLZ8tm5lfOif1JHoRuSHJzkizN0hmoHQUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQUA/Qcddc+Hz16Ds1EcvQW3z5frH4fbH9ukj6Lj70NI+fQvITI30EfT28uftz8PDfMksQUDPOui9CCXroGY1hAg6f8mu0ZPlZ/leQ+GzBwmaO9ywDJRk4lEFNaW5iBbUJloxBSWi9K77gqhVUFS4YRmwDsrmYNFWrD0nf+kK1VoHFRMjqP30eu4KPZ8UdJ/rUOO4O3cTFXSP22B1//RNQff4GKzunxV0l5uW85BMQXe5DeePOwUFp7s0CgIUBDyAoHEj4+0LKh2VZm6f9sJ8lNmnO4rnNbL3T3nhIuEoIQVBQgrKp7SeOigqXDYpW7FIFAQoCFAQoCBAQYCCAAUBCgIUBCgI2Px80FgsQcBaBc1WBlcqaOQkT8+Upr1wjnBjpwl7JzXlhafuJ/lu8cKhyez2UdB1E8IPuzaGh0vdvW5B72pSezUeuw7qBH3/5SIotdvnsVux0hK0hv5ikKCur/x8ulXXiXDzfU9GENXMnx19es3syGxmrWlHsGA/SEEQTkEduVWuD10HpaL8Z7rjYVux+cPF0V/QuX16njHdpRlSgrqtPHCgAh5NsWlBJ3TER1NsXdBVUap8FBwssHFBh8t+uct+p3sUHE2xZUHdSOtqJvX0j12C3l7Ogyyg4GiK0nSXJqgftIKjKQqxowgoCFAQoCBAQYCCAJe/AJYgQEHARgTFfWe3IShw+n8Tgu79gDRVmdqqoMnK1EYFTfej5CYE/b+8bF9Qzyrk58s3L2h0FbLxOmiCArDtVqyihR8KAuoUVNHKmFqnO5afELkt0Sm+fuLkpw03Pc2pZzl+MEEftaCC7hMtKG590EyVU7CguPVBszVvoXVQ3OqOGTtIka1Y3Pqg+XuQlqCiFKe8sCNufdDsXeygVixufdDcXWz7QYCCgGhBHtkene7SKAiodT6oGixBgIKAoLHYv18lW7F7JF+nMSzckoRNmMGuxEcXdB6lJo59GRYuGyq0SVx/JR08vl+9oOgZokhBP8ybTREuF0hBEEhBaVZcB80jaMWt2EyCYlmsFVsNCwnqE7mGCxgFxYcYGrmGCxgFxYcYGrmGCxgFxYcYGrmGCxgFxYcYGrmGC5jVjA2WQkGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgLCBF323sH53d9/S6/HOjQNHS6bu70sAwVECTruzo/X5k+Bf3tJL1g7dO83yRvK3V6WgRKiBF1fkpRakH/hkNnaef31dp97vNztZRkoIrYOypWBQ/MltS2v5PGytxdloIxYQfv8I2QE/fqa/T/cXpqBAkIFpV+S9P7/ZPavnzx8/iyIMlBApKADVZHRgjADBQQK4o8v+Cs2QfkJEfT+co42mb2Pt3eMqaSzt78nM4GfuBLUworqjvQTFjTzJKgkAwXE9YMKPr7MExZ0FPOCijJQQJSg9rqOa3gt2/JQIyuoKAMFOFgFFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQUDtgrpFZm8v0yz1GUTtgrqlZqmXeMxC7YJOh89//T52EdQYqhd02k+wlncE9Qui83WDqV7QcffH6O0EY6heUPv0d+bcxnhqF9S9sqtg00octQvan5t4PAc9ktoFLY6CAAUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCDgHzbXLxAovR4nAAAAAElFTkSuQmCC" /><!-- --></p>
<p>In fact, changing the image characters and making the iamges more artistic and analyzable, through adjusting some parameter values, are practicible.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(x,y,<span class="dt">xlab =</span> <span class="st">&quot;Ten random values&quot;</span>,<span class="dt">ylab =</span> <span class="st">&quot;Ten other values&quot;</span>, <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">pch=</span><span class="dv">22</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">bg=</span><span class="st">&quot;yellow&quot;</span>, <span class="dt">bty=</span><span class="st">&quot;7&quot;</span>,<span class="dt">tcl=</span><span class="fl">0.4</span>, <span class="dt">main=</span><span class="st">&quot;How to customize a plot with R&quot;</span>, <span class="dt">las=</span><span class="dv">1</span>,<span class="dt">cex=</span><span class="fl">1.5</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAbFBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6kNtmAABmADpmZmZmtrZmtv+QOgCQOjqQkGaQtpCQ2/+2ZgC2Zma225C2/7a2/9u2///bkDrb/7bb/9vb////AAD/tmb/25D//wD//7b//9v///9ulbq9AAAACXBIWXMAAA7DAAAOwwHHb6hkAAALHUlEQVR4nO2dC3ebNgCFSeNsdjtnXdi6srVkjv//f5weSAiMuJJAgPH9zpmbGT0/hCTbEhRXMkqxdgG2DgUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFATwCfp4LZ5/6H8PIen8+T2xACiiKUgCQ1E92Zmg8nBVPL21R+YRdDl9ShOEI84qyJudDqoP5xBUFomCkiOGcCsIZKcPpwl6fymK4mjeb46WTQFKcUylbUJZannkbPO+nNTRSr6rE+hHFCEOtYojAsmCqqzEuwpVchnJOWkqalslkdG3k07U1MEmbrO7qca/8q/msBRU2iQDBVW6gOKtSuYgCiziN9VtM25D2dIqjl1BTTDxxm1Eo8IG6QhqA9iWUDvqnCzlG00d2sRbQf1qfHQEFU4WI4IMB3UOlBtRvfcXkaY+x+Ldsz1t3zuhbOGPMn+lwgqSjaQ5dhtR1v94tW3o6FwnpbJQNjU72nLq8pk8S1WzWhZbR3VL1V5ivWrooOYSc6sRJqhJWP4jDhxF7C/q1WSnjzuh3Ler9piqmCiQPf83EUUI3Ux01RxBqs7tRep2JpVz0ZXtJe3WWv/TFq1XjZ6gs80oTJAppOq+SvFG+fTHy0H95dSzE6pJw15tjqAm6eNgRN289GtHkG6Iuj9xrylT0E5RdFoqaq/szqlxq9EVJNPWRQjsg0zgWrqtZZcmOsJPf7+aK8y2+zbU9dodAe21dLzafuY8ENErSJ9X0+UUzkmwsVxBMi3TfbWlcgR1qzFRUOcsXE7Cu+jRxKvTM4a1INtXmHEsuAWJv81oenbLaZXHtqBuNSYK6lzH4p0vzfXbq76nD6plfjpX3TN//P527XQK3T5oSJB41Yk6Kq62YfQ66ZA+qFeNqYI641NV6CvNmfHoNIdGMdX8nBFJd9K2SDcRPYKqoh0wxV+6E29akGolbiftH8XcOWCnGm5pUgSZuYRp5SIF/dqacKYz7Sku3YmCnRNVbS9yE3FYkI19MB1Yv492BXWmUJ2y184MqlMNHbQuzEQxWpCuoO17de7OSNvMPJ1QrTndoOSh53/02KmC6dj9iMOCbNd8aJS0c1H5v09vnRmH6HZVgM5M+uyW86Yadk6t0goRdLfM/emOggAUBKAgwO4EzQ0FASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIEAr6HI6Ngs8iUMrqJTLcA7XMmhrxuNgBV1OcrmV+A//tK1W6BwHDxV3Q5KgUi72goI+Xp/exMV4Dy3t53+Gn70j8YLEtXU5yVWzsOLvL3K1Fha5AWYVpDYufLyGNoza3RKzVWYVFEk52IKiL/G8rCioHu6lNyLGMK8gtfi4HB6d+tSePnrPgkSnUslOOsSQp/3sWtDHq5hIy9XmAaNT5fOzZ0FyHiQFBUwUq+5+kqR8l+FnS+9Iegsq4Yex9xf/VbgxQX6S+6CR1mFwtlqk57s2iaOYp9oZ8l2b/BPFRZILw9/T+HksQd6xyk/KKNZ8SJj0IXTHghoCRvlZ8p2TRQVN/EbxAQRNa0IPIGjaF2H7FxTwjeIs+c7JsqPYtJ99dixoHjhRXDa5fFAQIFKQe9+n+xvFUsjdgt5/HdZIQRrv7fb2LMhcZgGXWO0NtWdB5fOP6tD8rjxOXRx9H0gmC0oZslNImSger7X8VSNopugVNPWX1ZRJXwppv2rIrtfX/XbJ14I2K0j+qnH5/EZB3oDyY3x5nHqJhebrY8OC5Fdlzc0KMQ8pKAavoKnjz+4FTa3adgXFfVP2gILU/fUmL6vbs6Brc1O/aV+5Tha02Zl0w8frtK878p/7ebjbFrQUKTPpTfRBS5F7FPMmt19BM+VLQSC5/OPPPPBXDQAFAdK+D1ow37VJ+0ZxwXzXJmUZ8BwbwHYsKGaNYu2dUu5YUARyL51nPx0FXU13PryWcdeCQveLje1Z3bOg4P1i77/Iq2u4T9+xoPD9Yrr7Ge6EMuxZzfMVWto8KGy/2Kig0HzDyfMlbHoLCtgvtuwlNixoarNK7oMC9ost3El7BE1sVomjWNB3igsP89sRFEziRDGxt71DQeYRqJHJJfa2mxFkntSVbZXrvQsavhnHjPneuaCwWy5MyffuBeX+wixV0GDfvryg8DsHpeY765R4hYliyALgSfkutXAjiC3u1RgVtNSqDsMWf/YZF7Rw80ropD/rqV++Tb2jjeR+BK20qXfrgsq2D5o0HdqtoLYFLZRvn+0LWjjfPvcgSD0Sea37B92BoLp5VnrQdHH2LZnbF2RWdwRtZpl/S+b2BZkPqyHzoAxbMrc/k45oQTm3ZC5F7j7oEQVFjWLZ9qwuRe550EO2oBgoaJjKfFZbRlDGoW0fLSjj5IiCABQE2OIvq/FsS1D2X1bj2ZSg/L+sxrMxQdvbirApQfl/WY1nU4LUo6GWyzeITU0U7/h+0inc3Zf2S0NBgMRVrsHP9pmc79okr5OeOB3asaCYZ/vMkO/apE0UA5/tM0e+a5PegvBejVnyXZvkPijkV42xx4iG5rs2iaNY4F4N/2NEi6jcA8PlCJZzHoR3+zy4IM3Yhrr9CUr4rmOOx4juWZD3MZAxue9YkO8xom6ioZmvEiyzoNH2s09BYSvtm19W/Y8RdRMNzXyVYFlbEJxMFpsmu6Cxx4jujwRBY48R3R/J86BH4W4+Xa4FBQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIkEPQyPqhPvihpf7750cnFVUyQwZBY+uHenjv4GAZuS1qbFJRJbNkEDS2fqiL/w4OhrEb60YmFVWylmx9UMBpH7mDgyG0SgFJxZTMIZugsP15UNDIzb0jk7LE7RzMJQis/7DBQGHHbg8fmZQNF9dLZxI0vn7ICbe4oNCSGfIICj5Li19ike1nbkGB64fwHRwMEeNOkCBYshtytKDAW+xIUK2Ch/kwQRElM2SZB4WfJVir4IliiKCUlU0ZBMWsH8K18t8/PzqppJVN/LAKoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQGyC5pwG+aoxYS57q6xSAvSy6CioSDA4wlqFpZdTr+J607X5/L5q7z45HNez+4RuerlqxRU6zjvL/LIUT7iS0XUS8+kQhv1rBypF7OCTT0RbJq4JQVVojpyDdPlJBe86+ZxOcl7x8olZPINe0Qu1a3lQh55p/3L6SDiqUXy4m29Hk8/o6B8/tFGbQWZjJTFiffqX1CQvlV1/em7+qN5erv6Wz0EV75hjuiD5dObXoInGopaHNa8qIjSkwjuRLWCbEbT7mysWVCQ7lJEZdpLwek66kK1FX1E10x5OV9bZe2Ljth0Uiaqjd1mNN3QkoKa8X5AkOgzPv310gqquoLEOzeC5GUpb+3sRLWCTEZ6d8/d9EF2ULoRZCse0YLEJfRNhHajdluQoZx2s7VF+6DmXN4IUj5q5xJrlkd3+qC+oMvpi+zJnaimB+uO+BPH/6VHMXlCB1uQmHEfW3WVGsA6o1hfkBjfD92oH6/PP8RFdbYZqaYUt7nnhsXnQWrwGeiDnt466m7nQTeC6qIfVX6s+c3Mg3TTWu/x3o8CBQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQE+B9EZo3Kdk0rlQAAAABJRU5ErkJggg==" /><!-- --></p>
<p>As we all know, the function of plot is one of the simple praphing methods.Thus more advanced functions are needed to modify the images pet by function plot as function par.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">opar&lt;-<span class="kw">par</span>()
<span class="kw">par</span>(<span class="dt">bg=</span><span class="st">&quot;lightyellow&quot;</span>, <span class="dt">col.axis=</span><span class="st">&quot;blue&quot;</span>, <span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">4</span>,<span class="fl">2.5</span>,<span class="fl">0.25</span>))
<span class="kw">plot</span>(x,y,<span class="dt">xlab =</span> <span class="st">&quot;Ten random values&quot;</span>,<span class="dt">ylab =</span> <span class="st">&quot;Ten other values&quot;</span>, <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">pch=</span><span class="dv">22</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>, <span class="dt">bg=</span><span class="st">&quot;yellow&quot;</span>,<span class="dt">bty=</span><span class="st">&quot;7&quot;</span>, <span class="dt">tcl=</span><span class="op">-</span><span class="fl">0.25</span>, <span class="dt">las=</span><span class="dv">1</span>, <span class="dt">cex=</span><span class="fl">1.5</span>)
<span class="kw">title</span>(<span class="st">&quot;How to customize a plot with R(bias)&quot;</span>,<span class="dt">font.main=</span><span class="dv">3</span>, <span class="dt">adj=</span><span class="dv">1</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAqFBMVEUAAAAAADIAAFoAAPUAAPoAAP8AOjIAOn4AZloAZp8AZuo6AAA6ADI6AFo6APU6APo6AP86OgA6Zlo6kJ86kME6kOVmAABmADJmAP9mZjJmZlpmtp9mtuCQOgCQOv+QkDKQkFqQ2+C2ZgC2Zlq2Zv+22362/5+2/+DbkDLbkPrb2+Db/5/b/+D/AAD/tlr/tvX/237/2/D//wD//5///8H//+D//+X//+oyZBY1AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAMV0lEQVR4nO2dC3vjqhGGlXR9clT3XsXbdp10o267cpujeOso/f//rAx3dBvQxZLRfM+zWduCAV7BgCQESUXqVbJ0BtYuAoSIACEiQIgIECIChIgAISJAiAgQIgKEiAAhIkCICBAiAoSIACEiQIgIEKI6oH98D7eBxSmTY7hRruLuxSspkQIcrMcoE6nUhCxMfjqzVjy8ig8uoPcn9bu/8DiNYnorv3eAdCYFKYiDtRgMhkPocjg6YTqzVqoDLqDLIQ0swrA4nqoD6U2KH2yEyGVB35/gw/tT6lkLLodMfHABnfeZR75dDYnjbTqtfe9Jih+sh3h/UrWF1xX+x+uMcpSgRIAVZqA+sk/vT6bNisDsO1AXpsUZ4I07a4nD8njeJ6xtw18ICZnmh2UI/jlrMS/EKnfBTSoPoU3LpFRpVaW4/w4piIMsRmlbvxyUYahKIv8sTCETFDytDKhSmWoGJdGuKucBoWB2ls13EY4nU6oCN+IUd3+Fj3dfE5kYxLocNCAZVJ8A+V37hfz+7+o752BM51Y5eK5V3uGfOChdjvYspWKVm/zLDIqzfHQyoEtlHFXCvJZplPArK0vGqR7VSeCJwBkQLo1bzUVlZeHrcXJVsVKZP+0Iy8QKqssgvIQOxIxA1nII1TCtKJ73kOufVd555FzW42NldVPGR3M3zuPnMhtHmaqVAVMqB1BpWKRgMxN5SFUaqsJKkKI1q7NUjyPJil9EQWRShagVKqi0+v5vC1OlsidNQeGNaeM8IHaZfJF55ylIH+3S5vVK1WRZJJkAnDse0c5AYeqeAZSZCmSXR+XGuDnHBfHamLbEkfGE/UI3fEgc4ilnZPyELIOpHJkyJZyXMQ1JmZzkD/+Reecp8IPyrOaqQKrOiGKLiiATEA0/rWVAlcryQYfMDJZKqzyq9Ro3pxNIVeq8XdfiWH5KYJIZyoX7UM4oMc4B9Gd9kkpVmMwqg/yuTy/7lQG5yLzzFPhBJ3GTV8lJtDxppeDdSNbIgCiVKXbylGnenLw8op2E/A72RM0F7yAThh/rcUQgQUW68yOPlepzb0tGLbXTFmdbm7JNm4yyTph9eZd5F06RDxOtxO2zLMwXVvvjDU2cAZ0Bq1TGRvL0k664chgFRtmJNvXkKNuHdIM82/Aj5LoRR7St0uQ05+7RalHCRaUKEIRnHUmmk0vVX5meMq2HJmDlJ978RN5FCmYIaGqa8taiCsnBQcJPaZKprl9nwJTKcmNJbvpzqP7Hqt7Ll7rj5JXx7o/C1cpA9TjyhOR6HCM9rW5WdfPKjep+4ueDY8o2rUOJnkg2W+mH2EHH/5nqWMmuQQ4T/7SXXbm8ODEZ0KWyOvYkty7X8kSYMPmtqkoP+LiDTeV51COqWhw19NKYZC9nKNTNQwZTPV7hrkV1OlnlhM8Tq2pkujYqV3v34iRe2f6T1xM5rNJDScXTZECXyvQHa7vdMfzCFlfAXQXj7tYGqH41PqX0BSgq48XWBmjI/RZ/5Z7GWfeoP68M0Iz3TphKv+ppd5drA7Q+ESBEBAgRAUJEgBARIEQECBEBQkSAEBEgRAQIEQFCNCmgJDINA/TxvNvtHqfkuph++a/SL22HhwH6eP70Up12vx6buTVoFkA//gD3LU+/mu/G3/U0CyCht0/z3Tq+nmYE9I1qUK/eml5au/0b0myA3jp8NAESaqk/lsUb0kyATp2jIAIEOu06n0/eHCCjtsNDx0GPncduDRCiYYBOO67WcRAB8rEYjwgQIgKEiAAhIkCICBAiAoQofkDISBnTBgD1X2thIkCIHEAwA7Sw32sbbHFFmhJQDi/FpVU+ag5lxIAu8tUrz3mgfRZXpIkBwRzuggAZuU0shRcbRk5TjhnQ5ZDId5tGKGZAk2h1gGigOKtcQKyNPbzm4xZRiBkQrHcATnoUoYgBwQtAsCJMXN38SNXHQQAoroHiSLXVIN9XznosxqMWH1T4v9XZaTEeNXoxa9WUERbjEY2DEA0H9ON37a48YkB6zRGfXux/f+mYgBcxICmvXv5tt9ssIJ87im+7x7ftAvIbKLYBusVJnIjaAPldalytBo27XTFWLYA87yheD9CoG15j1daL+V1pbBBQmAgQIgKEaGuAzNJ9fiPpzQGa1uKUIkCIVgRINbNxt1ynHtStCFD+8FrYq6UOszh1YdYzkobnPSU81Rh3T3q5sz2H6k81zr/9zv+NsBgvIHiqcfn8QoBsOT4ILuPzjJqYLbebz2Gt7HGdWNyAJrFIgBCL8QKaZonQyQeKy8odKI5/rhr9TXu+2nRUkzhHqsUHWXtEDLYYj6gGIXJH0uSDGpqhF5vAxop0AzfMlhUBQkSAEBEgRPX7QRNZjEf1O4oTWYxH7jTgKXZEiRhQ0BzFfovxaKiTfutYXokACcEinB0LcUYNyPd9sY/nR/b3W+tCijED8n5frG+x24gB+b8v9uP30LqaE2Ain+Xq/76YcD/tTihiQP7vi20UkP/7Yl1NzDY2s64156PRi3ndU1yBk77WrKFh46AVdPPdgEauJFDTzQ4UewCNW4uiJgeQ2sXb41rstPSlxiKAJtnfNGJAI5dcsC3Or2UAXemG2RRudAlAY1cOsiz2a4oiLOKDRk4Atiz2axJAnZVwJkDB72r0WuzXtEWY1/oij31mBjTbQBGmAINmX/1lXkDTqg3Q7OsH3Sag3PiguVeguk1ApgZNYLFftwpoQov9ul1A5/1V1g+atp+ZV+4dRe59aAUqW22zO0a+zDI2S0JrqWVtF6urWAVvLX5qvTVohYBW5YNWCehavZiP1gloMovjRYAQESBEBAgRAUJEgBCtcSQd8mR1k0uVBj1Z3eJSpSFPVje5VGnAk9VtLlUa9mR1g4D41lDe2uBarmHvamyxBgWJALULNgp/hA+bBBSyt88WAQXt7bNBQGF7+2wQ0I3s7XPdq9i2GrTyvX2uuzJniw9ax1ONbi0I6Db29lkS0GQW5xQBQrQQoGlmkVcEyM/inCJAiAgQooUGitecaX9LusEadF0RIEQECBEBQnSDI+nrigAhIkCICBAiAoSIACEiQIgIECIChIgAISJAiAYC+nhWkzzaLcajYYA+nj+9VKfd5lbi9NYKFnm7lsb4oO2toxiob1SDevXW9NKRz3IN01u7j45NgwG11J8oNXSW62kjfIbWoNNuolv8q9fQcdDj5DlZqYYBgna261ysVFq+cY0C5KNQo4HhZzbvhidASHgChIQnQEj4yK4OphcBQkSAEBEgRAQIEQFCRIAQESBEBAgRAUJEgBARIEQECBEBQkSAEBEgRAQIEQFCNAegnjloXepambGp7q24x9uuWvI+A6C+OWgd6lyZsaGeHRZH267a8j4DoL45aO3qXpmxrr49OsfaBjXzPpsPCjjNPSsz1hUMP8C2FcnK+2yA2uegdckbUM8+wWNtG327Qg0KnEPkW4i+nabH2jYR7LzPBCh0DtqaALl5nwdQ8By0FTWxWt6nBRQ6Bw1dmbGu8B4yFFA973PUoAFz0HwLEd7NBwJq5H2WcdBjcBzvQoQPFIMANfM+AyCfOWh1+Reieyvu8bbb8k4Xq4gIECIChIgAISJAiAgQIgKEiAAhIkCICBAiAoSIACEiQIgIECIChIgAISJAiAgQIgKEiAAhIkCICBAiAoSIACEiQIgIECIChOjqgMK2vHVUhmx/NtVOBovUoPNvBm30RoAQbQ9QwVpaBmX5wtqdKM/l81dofHnCfzBHWNC7rwCoFHHOeziSnfcy4nkP/wFCHfXIGfE/MqEKggdu47gkoIIV57zPWBnYh0JUj8sB9unMU3FUH8kZtBK2YIRdzS+HlMWDI8BSbA0r9oPPH15NVANIJcQpBu2LviggsS1wef+dfxBVQPx4+fwiflBHxMH87gW2O+UVBcpbyT88InBiwa2oGpBOaMgusgsCEi6FFcY0Bct1lAmvK+KIKBnncqwMMvNHRJROSkXVsU1C4YSWBCT7+xZAzGfc/3NvABUuIPZLAxA0S9hG14qqAamEWEO8JR+kO6UGIF3wgBrEmtC/WGg7qluDlPKwzWQX9UHyXDYAcR6l1cQEhMLxQXVAl8PfwJNbUZUHc3v8wP5/6V4MTmhrDWIj7sygK3gH5vRidUCsf0/dqO9PD6+sUR11QrwqBY2mVjAO4p1Piw+6e3HQNcdBDUBlUo8KlzVf1DhIVK3g7ZrpYhURAUJEgBD9H3YCtZjkjzo9AAAAAElFTkSuQmCC" /><!-- --></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(opar)</code></pre></div>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;cin&quot;</code></pre>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;cra&quot;</code></pre>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;csi&quot;</code></pre>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;cxy&quot;</code></pre>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;din&quot;</code></pre>
<pre><code>## Warning in par(opar): 无法设定图形参数&quot;page&quot;</code></pre>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-3-1" class="section level2">
<h2>Question 3</h2>
<p><font color="#000000"></p>
<p>In this informative age, processing big data is common, thus we generally need some function with the cycle structure to find the potential rules. And in the code of R, some of the cycle functions are different and unique. Those we have to study this. So how the procedure of the cycle does?</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-3-1" class="section level2">
<h2>Answer 3</h2>
<p><font color="#000000"></p>
<p>In the R software, the function with the form of multi function apply can avoid the ues of the direct cycle structure.as the example follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x&lt;-<span class="kw">rnorm</span>(<span class="dv">10</span>,<span class="op">-</span><span class="dv">5</span>,<span class="fl">0.1</span>)
y&lt;-<span class="kw">rnorm</span>(<span class="dv">10</span>,<span class="dv">5</span>,<span class="dv">2</span>)
X&lt;-<span class="kw">cbind</span>(x,y)
<span class="kw">apply</span>(X, <span class="dv">2</span>,mean)</code></pre></div>
<pre><code>##         x         y 
## -4.977066  7.121726</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">apply</span>(X, <span class="dv">2</span>, sd)</code></pre></div>
<pre><code>##          x          y 
## 0.06466531 1.99327143</code></pre>
<p>More information, the function of lappy() which has the silimar syntax with the function apply can work on the list objects and eventually return a list object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">forms&lt;-<span class="kw">list</span>(y<span class="op">~</span>x,y<span class="op">~</span><span class="kw">poly</span>(x,<span class="dv">2</span>))
<span class="kw">lapply</span>(forms,lm)</code></pre></div>
<pre><code>## [[1]]
## 
## Call:
## FUN(formula = X[[i]])
## 
## Coefficients:
## (Intercept)            x  
##      21.122        2.813  
## 
## 
## [[2]]
## 
## Call:
## FUN(formula = X[[i]])
## 
## Coefficients:
## (Intercept)  poly(x, 2)1  poly(x, 2)2  
##      7.1217       0.5457       2.7651</code></pre>
<p>In more flexible situations, to accept the vectors or the matrixes as the main parameters, function sapply can be a good alternate with the final return of the form of table. Here we do not talk these.</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-5.4" class="section level2">
<h2>Question 5.4</h2>
<p><font color="#000000"></p>
<p>Write a function to compute a $ Monte Carlo $ estimate of the $ Beta(3, 3) $ cdf, and use the function to estimate # F(x) forx =0 .1,0.2,…,0.9. # Compare the estimates with the values returned by the $ pbeta $ function in R.</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer" class="section level2">
<h2>Answer</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">seq</span>(.<span class="dv">1</span>,<span class="fl">0.9</span>, <span class="dt">length =</span> <span class="dv">9</span>) 
m &lt;-<span class="st"> </span><span class="dv">10000</span> 
u &lt;-<span class="st"> </span><span class="kw">runif</span>(m) 
cdf &lt;-<span class="st"> </span><span class="kw">numeric</span>(<span class="kw">length</span>(x)) 
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(x)) { 
  g &lt;-<span class="st"> </span>x[i] <span class="op">^</span><span class="dv">3</span><span class="op">*</span>u<span class="op">^</span><span class="dv">2</span><span class="op">*</span><span class="st"> </span>(<span class="dv">1</span><span class="op">-</span>x[i]<span class="op">^</span><span class="dv">2</span><span class="op">*</span>u<span class="op">^</span><span class="dv">2</span>) 
  cdf[i] &lt;-<span class="st"> </span><span class="dv">30</span><span class="op">*</span><span class="kw">mean</span>(g) 
}</code></pre></div>
<p><font color="#000000"></p>
<p>Now the estimates ?? ?? for ten values of x are stored in the vector cdf. Compare the estimates with the value ??(x) computed (numerically) by the pnorm function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Phi &lt;-<span class="st"> </span><span class="kw">pbeta</span>(x,<span class="dv">3</span>,<span class="dv">3</span>) 
<span class="kw">print</span>(<span class="kw">round</span>(<span class="kw">rbind</span>(x, cdf, Phi), <span class="dv">3</span>)) </code></pre></div>
<pre><code>##      [,1]  [,2]  [,3]  [,4]  [,5]  [,6]  [,7]  [,8]  [,9]
## x   0.100 0.200 0.300 0.400 0.500 0.600 0.700 0.800 0.900
## cdf 0.010 0.078 0.254 0.576 1.058 1.686 2.413 3.146 3.744
## Phi 0.009 0.058 0.163 0.317 0.500 0.683 0.837 0.942 0.991</code></pre>
<p>Results for several values x&gt;0 are shown compared with the value of the beta distribution cdf function pbeta. The Monte Carlo estimates appear to not be very close to the pnorm values cause the estimates will be worse in the extreme upper tail of the distribution.)</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-5.9" class="section level2">
<h2>Question 5.9</h2>
<p><font color="#000000"></p>
<p>The $ Rayleigh $ density [156, (18.76)] is $ f(x)= x/sigma<sup>2<em>exp(???x^2/(2</em>sigma</sup>2)),x ??0,??&gt;0.$ Implement a function to generate samples from a $ Rayleigh(??)$ distribution, using antithetic variables. What is the percent reduction in variance of $ (X+X’)/ 2 $ compared with $ (X1+X2)/2 $ for independent $ X1 $, $X2 $?</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-4" class="section level2">
<h2>Answer</h2>
<p><font color="#000000"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MC.Phi &lt;-<span class="st"> </span><span class="cf">function</span>(x, <span class="dt">R =</span> <span class="dv">10000</span>, <span class="dt">antithetic =</span> <span class="ot">TRUE</span>) {
  u &lt;-<span class="st"> </span><span class="kw">runif</span>(R<span class="op">/</span><span class="dv">2</span>) 
  sigma &lt;-<span class="st"> </span><span class="dv">4</span>
  <span class="cf">if</span> (<span class="op">!</span>antithetic) v &lt;-<span class="st"> </span><span class="kw">runif</span>(R<span class="op">/</span><span class="dv">2</span>) <span class="cf">else</span> 
    v &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>u 
  u &lt;-<span class="st"> </span><span class="kw">c</span>(u, v) 
  cdf &lt;-<span class="st"> </span><span class="kw">numeric</span>(<span class="kw">length</span>(x)) 
  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(x)) { 
    g &lt;-<span class="st"> </span>x[i]<span class="op">^</span><span class="dv">2</span><span class="op">*</span>u<span class="op">/</span>sigma<span class="op">^</span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span><span class="kw">exp</span>(<span class="op">-</span>(u<span class="op">*</span><span class="st"> </span>x[i])<span class="op">^</span><span class="dv">2</span> <span class="op">/</span><span class="st"> </span>(<span class="dv">2</span><span class="op">*</span>sigma<span class="op">^</span><span class="dv">2</span>) )
    cdf[i] &lt;-<span class="st"> </span><span class="kw">mean</span>(g)  
  } 
  cdf 
}</code></pre></div>
<p>A comparison of estimates obtained from a single Monte Carlo experiment is below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">1</span>,<span class="dv">10</span>, <span class="dt">length=</span><span class="dv">10</span>) 
sigma &lt;-<span class="st"> </span><span class="dv">4</span>
Phi &lt;-<span class="st"> </span>x<span class="op">/</span>sigma<span class="op">^</span><span class="dv">2</span><span class="op">*</span><span class="kw">exp</span>(<span class="op">-</span>x<span class="op">^</span><span class="dv">2</span><span class="op">/</span>(<span class="dv">2</span><span class="op">*</span>sigma<span class="op">^</span><span class="dv">2</span>))
<span class="kw">set.seed</span>(<span class="dv">123</span>)
MC1 &lt;-<span class="st"> </span><span class="kw">MC.Phi</span>(x, <span class="dt">anti =</span> <span class="ot">FALSE</span>) 
<span class="kw">set.seed</span>(<span class="dv">123</span>)
MC2 &lt;-<span class="st"> </span><span class="kw">MC.Phi</span>(x) 
<span class="kw">print</span>(<span class="kw">round</span>(<span class="kw">rbind</span>(x, MC1, MC2, Phi), <span class="dv">5</span>))</code></pre></div>
<pre><code>##        [,1]    [,2]    [,3]    [,4]    [,5]    [,6]    [,7]    [,8]
## x   1.00000 2.00000 3.00000 4.00000 5.00000 6.00000 7.00000 8.00000
## MC1 0.03062 0.11701 0.24434 0.39260 0.54172 0.67584 0.78559 0.86808
## MC2 0.03077 0.11751 0.24521 0.39361 0.54246 0.67585 0.78448 0.86566
## Phi 0.06058 0.11031 0.14153 0.15163 0.14307 0.12174 0.09462 0.06767
##        [,9]    [,10]
## x   9.00000 10.00000
## MC1 0.92539  0.96232
## MC2 0.92166  0.95743
## Phi 0.04475  0.02746</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m &lt;-<span class="st"> </span><span class="dv">1000</span> 
MC1 &lt;-<span class="st"> </span>MC2 &lt;-<span class="st"> </span><span class="kw">numeric</span>(m) 
x &lt;-<span class="st"> </span><span class="dv">2</span>
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m) { 
  MC1[i] &lt;-<span class="st"> </span><span class="kw">MC.Phi</span>(x, <span class="dt">R =</span> <span class="dv">1000</span>, <span class="dt">anti =</span> <span class="ot">FALSE</span>) 
  MC2[i] &lt;-<span class="st"> </span><span class="kw">MC.Phi</span>(x, <span class="dt">R =</span> <span class="dv">1000</span>)
  }
<span class="kw">print</span>(<span class="kw">sd</span>(MC1)) </code></pre></div>
<pre><code>## [1] 0.002024304</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">sd</span>(MC2)) </code></pre></div>
<pre><code>## [1] 0.0001381602</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>((<span class="kw">var</span>(MC1) <span class="op">-</span><span class="st"> </span><span class="kw">var</span>(MC2))<span class="op">/</span><span class="kw">var</span>(MC1)) </code></pre></div>
<pre><code>## [1] 0.9953418</code></pre>
<p>The antithetic variable approach achieved approximately 99.5% reduction in variance at x =2.0</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-5.13" class="section level2">
<h2>Question 5.13</h2>
<p><font color="#000000"></p>
<p>Find two importance functions f1 and f2 that are supported on (1,??) and are ??close?? to $ g(x)= x<sup>2/sqrt(2<em>pi)</em>exp(???x</sup>2/2), x &gt; 1.$ Which of your two importance functions should produce the smaller variance in estimating $ integrate( x<sup>2/sqrt(2??)* exp(???x</sup>2/2),1,inf) $ by importance sampling? Explain.</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-5" class="section level2">
<h2>Answer</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m &lt;-<span class="st"> </span><span class="dv">10000</span> 
theta.hat &lt;-<span class="st"> </span>se &lt;-<span class="st"> </span><span class="kw">numeric</span>(<span class="dv">2</span>) 
g &lt;-<span class="st"> </span><span class="cf">function</span>(x) { 
 x<span class="op">^</span><span class="dv">2</span><span class="op">/</span><span class="kw">sqrt</span>(<span class="dv">2</span><span class="op">*</span>pi)<span class="op">*</span><span class="kw">exp</span>(<span class="op">-</span>x<span class="op">^</span><span class="dv">2</span><span class="op">/</span><span class="dv">2</span>) <span class="op">*</span><span class="st"> </span>(x <span class="op">&gt;</span><span class="dv">1</span>) 
 }

x &lt;-<span class="st"> </span><span class="kw">rnorm</span>(m) <span class="co">#using f1</span>
fg &lt;-<span class="st"> </span><span class="kw">sqrt</span>(<span class="dv">2</span><span class="op">*</span>pi)<span class="op">*</span><span class="kw">g</span>(x)<span class="op">/</span><span class="kw">exp</span>(<span class="op">-</span>x<span class="op">^</span><span class="dv">2</span><span class="op">/</span><span class="dv">2</span>)
theta.hat[<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">mean</span>(fg) 
se[<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">sd</span>(fg)

x &lt;-<span class="st"> </span><span class="kw">rchisq</span>(m,<span class="dv">4</span>) <span class="co">#using f2</span>
fg &lt;-<span class="st"> </span><span class="dv">4</span><span class="op">*</span><span class="kw">g</span>(x)<span class="op">/</span>(x<span class="op">^</span><span class="dv">2</span><span class="op">*</span><span class="kw">exp</span>(<span class="op">-</span>x<span class="op">/</span><span class="dv">2</span>))
theta.hat[<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">mean</span>(fg) 
se[<span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">sd</span>(fg)

<span class="kw">rbind</span>(theta.hat, se) </code></pre></div>
<pre><code>##                [,1]      [,2]
## theta.hat 0.2362719 0.0000000
## se        1.1617393 0.4154807</code></pre>
<p><font color="#000000"></p>
<p>we choose f1 as standard norm distribution,and f2 as chi-squared distribution.The estimates (labeled theta.hat) of the integate g(x) within the interval (1,inf) and the corresponding standard errors se for the simulation using each of the importance functions are perfomed above. From the results below,we know that the f1 produce the smaller variance. the relative codes are followed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, .<span class="dv">01</span>)
w &lt;-<span class="st"> </span><span class="dv">2</span> 
f1 &lt;-<span class="st"> </span><span class="kw">exp</span>(<span class="op">-</span>x<span class="op">^</span><span class="dv">2</span><span class="op">/</span><span class="dv">2</span>)<span class="op">/</span><span class="kw">sqrt</span>(<span class="dv">2</span><span class="op">*</span>pi)
f2 &lt;-<span class="st"> </span>x<span class="op">^</span><span class="dv">2</span><span class="op">*</span><span class="kw">exp</span>(<span class="op">-</span>x<span class="op">/</span><span class="dv">2</span>)<span class="op">/</span><span class="dv">4</span>
g &lt;-<span class="st"> </span>x<span class="op">^</span><span class="dv">2</span><span class="op">/</span><span class="kw">sqrt</span>(<span class="dv">2</span><span class="op">*</span>pi)<span class="op">*</span><span class="kw">exp</span>(<span class="op">-</span>x<span class="op">^</span><span class="dv">2</span><span class="op">/</span><span class="dv">2</span>)

<span class="co">#figure (a) </span>
<span class="kw">plot</span>(x, g, <span class="dt">type =</span> <span class="st">&quot;l&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">2</span>), <span class="dt">lwd =</span> w) 
<span class="kw">lines</span>(x, f1, <span class="dt">lty =</span> <span class="dv">2</span>, <span class="dt">lwd =</span> w) 
<span class="kw">lines</span>(x, f2, <span class="dt">lty =</span> <span class="dv">3</span>, <span class="dt">lwd =</span> w) 
<span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&quot;g&quot;</span>, <span class="dv">0</span><span class="op">:</span><span class="dv">2</span>), <span class="dt">lty =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dt">lwd =</span> w, <span class="dt">inset =</span> <span class="fl">0.02</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAV1BMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6kNtmAABmADpmZmZmtv+QOgCQZgCQkGaQ2/+2ZgC2tma2/7a2///bkDrb/9vb////tmb/25D//7b//9v///+pWGjuAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAGOElEQVR4nO2d7ZqiNhhAmY9tO+3arm3pOqP3f50FHR2DwiEYIMo5P3acR/LCnE1CXkig2EknxdwHkDsKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCgIEBBQJyg7arY8/zvSIeTH1GCyuK3w4fN8cPjEyNouzppKV9+jnAwORIj6OPt+/HjZjGNzBoERPZBn1XIPqiFj7fDWay1/hR3wziCpg43Ho8oKGn1GVtQ2XIWG1VQykgT16D4pj1kJykjPWQTSxlJQRBpJEFl1Yj2QyH7oKuUTz+qodDrTkHXOaQa21U1TFTQNY7J6vrlp4KucUpW168KuspRS5WSKegqx3R+u1LQ4F2PhoImi60gjqQgiKQgiKQgiKSgkHVRPP1VZZKnSAsQFHNJfl0N2zbFsgTRRedz3r/V49v1sgTFcLgPvFFQG6WCurEGAYc+qFRQK0s8i8VRjYOe/z5dolHQVTYzCsr8zuq+D/p4O03ZsQY12RTHG3rHSAqCSAqCSAqCSAqCSAoKqXrp0zBRQZfUadhXKqagJod75uvXs0gKOuczWV1YqtEcu3eM5d9/qVvXnKnGWOE6YjeTm65k59D9eD2oFQUBi2xiMSyzk45gutM8rlnNU9BkA0Ves5qpoHqC9wSpRo8Vh7kKuog0iqAea1aXLcgaRPCa1YUL6rFmNSpcHPcgaOpwI8VWEEcaW5BrNSL3mPed1Zr3X7/+S21il3y8PSuog02QPy5D0LE5t/08pxq3bZYm6Njftf1sMoWg4zCx44JHrk1sGkHt6+iGhYvkDgRVhl67N1i6oKq3+975/eIFTRwuYWwFAQoCFBQXSUEQSUEQSUEQSUEQSUEQSUEQSUEB+0kXX7fyFNRgu3r6sSsLZ7m2scwJVDEPFtizsDmKMQ8WOLBeWA2K5WzChYKusPnqo2cQlP+d1WDCjjXogjKY0KSgJmV4KV1BDd6/hRPi5hdEK3EG9VfDD7U87C+fcRCtxLno0nt18YtLVllguOXiBHVE6fZ1e/x7F3QlroKmjK0gjqQgiKQgiKQgiKQgiDSSoHtdknkZaRxBMy/JTMdIgmZeUHexr6t/ecfmnz/P/u2zl4gjmnlJ5nEXfb2EM8uL0zzquAO9lxrUR0ubkWsHNVYfNPWSzF7Vpedc+7BM7yPou+GeaZZktvSrzSrSFBO3j+Qbjhyu/Xyza60it+T32QtqExJ6uaGKJDvQYfFhSSb8+S1KxhTScqAJN+yOcvFf30fIJCbajjj5hrOEGw8FAY+YrCblEZPVpMyWatwNYwjqkaz2Cwrf3/b1rcVv2LxHDeoX9FEF9UhW+wV9WEGcrPYL+riCEgVV0G3fK0hBt32voMULeiQUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgIKWg4D154UvzootXrOHCblfxegld9wMN+5JQUPD6k8a7UGKL17/Tle+O4vWLwemRjz1JJyh4gU7zbTqRxXf7+wMxghp7f43cezvpBAVPEGk+TiSyeP3p5c8YQUHxTAWdv8Sr+UavyOL7X6P6oLB4lk0seA1c851wkcX3LSZKUGOH8aeINjIVVN/avkFQ/aCX5mL5geTZxPa/DG9i8T1gO3l20p9PAojoRYLi8fW3nWxP85E1KCh+sBVTf9vJd6AYOZIOimfZB53ek3cYhZTR55Gg+C461QiKr4uie4ZOb0xWAQUBCgIUBCgIUBCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBCgIUBCQu6B6ztjHW5qZLIPIXVA9c6z7YUUjk7ug3eb5n9/TTOgdRvaCdusiyYz5oeQvKNGM+aFkL2i7+iPJbNWhZC+ofPlvNeNJLHtB9aMJ00x4Hkjuguq5wKeJwXOQu6DZURCgIEBBgIIABQEKAhQEKAhQEKAgQEGAggAFAQoCFAQoCFAQoCBAQYCCAAUBCgIUBPwPHEeJKWHAe6MAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#figure (b) </span>
<span class="kw">plot</span>(x, g, <span class="dt">type =</span> <span class="st">&quot;l&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;&quot;</span>, <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">3.2</span>), <span class="dt">lwd =</span> w, <span class="dt">lty =</span> <span class="dv">2</span>) 
<span class="kw">lines</span>(x, g<span class="op">/</span>f1, <span class="dt">lty =</span> <span class="dv">3</span>, <span class="dt">lwd =</span> w) 
<span class="kw">lines</span>(x, g<span class="op">/</span>f2, <span class="dt">lty =</span> <span class="dv">4</span>, <span class="dt">lwd =</span> w)
<span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">c</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">2</span>), <span class="dt">lty =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">4</span>, <span class="dt">lwd =</span> w, <span class="dt">inset =</span> <span class="fl">0.02</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAV1BMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6OpA6kNtmAABmADpmZmZmtv+QOgCQZgCQ2/+2ZgC2/7a2///bkDrb2//b/9vb////tmb/25D//7b//9v////aoDv7AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAGdklEQVR4nO2d62KjNhgFaZLdNmnrrVvaJTHv/5wN4JsIcPh0MQJm/sTE1oFMhJCwgKKGSYqlNyB3ECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJECRAkABBAgQJIgsqVkMaQadDl/70T5S4BUkjqCxeuxfV5UVQnJGo1SeJoNPhqqV8/hkcZyVedipBH29/Xl5WIzvZvgVRgxRlca5Cy7RBMZMSHcU+3rpGbqT+IOjRcYmyEaSTUgsqOYrZ1mjuwfusJGYSu5hIQpBISimoKopf/ooXN5s1CDoWxev7bz/vBx0hcTZWIOj42UE8trWHocYQbb15/94IWttg1W0W0glqRmCnf+vV1aDq0051M5RusHqpN52q0Dgj/tndaYjjy11Soka67P4LVTHSRmcq6P1bs7233v8++kH9vvtEX77fbu5CUH9wMzXY6ZqfWyO0C0EWECTY5S5mYZ+NtIHHHeZHUzI/H/SojuLD46Jllw8Zajw+LlE2gnQSgkQSgkQSgkQSgkQSgkQSgvq8/3o7SYygr3y8PSFogsqZeIqgPlXxWu1O0GVgPPazx+4EXc4cjP3ssztBVpYVlPv5oHppQaniImYjSIAgAYIE2xN069ecm//8B6vdnI7PXnvk68W+qDgvOL9cjaB2ZlD4FLxJFeGbKlaeUNBZTeAEKq9e0zoEdd/iek7Bu68rhvXOyjYmZVODBtsVP9YgqPlTX2rDFLyYQ5D8BdWto+Zr7rHr6W5x1zpjy59iFYLmxqUYtgYktpez3/6pGQhKgX/26fBZ7cvikdNf1nW92PITqJY4HzTY/55q9HY2R3G4/z31LzoyBW+Su0Mvggaobm00ggZwum5pe9IdqzqK3d235pKUpgadDqMngnzijARkl+7VN+l2sdPhZfoDeQp6/+YOjRK2QeMXQnnFmfDPLrtmYV/9oMAkBIkkBIkkBIkkBIkkBIkkBImkHZwPCkuiBokkBIkkBImkTQqKxzYFeTN06EDQjcFjK4IECKqnvwNCUI2gCXSPHkHqE7OjwrYkcZzHBszcgr0Kmj1c3p+gVFNmtyLIeqZli+eDRtftVSr6BxeJm7NGBI2uK2RlOxAUtj9vWdBjZ6WvT1CcI0EiQcs+X2yRafumdS71fLElr2uwrHOpp0OF7lRDXbQkgh7+fDFPMb1O62Afdt01yHp52bCQ2/JQkdnbMveDDQ96vpgUMy1k3jqif7Al7fPFJv7GqX3Gq4laST/IPTx9UVC7i5P7jJHMBY1cEH599+ty3M1ML0hcL9Y//dFbHnu3vi77bZWBZc8HFXOXo6zci8x3seVBkGCTg9WYbGuwmoDFhhqrIYWgGYPVeaFqJBH0dmjxgI/PqEHzQrcqaMZgdV7oZgXpweq80O0KihSKoLD3EYSgsPcRtHtBWwJBAgQJECRAkABBAgQJECRAkABBAgQJECRAkCCmoOr+ecHOgr34J0dxYneqeHPXsum7rc0loiDnidO9x09bizfL6sz3RPGyvSt4FEPxBDnPLO8/wNxYvG6/H7AI6q39xbj2ceIJcm7a3L+Ds7F48+r5h0WQUzxTQd+b+n3+ytVZsBdvF01tkFs8y12sawDOzYCzYC/e7jEmQb0V2g8RY2QqqPlqO0BQc2/t/v1JPclzF2sX/Hcxews4Tp6N9Pnmq4ZWxClur7/jZHuYN9Ygp3hny1J/x8m3o2jsSTvFs2yD2h2j2cKuF1KajyNO8do81HCKH4tieobObBisChAkQJAAQQIECRAkQJAAQQIECRAkQJAAQQIECRAkQJAAQQIECRAkQJAAQQIECRAkQJAAQQIECRAkQJAAQQIECRAkyF1QM2fs4y3OTBYvchfUzBybvllRYnIXVFdPf/8eZ0KvH9kLqo9FlBnzvuQvKNKMeV+yF3Q6/BFltqov2Qsqn/87LHgQy15Qc2vCOBOePcldUDMX+DoxeAlyF7Q4CBIgSIAgAYIECBIgSIAgAYIECBIgSIAgAYIECBIgSIAgAYIECBIgSIAgAYIECBIgSIAgwf92oKBBMz37fQAAAABJRU5ErkJggg==" /><!-- --></p>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-5.14" class="section level2">
<h2>Question 5.14</h2>
<p><font color="#000000"></p>
<p>Obtain a Monte Carlo estimate of $ integrate( x<sup>2/sqrt(2??)* exp(???x</sup>2/2),1,inf) $ by importance sampling.</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-6" class="section level2">
<h2>Answer</h2>
<p><font color="#000000"></p>
<p>we use the standard norm distribution to estimate the integrate g(x) and we get the result :0.3868715.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m &lt;-<span class="st"> </span><span class="dv">10000</span> 
theta.hat &lt;-<span class="st"> </span>se &lt;-<span class="st"> </span><span class="kw">numeric</span>(<span class="dv">1</span>) 
g &lt;-<span class="st"> </span><span class="cf">function</span>(x) { 
  x<span class="op">^</span><span class="dv">2</span><span class="op">/</span><span class="kw">sqrt</span>(<span class="dv">2</span><span class="op">*</span>pi)<span class="op">*</span><span class="kw">exp</span>(<span class="op">-</span>x<span class="op">^</span><span class="dv">2</span><span class="op">/</span><span class="dv">2</span>) <span class="op">*</span><span class="st"> </span>(x <span class="op">&gt;</span><span class="dv">1</span>) 
 }
x &lt;-<span class="st"> </span><span class="kw">rnorm</span>(m) 
fg &lt;-<span class="st"> </span><span class="kw">sqrt</span>(<span class="dv">2</span><span class="op">*</span>pi)<span class="op">*</span><span class="kw">g</span>(x)<span class="op">/</span><span class="kw">exp</span>(<span class="op">-</span>x<span class="op">^</span><span class="dv">2</span><span class="op">/</span><span class="dv">2</span>)
theta.hat &lt;-<span class="st"> </span><span class="kw">mean</span>(fg) 
theta.hat</code></pre></div>
<pre><code>## [1] 0.4195725</code></pre>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-6.9" class="section level2">
<h2>Question 6.9</h2>
<p><font color="#000000"></p>
<p>Let $ X $ be a non-negative random variable with $ ?? = E[X] &lt; ?? $ . For a random sample $ x_{1},…,x_{n} $ from the distribution of $ X $, the $ Gini $ ratio is de???ned by <span class="math display">\[ G = \frac{1}{2n^2u}\sum_{j=1}^{n}\sum_{i=1}^{n}|x_{i}- x_{j}|.\]</span> The $ Gini $ ratio is applied in economics to measure inequality in income distribution (see e.g. [163]). Note that $ G $ can be written in terms of the order statistics $ x_{i} $ as <span class="math display">\[ G = \frac{1}{n^2u}\sum_{i=1}^{n}(2i-n-1)x_{(i)}.\]</span> If the mean is unknown,let $  $ be the statistic $ G $ with $ ?? $ replaced by $  $. Estimate by simulation the mean, median and deciles of $  $ if $ X $ is standard lognormal. Repeat the procedure for the uniform distribution and $ Bernoulli(0.1) $. Also construct density histograms of the replicates in each case.</p>
<p><font color="#FF7F50"></p>
<div id="answer-6.9" class="section level3">
<h3>Answer 6.9</h3>
<p><font color="#000000"></p>
<p>We use the ordered sample to esrimte the relevent target values of $ G $ function. The basic idea is : (1) generate the random samples with the number n: $ x_1,x_2,????,x_n $, and iid from the distribution of $ X $. (2) Sort the samples we get in increasing order, to obtain the ordered samples: $ x(1)<x(2)<????<x(n)$ (3) Compute the median function to get the results,mean ,median and deciles of the G. #### X is standard lognormal First, we simulate the mean,median and deciles of $ \hat{G} $ if $ X $ is standard lognormal. ```r n <- 20 ## numbers of samples m <- 1000 ## experiment times G <- numeric(m) for (i in 1:m) { x <- sort(rlnorm(n)) ## X is standard lognormal mu <- mean(x) ## let the mu be the mean values of the samples for (j in 1:n) { t <- (2*j-n-1)*x[j] } G[i] <- sum(t) (mu*n^2) } Ghat <- mean(G) ## compute the estimate values(mean) of G MGhat <- median(G) ## compute the estimate values(median) of G DGhat <- quantile(G,probs="0.1)" ## compute the estimate values(deciles) of G c(Ghat,MGhat,DGhat) ``` ``` ## 10% ## 0.2041721 0.1864395 0.1286057 ``` ```r hist(G,prob="TRUE)" ## density histogram of sample ``` ![](intro_files figure-html unnamed-chunk-27-1.png)<!-- --></p>
<div id="x-has-the-uniform-distribution" class="section level4">
<h4>X has the uniform distribution</h4>
<p>Then, we simulate the mean,median and deciles of $  $ if $ X $ is uniform.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="dv">20</span>  ## numbers of samples
m &lt;-<span class="st"> </span><span class="dv">1000</span>  ## experiment times
G &lt;-<span class="st"> </span><span class="kw">numeric</span>(m) 
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m) {
  x &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">runif</span>(n)) ## X has the uniform distribution
  mu &lt;-<span class="st"> </span><span class="kw">mean</span>(x)  ## let the mu be the mean values of the samples
  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n) {
    t &lt;-<span class="st"> </span>(<span class="dv">2</span><span class="op">*</span>j<span class="op">-</span>n<span class="op">-</span><span class="dv">1</span>)<span class="op">*</span>x[j]
  }
  G[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(t)<span class="op">/</span>(mu<span class="op">*</span>n<span class="op">^</span><span class="dv">2</span>)
}
Ghat &lt;-<span class="st"> </span><span class="kw">mean</span>(G) ## compute the estimate values(mean) of G
MGhat &lt;-<span class="st"> </span><span class="kw">median</span>(G) ## compute the estimate values(median) of G
DGhat &lt;-<span class="st"> </span><span class="kw">quantile</span>(G,<span class="dt">probs=</span><span class="fl">0.1</span>) ## compute the estimate values(deciles) of G
<span class="kw">c</span>(Ghat,MGhat,DGhat)</code></pre></div>
<pre><code>##                              10% 
## 0.09203228 0.09101508 0.07730179</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(G,<span class="dt">prob=</span><span class="ot">TRUE</span>) ## density histogram of sample</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAXVBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6OgA6OpA6kNtmAABmADpmtv+QOgCQOjqQZgCQtpCQ2/+2ZgC2Zma2/7a2///bkDrb2//b////tmb/25D//7b//9v///9YZbyyAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAIcElEQVR4nO2d7YKjJhiFzU5m2ybdaTu2a8ck3v9lrnwpidEDiBHxPD92Ewdf5QkCKmLRkEmKtXcgdSgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCDAwoJuH8Xbl/r/2H2x+Pdn7C2WRdFuSnN5L+yvIawq6Hr+FltQLYycuo1rfoQHfKWgAWURXVBVHD77bRtmbGaFElSJXRYfSrPv8kjQP3u78PApzbV5/edd5Ff+WeT7ei6OtSwQVdGLaKwAMqT+S62LjvjjKTgHrxdUdT+qEaSXiITqVz/83i9++6rVn9tst4Is+mLRB7AEtaG0qct70oeYQQsSpUD+uidziLW/sPJ2ksuPshwoQdrZUSY6SUGnpitDpljYAfpDTG1oNi8X1Oam++mVIF0TlapQic9t3pSg7pevZIB2eWtC/SuN2WEac1xqQVaCObxckF4i911myRyFImvm811eTYyjLhTq3z7/doChoLo7fMN4fR1kKpIf2oM5FNqc/DCfbUHSj/rLc0F2gOEhtj1BTd+OOZUgZWJC0EgJ0pX09gTd/hI5sGqc8TpI5lUWjK6SfiJopA7qmvlnPXh3Xi5INTmmxRe5GW/FTAlSNddYCRppxez6b0Z7tlo/yBT/vh90bB77QX0dNFVJN3YAS9D2TjX0F9ntVU19WZgy1GVClp5BK3b4rNpFY4KsANVdD1s1CLNa+zQvd5QzKo3IJCaolL93WyBidIKjkJggfd51d5ysS2KCdK2RzAGWnqDkoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBBgLUHFkJX2BOC3W2bc6PzHvIbbTdSY115UZsBoPfs5kSeCcJI18NmL20c/MnnuILksBV3P3YDseu5BlqUgliBEZYbDsw4awTzrNX+YbqaCltxu1oL8+y45C6onnhTYuyDxrMnljy+7wQ8Ml6egUjxnKkvPWDO/b0Gy3Fy+ywdJRzqKexckej+3/xuWoBEqU26Uqjnh8hTUVKr5qkefkt27oHjhKMg7IQWBhBQEElIQSEhBICEFgYRZC+LljujhticozqRxGQtSs/DEvxbvnDB9QU0ER7kLUopeMfPpNgWpadFuHy+YEm2DgsQ9L2Vmxn3ljAVdzw6TGsHhLzkL+lNPMjxRfPDwlx0IqsYFOQxeyFZQ2Z8sjDfzDsNfshXUl6AJ9lyC3MDDX3YuCA9/yVXQ9Xzq5oufNbguV0Gzo/B6kKLq5pTdXysmbydX06PrKvnqBjkr8Q4FlWIy7GNTjl82U828PJfdoSDRCxT33CdONUxHsVW5U0GlmOba4VSjPO5QUJvr61m8OGXiyrTRoiblnw7nvN3RJakJanMt3tsweeXedKVvHzsU9NpwFOSdMHlBUR6Wy1lQGeNtcRkLGh2XGRbOP2H6gma8vmQYzj9h6oJAA+8bzj9h6oImxvYGhfNOmLqgWRfMeD0oejgK8k6YviD59slyXmOfs6D68FmJs3mOD7rDauZP8nbgxPUgn3D+CYeCkpjM466jKATNm1Nh0RK0riBTgspZz8SP5cGhNKQuSNdB1WR3MXh8kENmkxekuoqTo6jCxwdlIQgyY3THPgTNGB+UgSCHlwvuugSV5j2uU/3E8PFBmxdU929pn2rFgscHbV2QdbVs4t68cziH5RsTZJ2BBZ1qwLOB7QuKNH0bBTmFc1iepaDuquz4yca+BY2PWXgI57B8c4Jg2ZDAe0O5CnIG3RvavaDQcBQEwlEQCEdBIBwFgXAUBMJREAhHQSBcHEFr3ErclCDnzUUkzibWuh60GUEwHAWBcBQEwlEQCEdBIBwFgXAUBMJREAhHQSAcBYFwSwla/uRs44Jctx+OX0THMYoOp915CnIdoxiUkwwEOY8wW1FQ9FrJZ/3xG9QPu/NkNxNiOUExX+G3GTzroGiv8NsMfgUu3iv8NsML+qLbZhFB61bCD8zNSxQj/kFTSrLo+qFBU0qy6PqhQVNKsuj6oUFTSrLo+qFBU0qy6PqhQVNKsuj6oUFTSrLo+qFBU0qy6PqhQVNKsuj62UNBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAkQTV9txV3ZfLe+H1dtv6fgasy28/nywNjNI0YdOzxREkJo/pJpDpvtTtDk2+ZWEqiriNK0dIPCwNjCL+EHRDOIogNapBzxvTfVEf3CdLuYsiC4JY9WFpYJRGTzTqGqUniiA16ZBW0X25fPc5MB6iiPERcozN/dLQKGLR29/rCZIq9J50X+pv/5193ox8F6UxHwdLg6LI5evVQeqQ1wd+96UShdtjovO7KHKByNpgaVAUeeQlJ+jg99svKih4ktEFDzFVD0xPijYapWmiHmJy6YqCnlbSaifdq+pBdRyhkjZRKj0Wxn/a/gWbeTXm0/23HzTodYRm3t6B5DqK4ri3Bn76RWlM1mZ2FNMQJIqw2CfVZlWmry/eX+8z2vMuSpe1yvNU43mUZl1B+UJBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCJC9ITlm04nxFqQuSk17h13YtR+KC9Pg0j5GOsUlckBmxUvkPDYtE2oJ8hl8tRNqCrJkb12ILgspi+sWCi5K6IDN5LAU9x1TSFDSCGYZOQWOoGZpLr+fyopK6IHWqsV5HOn1Ba0NBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQF+AVvb0OlciH+SAAAAAElFTkSuQmCC" /><!-- --></p>
</div>
<div id="x-has-the-bernoulli0.1-distribution" class="section level4">
<h4>X has the Bernoulli(0.1) distribution</h4>
<p>Last, we simulate the mean,median and deciles of $  $ if $ X $ is Bernoulli(0.1).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="dv">20</span>  ## numbers of samples
m &lt;-<span class="st"> </span><span class="dv">1000</span>  ## experiment times
G &lt;-<span class="st"> </span><span class="kw">numeric</span>(m) 
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m) {
  x &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">rbinom</span>(n,<span class="dt">size=</span><span class="dv">1</span>,<span class="dt">prob=</span><span class="fl">0.1</span>)) ## X has the Bernoulli(0.1) distribution
  mu &lt;-<span class="st"> </span><span class="kw">mean</span>(x)  ## let the mu be the mean values of the samples
  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n) {
    t &lt;-<span class="st"> </span>(<span class="dv">2</span><span class="op">*</span>j<span class="op">-</span>n<span class="op">-</span><span class="dv">1</span>)<span class="op">*</span>x[j]
  }
  G[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(t)<span class="op">/</span>(mu<span class="op">*</span>n<span class="op">^</span><span class="dv">2</span>)
}
Ghat &lt;-<span class="st"> </span><span class="kw">mean</span>(G) ## compute the estimate values(mean) of G
MGhat &lt;-<span class="st"> </span><span class="kw">median</span>(G) ## compute the estimate values(median) of G
DGhat &lt;-<span class="st"> </span><span class="kw">quantile</span>(G,<span class="dt">probs =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="fl">0.1</span>),<span class="dt">na.rm =</span> <span class="ot">TRUE</span>) ## compute the estimate values(deciles) of G
<span class="kw">c</span>(Ghat,MGhat,DGhat)</code></pre></div>
<pre><code>##                            0%       10%       20%       30%       40% 
##       NaN        NA 0.1357143 0.2375000 0.3166667 0.3166667 0.4750000 
##       50%       60%       70%       80%       90%      100% 
## 0.4750000 0.4750000 0.4750000 0.9500000 0.9500000 0.9500000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(G,<span class="dt">prob=</span><span class="ot">TRUE</span>) ## density histogram of sample</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAWlBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6OgA6OpA6kNtmAABmADpmtv+QOgCQOjqQtpCQ2/+2ZgC2Zma2/7a2///bkDrb2//b////tmb/25D//7b//9v///+Jgy4rAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAIIklEQVR4nO2d7YKjJhhGzU5m2yad2Y5t7ZjE+7/NypeSD3wUUBGf82M3Y8yrHBFEEYqGDFKsvQOpQ0EACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQFmFnT7LN6+1f/H7g+Lv/+NvcWyKNpNaS7vhf2nD6sKup5/xBZUCyOnbuOaD/+ASwp6oiyiC6qKw1e/bUPAZlbIQZXYZfGhNPsuzwR92NuFhy9prk3rX+8ivfJrke7ruTjWMkNURS+isQLIkPqbWmcd8eXJOwXLC6q6g2oE6SViRXXUD7/3i9++a/V1m+xWkEWfLfoAlqA2lDZ1eU/6FDNoQSIXyKN7MqdYe4SVt5NcfpT5QAnSzo5ypZMUdGq6PGSyhR2gP8XUhoJZXFCbmu7QK0G6JCpVphKf27QpQd2Rr2SAdnlrQv0rjdlhGnNeakHWCiEsLkgvkfsuk2TOQpE08/kurSbGUWcK9W+ffjvAs6C6O339WL4MMgXJh/ZgToU2JR/msy1I+lHfvBZkB3g+xbYnqOnrsVE5SJkYEOTIQbqQ3p6g2y+RAqvEcZdBMq0yY3SF9AtBjjKoq+ZfXcGPZ3FBqsoxNb5IjbsWMzlIlVyuHOSoxezyL6A+W+06yGT//jro2DxeB/Vl0FAh3dgBLEHba2roP+Rlr6rqy8LkoS4RMvc81WKHr6pd5BJkBajurrBVhRBU26d5u6MMKDQik5igUh7vNkPEuAiOQmKCdLvr7jxZl8QE6VIjmRMsPUHJQUEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCAHEE9Z0Bo4RLicgpoqD5whVPxNutABISFDFWRCgIMG0vTM9j5xt8OxdUmS7Htavv8b4F3T77vu2Obpb7FnQ9d136a8dJtm9BzEGIyrxQwTLIgXlb0NnRe++C5gxHQVN/umlBFWuxaVEiNDCzFhQjHAVN/elmBdUD77ulJyj0NtO01cUbk5c/vu1GR0i44Z9GEhQYdtLqpRgtQeae7TQ1lhQk883lpxwOYTPV/LKCRAvs9l/DHOSguzxUqkLDgZ9uUFBTqeqrdo71sHdBc4ajoKk/pSDw08QExRk0LmNBahSe0HHjshbU+DtK+HZH7DKoGnpu6rE7AT9NUJAaFu32ucqQaMkLEk8slBlXQ8tndwJ+mpig6znGoEY5C/pTDzIcNMTzDgS5nld47k7AT5MSVPYV9VqD6qUtqM9BYWQsKA4UNF+4pAVdz6duvPhlCunnpzFJC4rGeEF4CQWBJakJEjfiq9BBMHMWVIrBsI9N6XPbbPrtju0JEk8FxdOKhZoa2xRUimGuKcixenm8nsXEKctMNrVBQe2FkJi3IezOfc6CMOqBcx3lZZZ8BcluC+H9gzYoCL7qJAVpNcG9OzYoqITVlxCk5+oK7h+0PUHOHi09+85BzmKlRzX4j02M/kHbEzSugpfXAu6XfXIWNNAryivc5BVTF7T0DTO8JDFB0wh+mSV3QY9RdnC7o1GzT5ZhPYRyFlQfvirRml/mweH2BIlXmsXlH+8HOVYXF4pCEO8oOlY3Oah037W3p1reYS2myqBq6HLx9omyV86CVAYZ7kUF2yNZCxoDao/sXlC0cNsTFGdywXwFlWYeV15Jv1y97mdpD7rpkasgq3byejY/fesbE2S1wOZpaox4ez1xQXDwreEo6HbHiN3MWhDcOgWBrVMQ2Pr2BcF2uufuuJZvTFA0KMgzHAWBcBQEwlEQCEdBIBwFgXAUBMJREAhHQSDc3gU93O7welsua0EP4bxST0FguxQEtktBYLtZCBo5M8tuBY2dmWWvgkbPq7FXQaNnZtmrIOYgxNiZWXYraOzMLPsVNDIcBYFwFAReZqGgpyjwdkdCrCEoYygIELmxmh+RG6v5Ebmp4dxMQkza8diN1YDN4FWiBFm5sRqwmQwEjWisBmwmB0G4sRqwmSwEeUNB4ZuhoOBVKGiJIBQUYZWA1fcHBQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAswrqLaGtJIPHX0ep9lBmst74Tth7uW37klMjYba6plVkBhTxowrc/tsP1QeibODNPXbd+M5ccP13D2qugsJmFOQekykh5NR4+5MHzrlLoj6w2v8lbp/YH4XEjGnoBdOxh43R5DLT99p9Ori1D3snHSsZhUkk3P3EBYPnD8YpP7xz9l7FsZe0PN+uZlTkMoudqbx6PRwF6QSp4nv5Cidkef9GmBRQbVnGd0LOkw49k+RkhP0mJW9Os3cBVHlhuc4a+mdYg+Fod8Ie3dBVKI8i+r0Cun76nRwlNiRQVQHnNBTLJlq/u6C7PLuW/nYV3Wi243VC2daoOQuFEWukZf0ot6pVL8QjynJrSBqbqqQal5FqRJpauQABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCJC8IDnYQ+jknQGkLkh2msFTKM9H4oJ0ZzLfbokRSFyQmbu8CprcNIS0Bfn2lYpI2oKsMa/WYguCymLFcfdSF2SG3aOg15hCmoIcmD7jFORC9T4vfV+iCyd1QaqpseLQqMkLWhsKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAjwP8f2nrJKy2JfAAAAAElFTkSuQmCC" /><!-- --></p>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
</div>
<div id="question-6.10" class="section level2">
<h2>Question 6.10</h2>
<p><font color="#000000"></p>
<p>Construct an approximate 95% con???dence intervalfor the $ Gini $ ratio $ = E[G] $ if $ X $ is lognormal with unknown parameters. Assess the coverage rate of the estimation procedure with a $ Monte Carlo $ experiment.</p>
<p><font color="#FF7F50"></p>
<div id="answer-6.10" class="section level3">
<h3>Answer 6.10</h3>
<p><font color="#000000"></p>
<p>From the codes below, we can get the estimate of G and a 95% CI. Accoding to the relevent theory, we know that the CI is <span class="math display">\[ [RG-sd(G)/sqrt(n)*t_{(1-\frac{\alpha}2)}(n-1),RG+sd(G)/sqrt(n)*t_{(1-\frac{\alpha}2)}(n-1)]\]</span>,and the RG is the real value of the expectation of $ G $ computed with the specific distribution of $ X $.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="dv">20</span>  ## numbers of samples
m &lt;-<span class="st"> </span><span class="dv">1000</span>  ## experiment times
G &lt;-<span class="st"> </span><span class="kw">numeric</span>(m) 
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m) {
  x &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">rlnorm</span>(n)) ## X is standard lognormal
  mu &lt;-<span class="st"> </span><span class="kw">mean</span>(x)  ## let the mu be the mean values of the samples
  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n) {
    t[j] &lt;-<span class="st"> </span>(<span class="dv">2</span><span class="op">*</span>j<span class="op">-</span>n<span class="op">-</span><span class="dv">1</span>)<span class="op">*</span>x[j]
    <span class="kw">sum</span>(t)
  }
  G[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(t)<span class="op">/</span>(mu<span class="op">*</span>n<span class="op">^</span><span class="dv">2</span>)
  G[i]
}
<span class="kw">mean</span>(G) ## compute the estimate values(mean) of G</code></pre></div>
<pre><code>## [1] 0.4814059</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha &lt;-<span class="st"> </span><span class="fl">0.05</span> 
sd &lt;-<span class="st"> </span><span class="dv">1</span>
EG &lt;-<span class="st"> </span><span class="dv">2</span><span class="op">*</span><span class="kw">pnorm</span>(sd<span class="op">/</span><span class="kw">sqrt</span>(<span class="dv">2</span>))<span class="op">-</span><span class="dv">1</span> ## compute the integration of G to get the real mean values
  UCL &lt;-<span class="st"> </span><span class="kw">mean</span>(G)<span class="op">+</span><span class="kw">sd</span>(G)<span class="op">*</span><span class="kw">qt</span>(alpha<span class="op">/</span><span class="dv">2</span>, <span class="dt">df =</span> n<span class="op">-</span><span class="dv">1</span>)<span class="op">/</span><span class="kw">sqrt</span>(n) ## compute the upper confidence limit
  LCL &lt;-<span class="st"> </span><span class="kw">mean</span>(G)<span class="op">-</span><span class="kw">sd</span>(G)<span class="op">*</span><span class="kw">qt</span>(alpha<span class="op">/</span><span class="dv">2</span>, <span class="dt">df =</span> n<span class="op">-</span><span class="dv">1</span>)<span class="op">/</span><span class="kw">sqrt</span>(n) ## compute the lower confidence limit
<span class="kw">c</span>(UCL,LCL)</code></pre></div>
<pre><code>## [1] 0.4436828 0.5191289</code></pre>
<p>Then, we estimate the coverage rate according to the hypothesis test.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="dv">20</span>  ## numbers of samples
m &lt;-<span class="st"> </span><span class="dv">1000</span>  ## experiment times
G &lt;-<span class="st"> </span><span class="kw">numeric</span>(m) 
UCL=<span class="kw">numeric</span>(<span class="dv">1000</span>)
LCL=<span class="kw">numeric</span>(<span class="dv">1000</span>)
<span class="cf">for</span>(ii <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">1000</span>) {
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m) {
  x &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">rlnorm</span>(n)) ## X is standard lognormal
  mu &lt;-<span class="st"> </span><span class="kw">mean</span>(x)  ## let the mu be the mean values of the samples
  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n) {
    t[j] &lt;-<span class="st"> </span>(<span class="dv">2</span><span class="op">*</span>j<span class="op">-</span>n<span class="op">-</span><span class="dv">1</span>)<span class="op">*</span>x[j]
    <span class="kw">sum</span>(t)
  }
  G[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(t)<span class="op">/</span>(mu<span class="op">*</span>n<span class="op">^</span><span class="dv">2</span>)
  G[i]
}
<span class="kw">mean</span>(G) ## compute the estimate values(mean) of G
alpha &lt;-<span class="st"> </span><span class="fl">0.05</span> 
sd &lt;-<span class="st"> </span><span class="dv">1</span>
EG &lt;-<span class="st"> </span><span class="dv">2</span><span class="op">*</span><span class="kw">pnorm</span>(sd<span class="op">/</span><span class="kw">sqrt</span>(<span class="dv">2</span>))<span class="op">-</span><span class="dv">1</span> ## compute the integration of G to get the real mean values
  UCL[ii] &lt;-<span class="st"> </span><span class="kw">mean</span>(G)<span class="op">+</span><span class="kw">sd</span>(G)<span class="op">*</span><span class="kw">qt</span>(alpha<span class="op">/</span><span class="dv">2</span>, <span class="dt">df =</span> n<span class="op">-</span><span class="dv">1</span>)<span class="op">/</span><span class="kw">sqrt</span>(n) ## compute the upper confidence limit
  LCL[ii] &lt;-<span class="st"> </span><span class="kw">mean</span>(G)<span class="op">-</span><span class="kw">sd</span>(G)<span class="op">*</span><span class="kw">qt</span>(<span class="dv">1</span><span class="op">-</span>alpha<span class="op">/</span><span class="dv">2</span>, <span class="dt">df =</span> n<span class="op">-</span><span class="dv">1</span>)<span class="op">/</span><span class="kw">sqrt</span>(n) ## compute the lower confidence limit
}
<span class="kw">sum</span>(EG<span class="op">&lt;</span>UCL <span class="op">&amp;</span><span class="st"> </span>EG<span class="op">&gt;</span>LCL)  <span class="co">#count the number of intervals that satisfy the conditions</span></code></pre></div>
<pre><code>## [1] 0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mean</span>(UCL<span class="op">&gt;</span>EG <span class="op">&amp;</span><span class="st"> </span>EG<span class="op">&gt;</span>LCL) ## compute the coverage rate</code></pre></div>
<pre><code>## [1] 0</code></pre>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="question-6.b" class="section level2">
<h2>Question 6.B</h2>
<p><font color="#000000"></p>
<p>Tests for association based on $ Pearson $ product moment correlation $ $, $ Spearman??s $ rank correlation coe???cient $ <em>{s} $, or $ Kendall??s $ coe???cient $ $ ,are implemented in $ cor.test $. Show (empirically) that the nonparametric tests based on $ </em>{s} $ or $ $ are less powerful than the correlation test when the sampled distribution is bivariate normal. Find an example of an alternative (a bivariate distribution $ (X,Y) $ such that $ X $ and $ Y $ are dependent) such that at least one of the nonparametric tests have better empirical power than the correlation test against this alternative.</p>
<p><font color="#FF7F50"></p>
<div id="answer-6.b" class="section level3">
<h3>Answer 6.B</h3>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="question-7.1" class="section level1">
<h1>Question 7.1</h1>
<p><font color="#000000"></p>
<p>Compute a <span class="math inline">\(jackknife\)</span> estimate of the bias and the standard error of the correlation statistic in <span class="math inline">\(Example 7.2\)</span> (p202).</p>
<p><font color="#800000"></p>
<div id="answer-7.1" class="section level2">
<h2>Answer 7.1</h2>
<p><font color="#000000"></p>
<div id="ideas" class="section level3">
<h3>Ideas</h3>
<p>If <span class="math inline">\(\hat\theta\)</span> is a smooth (plug-in) statistic, then <span class="math inline">\(\hat\theta_{(i)} = t(F_{n-1}(x_{(i)}))\)</span>, and the jackknife estimate of bias is <span class="math display">\[\hat{bias}_{jack}=(n-1)(\overline{\hat\theta_{(.)}}-\hat\theta),\]</span> when <span class="math inline">\(\overline{\hat\theta_{(.)}}=\frac1n\sum_{i=1}^{n}\hat\theta_{(i)}\)</span> is the mean of the estimates from the leave-one-out samples, and <span class="math inline">\(\hat\theta= \hat\theta_{(x)}\)</span> is the estimate computed from the original observed sample.</p>
<p>A <span class="math inline">\(jackknife\)</span> estimate of standard error is <span class="math display">\[\hat{se}_{jack}=\sqrt{\frac{n-1}{n}\sum_{i=1}^{n}(\hat\theta_{(i)}-\overline{\hat\theta_{(.)}})^2},\]</span> for smooth statistics <span class="math inline">\(\hat\theta.\)</span></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(bootstrap) <span class="co">#for the law data</span>
<span class="kw">data</span>(law, <span class="dt">package =</span> <span class="st">&quot;bootstrap&quot;</span>) 
n &lt;-<span class="st"> </span><span class="kw">nrow</span>(law) <span class="co"># for all data available in the samples popuation</span>
theta.hat &lt;-<span class="st"> </span><span class="kw">cor</span>(law<span class="op">$</span>LSAT,law<span class="op">$</span>GPA) <span class="co">#compute the correlation between law$LSAT and law$GPA</span>
<span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">'theta.hat=%f'</span>,theta.hat))</code></pre></div>
<pre><code>## [1] &quot;theta.hat=0.776374&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#compute the jackknife replicates, leave-one-out estimates </span>
theta.jack &lt;-<span class="st"> </span><span class="kw">numeric</span>(n) 
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n) 
  theta.jack[i] &lt;-<span class="st"> </span><span class="kw">cor</span>(law<span class="op">$</span>LSAT[<span class="op">-</span>i],law<span class="op">$</span>GPA[<span class="op">-</span>i])
bias &lt;-<span class="st"> </span>(n <span class="op">-</span><span class="st"> </span><span class="dv">1</span>) <span class="op">*</span><span class="st"> </span>(<span class="kw">mean</span>(theta.jack) <span class="op">-</span><span class="st"> </span>theta.hat)
<span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">'bias=%f'</span>,bias)) <span class="co">#jackknife estimate of bias </span></code></pre></div>
<pre><code>## [1] &quot;bias=-0.006474&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">se &lt;-<span class="st"> </span><span class="kw">sqrt</span>((n<span class="op">-</span><span class="dv">1</span>) <span class="op">*</span><span class="st"> </span><span class="kw">mean</span>((theta.jack <span class="op">-</span><span class="st"> </span><span class="kw">mean</span>(theta.jack))<span class="op">^</span><span class="dv">2</span>)) 
<span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">'se=%f'</span>,se)) <span class="co">#jackknife estimate of standard error</span></code></pre></div>
<pre><code>## [1] &quot;se=0.142519&quot;</code></pre>
<p><font color="#800000"></p>
</div>
<div id="explain-the-results-we-get" class="section level3">
<h3>Explain the results we get</h3>
<p><font color="#000000"></p>
<p>The correlation statistic is the correlation betwwen <span class="math inline">\(LSAT\)</span> and <span class="math inline">\(GPA\)</span>, we use the <span class="math inline">\(\theta\)</span> to delegate the target correlation and denote the estimstor <span class="math inline">\(\theta\)</span> as <span class="math inline">\(\hat\theta\)</span>, we compute and get the value <span class="math inline">\(\hat\theta=0.776374\)</span>.Then we use the method of jackknife to estimate its bias and standard error.The final results are <span class="math inline">\(\hat{bias}_{jack}=-0.006474\)</span> and <span class="math inline">\(\hat{se}_{jack}=0.142519.\)</span></p>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
</div>
<div id="question-7.5" class="section level1">
<h1>Question 7.5</h1>
<p><font color="#000000"></p>
<p>Refer to <span class="math inline">\(Exercise 7.4\)</span>(p206). Compute 95% bootstrap confidence intervals for the mean time between failures <span class="math inline">\(\frac1\lambda\)</span> by the standard normal, basic, percentile, and <span class="math inline">\(BCa\)</span> methods. Compare the intervals and explain why they may differ.</p>
<p><font color="#800000"></p>
<div id="answer-7.5" class="section level2">
<h2>Answer 7.5</h2>
<p><font color="#000000"></p>
<div id="ideas-1" class="section level3">
<h3>Ideas</h3>
<p>1,The standard normal bootstrap confidence interval : If <span class="math inline">\(\theta\)</span> is a sample mean and the sample size is large, then the Central Limit Theorem implies that <span class="math display">\[Z=\frac{\hat\theta-E[\hat\theta]}{se(\hat\theta)}\]</span> is approximately standard normal. Hence, if <span class="math inline">\(\hat\theta\)</span> is unbiased for <span class="math inline">\(\theta\)</span>, then an approximate <span class="math inline">\(100(1-\alpha)\)</span>% CI for <span class="math inline">\(\theta\)</span> is the <span class="math inline">\(Z\)</span>-interval <span class="math display">\[[\hat\theta-z_\frac{\alpha}{2}se(\hat\theta),\hat\theta+z_\frac{\alpha}{2}se(\hat\theta)]\]</span> where <span class="math display">\[z_\frac{\alpha}{2} =\Phi^{-1}(1-\frac{\alpha}{2}).\]</span></p>
<p>2,The basic bootstrap confidence interval : The basic bootstrap confidence interval transforms the distribution of the replicates by subtracting the observed statistic.The <span class="math inline">\(100(1-\alpha)\)</span>% confidence limits for the basic bootstrap confidence interval are <span class="math display">\[(2\hat\theta-\hat\theta_{1-\frac{\alpha}{2}},2\hat\theta-\hat\theta_\frac{\alpha}{2}).\]</span></p>
<p>3,The percentile bootstrap confidence interval : Suppose that <span class="math inline">\(\hat\theta^{(1)},\cdots,\hat\theta^{(B)}\)</span> are the bootstrap replicates of the statistic <span class="math inline">\(\hat\theta\)</span>. From the ecdf of the replicates, compute the <span class="math inline">\(\frac\alpha2\)</span> quantile <span class="math inline">\(\hat\theta_\frac\alpha2\)</span>,and the <span class="math inline">\(1-\frac\alpha2\)</span>quantile<span class="math inline">\(\hat\theta_{1-\frac\alpha2}.\)</span></p>
<p>4,The better bootstrap confidence interval(BCa) : For a <span class="math inline">\(100(1-\alpha)\)</span>% BCa bootstrap confidence interval is <span class="math display">\[(\hat\theta_{\alpha1}^*,\hat\theta_{\alpha2}^*).\]</span></p>
<p>But we can obtain the normal, basic, percentile and the better bootstrap confidence intervals using the <span class="math inline">\(boot\)</span> and <span class="math inline">\(boot.ci\)</span> functions in the <span class="math inline">\(boot\)</span> package.The code generates 95% confidence intervals are follow.</p>
<p><font color="#000000"></p>
<hr />
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(boot) <span class="co">#for boot and boot.ci </span>
<span class="kw">data</span>(aircondit, <span class="dt">package =</span> <span class="st">&quot;boot&quot;</span>) <span class="co">#get the data set and package</span>
theta.boot &lt;-<span class="st"> </span><span class="cf">function</span>(x,ind) { 
  <span class="co">#function to compute the target values </span>
  x &lt;-<span class="st"> </span>x[ind]
  <span class="kw">mean</span>(x) <span class="co">#acording to the exponetial distribution,the $\frac1\lambda$ is exactly the values of the mean x</span>
}
<span class="co">#Run the bootstrap and compute con???dence interval estimates for the bioequivalence ratio</span>
x &lt;-<span class="st"> </span>aircondit<span class="op">$</span>hours <span class="co">#give x the values in the data set aircondit to denote the time intervals</span>
dat &lt;-<span class="st"> </span>x
boot.obj &lt;-<span class="st"> </span><span class="kw">boot</span>(dat, <span class="dt">statistic =</span> theta.boot, <span class="dt">R =</span> <span class="dv">2000</span>)
<span class="co">#The output for the bootstrap and bootstrap con???dence intervals is below.</span>
<span class="kw">print</span>(boot.obj) </code></pre></div>
<pre><code>## 
## ORDINARY NONPARAMETRIC BOOTSTRAP
## 
## 
## Call:
## boot(data = dat, statistic = theta.boot, R = 2000)
## 
## 
## Bootstrap Statistics :
##     original    bias    std. error
## t1* 108.0833 0.7977083    37.27027</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">boot.ci</span>(boot.obj, <span class="dt">type =</span> <span class="kw">c</span>(<span class="st">&quot;basic&quot;</span>, <span class="st">&quot;norm&quot;</span>, <span class="st">&quot;perc&quot;</span>,<span class="st">&quot;bca&quot;</span>)))</code></pre></div>
<pre><code>## BOOTSTRAP CONFIDENCE INTERVAL CALCULATIONS
## Based on 2000 bootstrap replicates
## 
## CALL : 
## boot.ci(boot.out = boot.obj, type = c(&quot;basic&quot;, &quot;norm&quot;, &quot;perc&quot;, 
##     &quot;bca&quot;))
## 
## Intervals : 
## Level      Normal              Basic         
## 95%   ( 34.2, 180.3 )   ( 22.6, 168.9 )  
## 
## Level     Percentile            BCa          
## 95%   ( 47.3, 193.6 )   ( 60.7, 224.8 )  
## Calculations and Intervals on Original Scale
## Some BCa intervals may be unstable</code></pre>
</div>
<div id="explain-the-results-we-get-1" class="section level3">
<h3>Explain the results we get</h3>
<p>From the codes above, we get the standard normal,the basic,the percentile and the better bootstrap 95% CI are <span class="math inline">\(( 33.8, 181.2 ),\)</span> <span class="math inline">\(( 25.6, 172.1 ),\)</span> <span class="math inline">\(( 44.1, 190.6 ),\)</span> <span class="math inline">\(( 55.2, 229.3 )\)</span> respectively. And we compare them then know that the results of the standard normal,the basic,the percentile methods are different but with subtle difference.Meanwhile, the results of the BCa method are different from that of other methods notably.</p>
<p>(1)The standard normal bootstrap CI is the simplest approach, but not necessarily the best.In this method, we have made several assumptions:to apply the normal distribution, we assume that the distribution of <span class="math inline">\(\hat\theta\)</span> is normal or <span class="math inline">\(\hat\theta\)</span>is a sample mean and the sample size is large; We have also implicitly assumed that <span class="math inline">\(\hat\theta\)</span> is unbiased for <span class="math inline">\(\theta.\)</span></p>
<p>(2)The basic bootstrap CI transforms the distribution of the replicates by subtracting the observed statistic. We reuse the ecdf and depend on it strongly than other methods and we do not utilize the information sufficently,but we make no assumptions,so the result is more fitted to some extent.</p>
<p>(3)A bootstrap percentile interval uses the empirical distribution of the bootstrap replicates as the reference distribution. The quantiles of the empirical distribution are estimators of the quantiles of the sampling distribution of <span class="math inline">\(\hat\theta\)</span>, so that these (random) quantiles may match the true distribution better when the distribution of <span class="math inline">\(\hat\theta\)</span> is not normal.</p>
<p>(4)The bias-corrected and accelerated (BCa)is a modified version of percentile intervals that have better theoretical properties and better performance in practice.And also we can get the same conclusion from the results.</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
</div>
<div id="question-7.8" class="section level1">
<h1>Question 7.8</h1>
<p><font color="#000000"></p>
<p>Refer to <span class="math inline">\(Exercise 7.7\)</span>(p210). Obtain the <span class="math inline">\(jackknife\)</span> estimates of bias and standard error of <span class="math inline">\(\hat\theta\)</span>.</p>
<p><font color="#800000"></p>
<div id="answer-7.8" class="section level2">
<h2>Answer 7.8</h2>
<p><font color="#000000"></p>
<div id="ideas-2" class="section level3">
<h3>Ideas</h3>
<p>If we want to compute the jackknife estimates of <span class="math inline">\(\hat\theta\)</span>, we must to get the covariance matrix of the samples popuation and compute its all eigenvalues. Then it is easier for us to compute the <span class="math inline">\(\hat\theta\)</span>. Finally, we just need to estimate with three methods of jackknife.The simple procedures are:</p>
<ol style="list-style-type: decimal">
<li><p>Get the data set and generate a covariance matrix <span class="math inline">\(\Sigma.\)</span></p></li>
<li><p>Compute all eigenvalues of the covariance matrix : <span class="math inline">\(\lambda_1,\cdots,\lambda_5\)</span> and compute the will-estimated parameter <span class="math inline">\(\hat\theta=\frac{\hat\lambda_1}{\sum_{i=1}^{5}\hat\lambda_i}.\)</span></p></li>
<li><p>Estimate <span class="math inline">\(\hat\theta\)</span> with the jackknife : its <span class="math inline">\(jackknife\)</span> estimate of bias is <span class="math display">\[\hat{bias}_{jack}=(n-1)(\overline{\hat\theta_{(.)}}-\hat\theta),\]</span> and its <span class="math inline">\(jackknife\)</span> estimate of standard error is <span class="math display">\[\hat{se}_{jack}=\sqrt{\frac{n-1}{n}\sum_{i=1}^{n}(\hat\theta_{(i)}-\overline{\hat\theta_{(.)}})^2},\]</span></p></li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(bootstrap) <span class="co">#for the sample data</span>
<span class="kw">data</span>(scor) <span class="co"># duplicate the data scor to &quot;data&quot;</span>
<span class="kw">str</span>(scor) <span class="co"># the structure of the sample data</span></code></pre></div>
<pre><code>## 'data.frame':    88 obs. of  5 variables:
##  $ mec: num  77 63 75 55 63 53 51 59 62 64 ...
##  $ vec: num  82 78 73 72 63 61 67 70 60 72 ...
##  $ alg: num  67 80 71 63 65 72 65 68 58 60 ...
##  $ ana: num  67 70 66 70 70 64 65 62 62 62 ...
##  $ sta: num  81 81 81 68 63 73 68 56 70 45 ...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="kw">nrow</span>(scor) <span class="co"># for all data available in the samples popuation</span>
cov &lt;-<span class="st"> </span><span class="kw">cov</span>(scor) <span class="co"># compute the covariance matrix</span>
cov</code></pre></div>
<pre><code>##          mec       vec       alg       ana       sta
## mec 305.7680 127.22257 101.57941 106.27273 117.40491
## vec 127.2226 172.84222  85.15726  94.67294  99.01202
## alg 101.5794  85.15726 112.88597 112.11338 121.87056
## ana 106.2727  94.67294 112.11338 220.38036 155.53553
## sta 117.4049  99.01202 121.87056 155.53553 297.75536</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">e &lt;-<span class="st"> </span><span class="kw">eigen</span>(cov) <span class="co">#compute the eigenvalues of the covariance matrix</span>
e</code></pre></div>
<pre><code>## eigen() decomposition
## $values
## [1] 686.98981 202.11107 103.74731  84.63044  32.15329
## 
## $vectors
##            [,1]        [,2]       [,3]         [,4]        [,5]
## [1,] -0.5054457  0.74874751 -0.2997888  0.296184264 -0.07939388
## [2,] -0.3683486  0.20740314  0.4155900 -0.782888173 -0.18887639
## [3,] -0.3456612 -0.07590813  0.1453182 -0.003236339  0.92392015
## [4,] -0.4511226 -0.30088849  0.5966265  0.518139724 -0.28552169
## [5,] -0.5346501 -0.54778205 -0.6002758 -0.175732020 -0.15123239</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>e<span class="op">$</span>values[<span class="dv">1</span>] <span class="co"># The maximum eigenvaues of the covariance matrix</span>
<span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">'x=%f'</span>,x))</code></pre></div>
<pre><code>## [1] &quot;x=686.989810&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw">sum</span>(e<span class="op">$</span>values[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]) <span class="co"># The sum values of all eigenvaues of the covariance matrix</span>
<span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">'y=%f'</span>,y))</code></pre></div>
<pre><code>## [1] &quot;y=1109.631923&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">theta.hat &lt;-<span class="st"> </span>x<span class="op">/</span>y <span class="co">#The contribution rate of the covariance matrix</span>
<span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">'theta.hat=%f'</span>,x<span class="op">/</span>y))</code></pre></div>
<pre><code>## [1] &quot;theta.hat=0.619115&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#compute the jackknife replicates, leave-one-out estimates </span>
theta.jack &lt;-<span class="st"> </span><span class="kw">numeric</span>(n)
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n) 
  cov.jack &lt;-<span class="st"> </span><span class="kw">cov</span>(scor[<span class="op">-</span>i])  <span class="co"># compute the covariance matrix in the jackknife</span>
  e &lt;-<span class="st"> </span><span class="kw">eigen</span>(cov.jack) <span class="co">#compute the eigenvalues of the covariance matrix in the jackknife</span>
  x &lt;-<span class="st"> </span>e<span class="op">$</span>values[<span class="dv">1</span>] <span class="co"># The maximum eigenvaues of the jackknife covariance matrix</span>
  y &lt;-<span class="st"> </span><span class="kw">sum</span>(e<span class="op">$</span>values[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]) <span class="co"># The sum values of all eigenvaues of the jackknife covariance matrix</span>
  theta.jack[i] &lt;-<span class="st"> </span>x<span class="op">/</span>y <span class="co">#The contribution rate of the jackknife covariance matrix</span>
bias &lt;-<span class="st"> </span>(n <span class="op">-</span><span class="st"> </span><span class="dv">1</span>) <span class="op">*</span><span class="st"> </span>(<span class="kw">mean</span>(theta.jack) <span class="op">-</span><span class="st"> </span>theta.hat)
<span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">'bias=%f'</span>,bias)) <span class="co">#jackknife estimate of bias </span></code></pre></div>
<pre><code>## [1] &quot;bias=-53.250929&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">se &lt;-<span class="st"> </span><span class="kw">sqrt</span>((n<span class="op">-</span><span class="dv">1</span>) <span class="op">*</span><span class="st"> </span><span class="kw">mean</span>((theta.jack <span class="op">-</span><span class="st"> </span><span class="kw">mean</span>(theta.jack))<span class="op">^</span><span class="dv">2</span>))
<span class="kw">print</span>(<span class="kw">sprintf</span>(<span class="st">'se=%f'</span>,se)) <span class="co">#jackknife estimate of standard error</span></code></pre></div>
<pre><code>## [1] &quot;se=0.612080&quot;</code></pre>
</div>
<div id="explain-the-results-we-get-2" class="section level3">
<h3>Explain the results we get</h3>
<p><font color="#000000"></p>
<p>From the results, we can see that the eigrn values of the covariance matrix are <span class="math inline">\(\lambda_1 = 686.98981, \lambda_2 = 202.11107,\lambda_3 = 103.74731, \lambda_4 = 84.63044,\lambda_5 = 32.15329\)</span>. So we compute the <span class="math inline">\(\hat\theta=\frac{\hat\lambda_1}{\sum_{i=1}^{5}\hat\lambda_i}=0.619115.\)</span> Then we estimate the jackknife estimate of standard error and bias of <span class="math inline">\(\hat\theta\)</span> are <span class="math inline">\(\hat{bias_{jack}}=-53.250929, \hat{se_{jack}}=0.612080\)</span> respectively.</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
</div>
<div id="question-7.11" class="section level1">
<h1>Question 7.11</h1>
<p><font color="#000000"></p>
<p>In <span class="math inline">\(Example 7.18\)</span>(p227), leave-one-out (n-fold) cross validation was used to select the best fitting model. Use leave-two-out cross validation to compare the models.</p>
<p><font color="#800000"></p>
<div id="answer-7.11" class="section level2">
<h2>Answer 7.11</h2>
<p><font color="#000000"></p>
<div id="ideas-3" class="section level3">
<h3>Ideas</h3>
<p>The proposed models for predicting magnetic measurement (Y) from chemical measurement (X) are:</p>
<p>1.<span class="math inline">\(Linear: Y=\beta_0+\beta_1X+\epsilon.\)</span></p>
<p>2.<span class="math inline">\(Quadratic: Y=\beta_0+\beta_1X+\beta_2X^2+\epsilon.\)</span></p>
<p>3.<span class="math inline">\(Exponential: log(Y) = log(\beta_0)+\beta_1X + \epsilon.\)</span></p>
<p>4.<span class="math inline">\(Log-Log: log(Y)=\beta_0 + \beta_1log(X)+\epsilon.\)</span></p>
<p>And the leave-two-out cross validation procedure is:</p>
<p>1.For <span class="math inline">\(k =1,\cdots,n\)</span>,let observation <span class="math inline">\((x_k,y_k)\)</span> and <span class="math inline">\((x_{n-k+1},y_{n-k+1})\)</span> be the test point and use the remaining observations to fit the model.</p>
<p>(a)Fit the model(s) using only the n-2 observations in the training set <span class="math inline">\((x_i,y_i),i \neq k,i \neq n-k+1.\)</span></p>
<ol start="2" style="list-style-type: lower-alpha">
<li><p>Compute the predicted response <span class="math inline">\(\hat{y}_k=\hat\beta_0+\hat\beta_1x_k\)</span>for the test point.</p></li>
<li><p>Compute the prediction error <span class="math inline">\(e_{k1}=y_k-\hat{y}_k.\)</span> and <span class="math inline">\(e_{k2}=y_{n-k+1}-\hat{y}_{n-k+1}.\)</span></p></li>
<li><p>compute the mean values of <span class="math inline">\(e_{k1}=y_k-\hat{y}_k.\)</span> and <span class="math inline">\(e_{k2}=y_{n-k+1}-\hat{y}_{n-k+1}\)</span> and denote as <span class="math inline">\(e_k.\)</span></p></li>
</ol>
<p>2.Estimate the mean of the squared prediction errors <span class="math inline">\(\hat\sigma_\epsilon^2=\frac1n\sum_{k=1}^{n}e_k^2.\)</span></p>
<p>The code to estimate the parameters of the four models follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(DAAG)</code></pre></div>
<pre><code>## 载入需要的程辑包：lattice</code></pre>
<pre><code>## 
## 载入程辑包：'lattice'</code></pre>
<pre><code>## The following object is masked from 'package:boot':
## 
##     melanoma</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(ironslag, <span class="dt">package =</span> <span class="st">&quot;DAAG&quot;</span>)
<span class="kw">attach</span>(ironslag)
n &lt;-<span class="st"> </span><span class="kw">length</span>(magnetic) <span class="co">#in DAAG ironslag </span>
e1&lt;-<span class="st"> </span>e11&lt;-<span class="st"> </span>e12&lt;-<span class="st"> </span>e2&lt;-<span class="st"> </span>e21&lt;-<span class="st"> </span>e22&lt;-<span class="st"> </span>e3&lt;-<span class="st"> </span>e31&lt;-<span class="st"> </span>e32&lt;-<span class="st"> </span>e4&lt;-<span class="st"> </span>e41&lt;-<span class="st"> </span>e42 &lt;-<span class="st"> </span><span class="kw">numeric</span>(n)
<span class="co"># for n-fold cross validation </span>
<span class="co"># fit models on leave-two-out samples </span>
<span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n) { 
  y &lt;-<span class="st"> </span>magnetic[{<span class="op">-</span>k}<span class="op">&amp;</span>{<span class="op">-</span>(n<span class="op">-</span>k<span class="op">+</span><span class="dv">1</span>)}] ## leave-two-out samples
  x &lt;-<span class="st"> </span>chemical[{<span class="op">-</span>k}<span class="op">&amp;</span>{<span class="op">-</span>(n<span class="op">-</span>k<span class="op">+</span><span class="dv">1</span>)}] ## leave-two-out samples
  
  J1 &lt;-<span class="st"> </span><span class="kw">lm</span>(y <span class="op">~</span><span class="st"> </span>x) <span class="co">#the Linear model</span>
  yhat11 &lt;-<span class="st"> </span>J1<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>J1<span class="op">$</span>coef[<span class="dv">2</span>] <span class="op">*</span><span class="st"> </span>chemical[k] 
  e11[k] &lt;-<span class="st"> </span>magnetic[k] <span class="op">-</span><span class="st"> </span>yhat11
  yhat12 &lt;-<span class="st"> </span>J1<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>J1<span class="op">$</span>coef[<span class="dv">2</span>] <span class="op">*</span><span class="st"> </span>chemical[n<span class="op">-</span>k<span class="op">+</span><span class="dv">1</span>] 
  e12[k] &lt;-<span class="st"> </span>magnetic[n<span class="op">-</span>k<span class="op">+</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span>yhat12
  e1[k] &lt;-<span class="st"> </span>e11[k]<span class="op">+</span>e12[k]
  
  J2 &lt;-<span class="st"> </span><span class="kw">lm</span>(y <span class="op">~</span><span class="st"> </span>x <span class="op">+</span><span class="st"> </span><span class="kw">I</span>(x<span class="op">^</span><span class="dv">2</span>)) <span class="co">#the Quadratic model</span>
  yhat21 &lt;-<span class="st"> </span>J2<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>J2<span class="op">$</span>coef[<span class="dv">2</span>] <span class="op">*</span><span class="st"> </span>chemical[k] <span class="op">+</span><span class="st"> </span>J2<span class="op">$</span>coef[<span class="dv">3</span>] <span class="op">*</span><span class="st"> </span>chemical[k]<span class="op">^</span><span class="dv">2</span> 
  e21[k] &lt;-<span class="st"> </span>magnetic[k] <span class="op">-</span><span class="st"> </span>yhat21
  yhat22 &lt;-<span class="st"> </span>J2<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>J2<span class="op">$</span>coef[<span class="dv">2</span>] <span class="op">*</span><span class="st"> </span>chemical[n<span class="op">-</span>k<span class="op">+</span><span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>J2<span class="op">$</span>coef[<span class="dv">3</span>] <span class="op">*</span><span class="st"> </span>chemical[n<span class="op">-</span>k<span class="op">+</span><span class="dv">1</span>]<span class="op">^</span><span class="dv">2</span> 
  e22[k] &lt;-<span class="st"> </span>magnetic[n<span class="op">-</span>k<span class="op">+</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span>yhat22
  e2[k] &lt;-<span class="st"> </span>e21[k]<span class="op">+</span>e22[k]
  
  J3 &lt;-<span class="st"> </span><span class="kw">lm</span>(<span class="kw">log</span>(y) <span class="op">~</span><span class="st"> </span>x) <span class="co">#the Exponential model</span>
  logyhat31 &lt;-<span class="st"> </span>J3<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>J3<span class="op">$</span>coef[<span class="dv">2</span>] <span class="op">*</span><span class="st"> </span>chemical[k] 
  yhat31 &lt;-<span class="st"> </span><span class="kw">exp</span>(logyhat31) 
  e31[k] &lt;-<span class="st"> </span>magnetic[k] <span class="op">-</span><span class="st"> </span>yhat31
  logyhat32 &lt;-<span class="st"> </span>J3<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>J3<span class="op">$</span>coef[<span class="dv">2</span>] <span class="op">*</span><span class="st"> </span>chemical[n<span class="op">-</span>k<span class="op">+</span><span class="dv">1</span>] 
  yhat32 &lt;-<span class="st"> </span><span class="kw">exp</span>(logyhat32) 
  e32[k] &lt;-<span class="st"> </span>magnetic[n<span class="op">-</span>k<span class="op">+</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span>yhat32
  e3[k] &lt;-<span class="st"> </span>e31[k]<span class="op">+</span>e32[k]
  
  J4 &lt;-<span class="st"> </span><span class="kw">lm</span>(<span class="kw">log</span>(y) <span class="op">~</span><span class="st"> </span><span class="kw">log</span>(x)) <span class="co">#the Log-Log model</span>
  logyhat41 &lt;-<span class="st"> </span>J4<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>J4<span class="op">$</span>coef[<span class="dv">2</span>] <span class="op">*</span><span class="st"> </span><span class="kw">log</span>(chemical[k]) 
  yhat41 &lt;-<span class="st"> </span><span class="kw">exp</span>(logyhat41) 
  e41[k] &lt;-<span class="st"> </span>magnetic[k] <span class="op">-</span><span class="st"> </span>yhat41
  logyhat42 &lt;-<span class="st"> </span>J4<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>J4<span class="op">$</span>coef[<span class="dv">2</span>] <span class="op">*</span><span class="st"> </span><span class="kw">log</span>(chemical[n<span class="op">-</span>k<span class="op">+</span><span class="dv">1</span>]) 
  yhat42 &lt;-<span class="st"> </span><span class="kw">exp</span>(logyhat42) 
  e42[k] &lt;-<span class="st"> </span>magnetic[n<span class="op">-</span>k<span class="op">+</span><span class="dv">1</span>] <span class="op">-</span><span class="st"> </span>yhat42
  e4[k] &lt;-<span class="st"> </span>e41[k]<span class="op">+</span>e42[k]
} 
 <span class="kw">c</span>(<span class="kw">mean</span>(e1<span class="op">^</span><span class="dv">2</span>), <span class="kw">mean</span>(e2<span class="op">^</span><span class="dv">2</span>), <span class="kw">mean</span>(e3<span class="op">^</span><span class="dv">2</span>), <span class="kw">mean</span>(e4<span class="op">^</span><span class="dv">2</span>)) </code></pre></div>
<pre><code>## [1] 45.95909 42.44307 43.84268 46.08802</code></pre>
<p>Now according the results, we display the specific properties and plots for the model 2:</p>
</div>
<div id="explain-the-results-we-get-3" class="section level3">
<h3>Explain the results we get</h3>
<p>According to the prediction error criterion, Model 2, the <span class="math inline">\(Quadratic\)</span> model, would be the best fit for the data. And now we show some relative parameter values and plots of this model. The best model is L2(the <span class="math inline">\(Quadratic\)</span> model). The fitted regression equation for Model 2 is <span class="math display">\[\hat{Y}= 24.49262 -1.39334X + 0.05452X^2.\]</span></p>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="question" class="section level2">
<h2>Question</h2>
<div id="exercise-8.1" class="section level3">
<h3>Exercise 8.1</h3>
<p><font color="#000000"></p>
<p>Implement the two-sample Cramer-von Mises test for equal distributions as a permutation test. Apply the test to the data in Examples 8.1 and 8.2.</p>
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="answer-exercise-8.1" class="section level2">
<h2>Answer Exercise 8.1</h2>
<p><font color="#000000"></p>
<div id="ideas-4" class="section level3">
<h3>Ideas</h3>
<p>The Cramer-von Mises statistic, which estimates the integrated squared distance between the distributions, is defined by <span class="math display">\[W_2=\frac{mn}{(m+n)^2}[\sum^n_{i=1}(F_n(x_i)-G_m(x_i))^2+\sum^m_{j=1}(F_n(x_j)-G_m(x_j))^2].\]</span><br />
where <span class="math inline">\(F_n\)</span> is the ecdf of the sample <span class="math inline">\(x_1,...,x_n\)</span> and <span class="math inline">\(G_m\)</span> is the ecdf of the sample <span class="math inline">\(y_1,...,y_m\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">attach</span>(chickwts)
x &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">as.vector</span>(weight[feed <span class="op">==</span><span class="st"> &quot;soybean&quot;</span>]))
y &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">as.vector</span>(weight[feed <span class="op">==</span><span class="st"> &quot;linseed&quot;</span>]))
<span class="kw">detach</span>(chickwts)

R &lt;-<span class="st"> </span><span class="dv">999</span><span class="co">#number of replicates</span>
z &lt;-<span class="st"> </span><span class="kw">c</span>(x,y)<span class="co">#pooled sample</span>
K &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(z)
D &lt;-<span class="st"> </span><span class="kw">numeric</span>(R) <span class="co">#storage for replicates</span>

CM &lt;-<span class="st"> </span><span class="cf">function</span>(x,y){<span class="co">#a function to compute the statistic of Cramer-von Mises</span>
  ecdfx &lt;-<span class="st"> </span><span class="kw">ecdf</span>(x)
  ecdfy &lt;-<span class="st"> </span><span class="kw">ecdf</span>(y)
  n1 &lt;-<span class="st"> </span><span class="kw">length</span>(x)
  n2 &lt;-<span class="st"> </span><span class="kw">length</span>(y)
  
  sum1 &lt;-<span class="st"> </span><span class="kw">sum</span>((<span class="kw">ecdfx</span>(x)<span class="op">-</span><span class="kw">ecdfy</span>(x))<span class="op">^</span><span class="dv">2</span>)
  sum2 &lt;-<span class="st"> </span><span class="kw">sum</span>((<span class="kw">ecdfx</span>(y)<span class="op">-</span><span class="kw">ecdfy</span>(y))<span class="op">^</span><span class="dv">2</span>)
  w &lt;-<span class="st"> </span>n1<span class="op">*</span>n2<span class="op">/</span>(n1<span class="op">+</span>n2)<span class="op">^</span><span class="dv">2</span><span class="op">*</span>(sum1<span class="op">+</span>sum2)
  <span class="kw">return</span>(w)
}

D0 &lt;-<span class="st"> </span><span class="kw">CM</span>(x,y)

<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>R) {
  k &lt;-<span class="st"> </span><span class="kw">sample</span>(K, <span class="dt">size =</span> <span class="kw">length</span>(x), <span class="dt">replace =</span> F)
  x1 &lt;-<span class="st"> </span>z[k]
  y1 &lt;-<span class="st"> </span>z[<span class="op">-</span>k] <span class="co">#complements of X1</span>
  D[i] &lt;-<span class="st"> </span><span class="kw">CM</span>(x1,y1)
}

p &lt;-<span class="st"> </span><span class="kw">mean</span>(<span class="kw">c</span>(D0, D) <span class="op">&gt;=</span><span class="st"> </span>D0)
<span class="kw">print</span>(p)</code></pre></div>
<pre><code>## [1] 0.408</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(D, <span class="dt">main =</span> <span class="st">&quot;&quot;</span>, <span class="dt">freq =</span> <span class="ot">FALSE</span>, <span class="dt">xlab =</span> <span class="st">&quot;D (p = 0.412)&quot;</span>, <span class="dt">breaks =</span> <span class="st">&quot;scott&quot;</span>)
<span class="kw">points</span>(D0, <span class="dv">0</span>, <span class="dt">cex =</span> <span class="dv">1</span>, <span class="dt">pch =</span> <span class="dv">16</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAUVBMVEUAAAAAADoAAGYAOjoAOpAAZrY6AAA6ADo6AGY6kNtmAABmADpmkJBmtv+QOgCQZgCQtpCQ2/+2ZgC2///bkDrb////tmb/25D//7b//9v////Ql8DGAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAHZklEQVR4nO2da2OiOBhG7WxrV2fH2ToUlf//Q5cQblHDCRcXCM/5Um3jK56GQPAx7jLRyW7uDVg6EgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBLxS0K3ndM/wf9Nv8848/2WW/2719hZfekqDCz2cu53r8FVx6Q4Kux0Mu6cPcTN6/Q0tvStCv7HY6mJtp3pfaVXZPBpztCSp6T9LZgzYu6Hr88afoQqlvlN64oNyM3ZE+QsptUlCfchIE5SQIykkQlJMgKCdBUE6CoJwEQTkJgnISBOUkCMpJEJSTICgnQa0qumA2oJwEQTkJgnISBOUiE3Q9+i+kDiiXRScoy5L8IHSYrlx8grIJHMUuyCpy3xQcXC5CQeZNHfPmacD7ygHlYhOUj9I7ayYd3oUiFnQ9BmVaQstl8Qn6af2M6D7ZJgQlLOiy96aDohV0bubj/sO8GaUqPBqjFdT0oC5Sa2+LPSiQ69Ec5x4FdV4PWneUs++Gn/Nj3ZAetHJB1+OhHmBgkE52hw0K6sNl/5cEdXI77TYpyIR8k2q2Mb5cfILO79+X/UeZgx5fLjpBJgSd5vvOi6YacQgyHzUImGqElMuiE5TvW+Y08HrULubgXA96+7qdxl25j1nQ5OUkCMrFJig//wuZaoSWi0/QeZSZh3LRCSo+LTdduSxCQQGfsgwvl0UnaNQBngNU6xdUzDMmLBedoNALZoHlsugETV9OgqCcK2itl+5bm5zvZO/f53EHe+hBqxaUvn0lZjb/mnzQ+gWZj3ubT8O/9HrQmgWZE0Uj6KVXFNcsqOpB51FX7SMWVI5BybjTxZgF2VPFkSmqqAVNXW6LgpIi5Zn5Y1YRC7Jzse4ROsl3QPu2x/YEncue0ZUjt4srFSnhzQmqFwTqiLbU19TyM4Gk7wpUKxfUulrW8d58uTyXabO1HtSagXVNNaq/5ePV5gTVO1bnVKM6jbydJGhMOee2BD2Wc25L0GM557YjaH3XFStBmKHvU87x8NiDViho4nKOBwl6LCdBUE6CoJwEQTkJgnIkaE0H+/l6kARJkLecBD1UcS+YSRCUkyAoJ0FQToKgnARBuTBB6zhdnLsHSZAEPSknQVCuj6Clj0SzC3rFVkzJkgQtsjctRNDDRG4x9Nsc/FSiT1D56lfYj3ptSJ0eSn0xIo+g6iX32N36bNcr6bMhdfyFvoDtXlDdKXqPR20ef99j2wfT50n8b1B3vYrF8TpBAT3I8yRDtmwh9ByDyi7kHYN8z7JaPz0Hwyrk0PvjCqv1s5yjxVJ5iaBZxt5ORryW6bQEFPX9Ya7fByBBgAQBEgRIECBBgAQBEgRIEKCpBiBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIEDCloLS9gFUasJrVfRtaP81pX69l1MXl7ypDUCR3BiyBOKEgs4JMvYqMcyfkAeY+vGPrtDdrGeH6vNdjFbK4nfL2yYAvI55OkI02lIvHOHdCHpCVK4GGP8EHPoHpcZUgu+zPgEUipxPkbELI9ty3Sd5/dwpy2ocISneHu5gO9eknTCjo0zx5uUXOnZAHFHe7xyC3fdAudr8FA9btn06Q/e+U/yPnTsgDih2oW9BdzZCjwJ2gvqkdw2IE8SqgbnvTGS57esGOoHTIF8YvZRcr7vTYxQIH3fYWDOk/yxmkkzKn0jGoOO1DumjmCOpc4M/Pgg7z1IOc9tYWLnbUNBi6SO2SThTpTNpp33MM4qYeppxqJPa4Yk9RkoCDjPOAjKcaTvtzyNShEGTal7vwnOdBkSJBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBMwk634ddbqfO8IUvMGxjvlXE9/pz3NfoPmMuQcULvOzr7EvSHcf0BIZtzLeJ+Kajvqj6KbMKal50a/24J/gCw2XMt0lWQT8cwryC6uCLialcPv/dP0+0eALDbsy3kN1v8bkQ5hXU/PwwGvIxqUr25neaPI8/MNwSVER8BwShgZkFlb2i+BpTGwJ71gf8geFGkI1odn2P7jCWIuhXHTt80gf8geFaUBnxbX1h7EQsYxcr3JQ70hNB/sBwJaiK+MYmqBqkbQ9qC3LHIG9gOK1+V3qJTVDVXVpj0LMgpz8wbAU1Ed/IxqAmm2uPYt7Pp3gDw6k9Dzq0G07LQqYaxXnQ/p+9L2rvCwwXgloR32jOg+4xg9AEu0f3CfkgFiLIzMUmEBTPXOyBfBY1XlBEs/n1IEGABAESBEgQIEGABAESBEgQIEGABAESBEgQIEGABAESBEgQIEGABAESBEgQ8B/2XyxzHgDTwAAAAABJRU5ErkJggg==" /><!-- --></p>
</div>
<div id="explains" class="section level3">
<h3>Explains</h3>
<p>The approximate achieved significance level 0.412 does not support the alternative hypothesis that distributions differ. So we may say that soybean and linseed groups may be similar.</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="question-8-2" class="section level2">
<h2>Question 8-2</h2>
<p><font color="#000000"></p>
<p>Design experiments for evaluating the performance of the NN, energy, and ball methods in various situations</p>
<ul>
<li>Unequal variances and equal expectations</li>
<li>Unequal variances and unequal expectations</li>
<li>Non-normal distributions: t distribution with 1 df (heavy-tailed distribution), bimodel distribution (mixture of two normal distributions)</li>
<li>Unbalanced Unbalanced samples (say, 1 case versus 10 controls)</li>
<li>Note: The parameters should be choosen such that the powers are distinguishable (say, range from 0.3 to 0.9).</li>
</ul>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-8.2" class="section level2">
<h2>Answer 8.2</h2>
<p><font color="#000000"></p>
<div id="unequal-variances-and-equal-expectations" class="section level3">
<h3>Unequal variances and equal expectations</h3>
<p>In the following code, the first case, where unbalanced samples are with unequal variances and equal expectations is discussed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RANN)## for nn function
<span class="kw">library</span>(boot)
<span class="kw">library</span>(energy)
<span class="kw">library</span>(Ball)
<span class="kw">library</span>(ggplot2)

m &lt;-<span class="st"> </span><span class="dv">30</span><span class="co">#times to loop</span>
k&lt;-<span class="dv">3</span>
p&lt;-<span class="dv">2</span><span class="co">#ncol</span>
<span class="co"># mu &lt;- 0.5</span>
n1 &lt;-<span class="st"> </span>n2 &lt;-<span class="st"> </span><span class="dv">50</span><span class="co">#nrow</span>
R&lt;-<span class="dv">999</span><span class="co">#the number of replications in the bd.test function</span>
n &lt;-<span class="st"> </span>n1<span class="op">+</span>n2
N =<span class="st"> </span><span class="kw">c</span>(n1,n2)

Tn &lt;-<span class="st"> </span><span class="cf">function</span>(z, ix, sizes,k) {
  n1 &lt;-<span class="st"> </span>sizes[<span class="dv">1</span>]
  n2 &lt;-<span class="st"> </span>sizes[<span class="dv">2</span>]
  n &lt;-<span class="st"> </span>n1 <span class="op">+</span><span class="st"> </span>n2
  <span class="cf">if</span>(<span class="kw">is.vector</span>(z)) z &lt;-<span class="st"> </span><span class="kw">data.frame</span>(z,<span class="dv">0</span>)
  z &lt;-<span class="st"> </span>z[ix, ]
  NN &lt;-<span class="st"> </span><span class="kw">nn2</span>(<span class="dt">data=</span>z, <span class="dt">k=</span>k<span class="op">+</span><span class="dv">1</span>)
  block1 &lt;-<span class="st"> </span>NN<span class="op">$</span>nn.idx[<span class="dv">1</span><span class="op">:</span>n1,<span class="op">-</span><span class="dv">1</span>] 
  block2 &lt;-<span class="st"> </span>NN<span class="op">$</span>nn.idx[(n1<span class="op">+</span><span class="dv">1</span>)<span class="op">:</span>n,<span class="op">-</span><span class="dv">1</span>] 
  i1 &lt;-<span class="st"> </span><span class="kw">sum</span>(block1 <span class="op">&lt;</span><span class="st"> </span>n1 <span class="op">+</span><span class="st"> </span>.<span class="dv">5</span>); i2 &lt;-<span class="st"> </span><span class="kw">sum</span>(block2 <span class="op">&gt;</span><span class="st"> </span>n1<span class="op">+</span>.<span class="dv">5</span>) 
  (i1 <span class="op">+</span><span class="st"> </span>i2) <span class="op">/</span><span class="st"> </span>(k <span class="op">*</span><span class="st"> </span>n)
}

eqdist.nn &lt;-<span class="st"> </span><span class="cf">function</span>(z,sizes,k){<span class="co">#NN</span>
  boot.obj &lt;-<span class="st"> </span><span class="kw">boot</span>(<span class="dt">data=</span>z,<span class="dt">statistic=</span>Tn,<span class="dt">R=</span>R,
  <span class="dt">sim =</span> <span class="st">&quot;permutation&quot;</span>, <span class="dt">sizes =</span> sizes,<span class="dt">k=</span>k)
  ts &lt;-<span class="st"> </span><span class="kw">c</span>(boot.obj<span class="op">$</span>t0,boot.obj<span class="op">$</span>t)
  p.value &lt;-<span class="st"> </span><span class="kw">mean</span>(ts<span class="op">&gt;=</span>ts[<span class="dv">1</span>])
  <span class="kw">list</span>(<span class="dt">statistic=</span>ts[<span class="dv">1</span>],<span class="dt">p.value=</span>p.value)
}

p.values &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>,m,<span class="dv">3</span>)<span class="co">#to store p values</span>

<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m) {
  x &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(n1 <span class="op">*</span><span class="st"> </span>p, <span class="dt">sd =</span> <span class="dv">1</span>), <span class="dt">ncol =</span> p)<span class="co">#unequal variances</span>
  y &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(n2 <span class="op">*</span><span class="st"> </span>p, <span class="dt">sd =</span> <span class="fl">1.4</span>), <span class="dt">ncol =</span> p)
  z &lt;-<span class="st"> </span><span class="kw">rbind</span>(x, y)
  p.values[i, <span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">eqdist.nn</span>(z, N, k)<span class="op">$</span>p.value<span class="co">#NN</span>
  p.values[i, <span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">eqdist.etest</span>(z, <span class="dt">sizes =</span> N, <span class="dt">R =</span> R)<span class="op">$</span>p.value<span class="co">#in the energy package</span>
  p.values[i, <span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw">bd.test</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y, <span class="dt">R =</span> <span class="dv">999</span>, <span class="dt">seed =</span> i)<span class="op">$</span>p.value<span class="co">#&quot;Ball Divergence&quot; in the ball package</span>
}

alpha &lt;-<span class="st"> </span><span class="fl">0.1</span><span class="co">#confidence level</span>
pow &lt;-<span class="st"> </span><span class="kw">apply</span>(p.values<span class="op">&lt;</span>alpha,<span class="dv">2</span>,mean)<span class="co">#compute the number of p.values which is less than 0.1 in each column</span>
<span class="kw">print</span>(pow)</code></pre></div>
<pre><code>## [1] 0.1666667 0.3000000 0.8666667</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">power &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">methods =</span> <span class="kw">c</span>(<span class="st">'NN'</span>,<span class="st">'energy'</span>,<span class="st">'Ball'</span>),pow)
<span class="kw">ggplot</span>(power,<span class="kw">aes</span>(methods,pow))<span class="op">+</span><span class="co">#plot</span>
<span class="st">  </span><span class="kw">geom_col</span>(<span class="dt">fill =</span> <span class="st">'orange'</span>)<span class="op">+</span>
<span class="st">  </span><span class="kw">coord_flip</span>()</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAA3lBMVEUAAAAAADoAAGYAOpAAZmYAZrYzMzM6AAA6ADo6AGY6Ojo6OpA6kNtNTU1NTW5NTY5Nbo5NbqtNjqtNjshmAABmADpmAGZmtttmtv9uTU1uTW5uTY5ubo5ubqtujshuq+SOTU2OTW6OTY6Obk2Obm6ObquOjo6OyP+QOgCQOjqQZgCQkGaQ2/+rbk2rbm6rbo6rjk2ryKur5Mir5P+2ZgC2///Ijk3Ijm7IyI7I///bkDrb/7bb///kq27k///r6+v/pQD/tmb/yI7/25D/5Kv//7b//8j//9v//+T///9rL/8wAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAG6UlEQVR4nO3cj1/aRhzGcaqj1k20nW5DncO1c7jV2m0IG7TQgQrc//8P7S4JhB/ic4kBZPk8r1cl8E04eHN3iamkYMijKaz7BTz3ACQCkAhAIgCJACSSDdCXcSYWZ/NIKW1tiU8JkCgBJEoAiRJAogSQKAEkSgCJEkCiBJAoASRKywL61yeZv5sMawB5PiVAogSQKAEkSgCJEkCiBJAoASRKAIkSQKIEkCgBJEoAiRJAogSQKAEkSgCJEkCiBJAoASRKAIkSQKIEkCgBJEoAiRJAogSQKAEkSgCJUiqgu+Oq+3naHC8ANAN00IyAogWApoFO/zkMgUYLAM0ANRvVEChasA++tIlX8QJK0ua6kxSof9EJgcKFqBB/ALnvQaZdCYHCBYDmgIbXf4ZAwQJAc0DBrmy8ANA8kGlHQG4BIL/Ezw8QQACFtwA9XgJIlAASJYBECSBRAkiUABIlgEQJIFECSJQAEiWARAkgUQJIlAASJYBECSBRAkiUABIlgEQJIFECSJQAEiWARGlZQOt5N0t4SoBECSBRAkiUABIlgEQJIFFaFpDXcVCygySAAAIIIIAA8gG6PyrfHxW26gAtALopmtZWvVUE6GEg24EGl0XTTd6F8gN0f7QH0EKgweVe98WVG2gAPTwH3e4UiuZm+zNA7OYByhzIHgCNwiT9EJBLy9HYfRk9aOFu3t2wmwfoiUNsL6lPboBM183Ryaeg/AClDUAAhWm5IZZ8CsoNEJM0u3mAlgnEEFNATNLs5gFaKtDtDueDHj9pn/x0fa6AZk6V9c9LB01zd/ZHqVQd3/vx3UHTLr7+peYupHjWyRXQ4HIKqFEx7UNzd1wxvTed8b1qUOjt13r2Z8WkutDko0nZiZeVqTlo6hCx/3PN9N8G13Nz15WM79kbM7yu9S86f9eidfPQg+ZO2tuBVCrt10ZAE/fswLJAw+u/Ljo5ApqL6ygmvDSgA4rvRT3ItH+tjNbNCdDM72KNYPKJgCbuRXOQuTup5RvIjirHEAHF99zit7YHDT82cwXUGs9BHkdDAdtP47u5APL+L8PhBzddm/b+eITlBSh9cgNkh1m5xZ+/LAS62f4U/hUeQIv/BK/MKVeA0p+TdkOMc9KLgcL/m+ec9GKgtAEIoCCck+ac9JOA0vz5Zq6AZs5JAzQLlOYQMV9ATNJqiKU4RswTEJO0AGKSFkDmdpdJ+vEhxrd9+F0MIIAAAijPQOt5N0t4SoBECSBRAkiUABIlgEQJIFECSJSWBfT0A8X1BiARgEQAEgFIBCARgEQAEgFIBCARgEQAEgFIBCARgEQAEgFIBCARgEQAEgFIBCARgEQAEgFIBCARgEQAEgFIBCCRJwDdHZdKwRUVo7unwSXxAIqBHEd83TKAHgRyV79zXakK0IIedNZxl5wMdUKgrK/Eud7Mve2kc9Cb4NqS4TU56UHzPWj4wV3LtVQ6AOhhINOo9s+rU0MMoEkgOwEFl1A8qQE0BzQ6DmqXSq/fVQFKEoAAAigMQCIAiQAkApAIQCIAiQAkApAIQCIAiQAkApAIQCIAiQAkApAIQCIAiQAkApAIQCIAiQAkApAIl+jyfUqARAkgUQJIlAASJYBECSBRAkiUABIlgEQJIFHKFijOS73K+jdLsh1AIgCJACSSNdD/LgCJACQCkAhAIpkA9c+j77KOluIH/DYLv0jtvolWOmiKrSY2i9b3ay1ezW3m2vNqLgsg90XW9uHEUvyA32bBF6lPaqZRVdtMbRat79fazGo9a+XVXBZA/bfN6But0VL8gN9mPfe6G1X3pfwkrUXr+7U2vZr7SPyaywLo7qwTtDheih/w28zFLtkxMHHtC71ZtL5fa9Orua7k11wWQK67hk1HS/EDfpsZNwAqwSjTH2u8WbS+X2tTzQW3fs09kx7UP69Ej8qJYebJG9UUPag3ntRlc89iDrJ7sfHrlK945skb1RRzUGP0cawGyA2P0V6sEu7FKl57sdFakY/7XIcf1TuNN4vW92tt8kWFA8uvuQyPg9zHk+I4yG42cWCyr/cscWvR+omOg4JXGY40r+Y4khYBSAQgEYBEABIBSAQgEYBEABIBSOT5A92+el8obH82ZnBZKBTtzz1juoWyMTfFVTS/AUA7W/XBpZMJ/7Us1s3Xe3a5vIrmNwGo7LrRVXerbrvOVv12tz747f32Z3u7iuY3AOjVlTH3R+WuG2b2ju05t998+r7uetIKsnlAdu7pFgeXP7i5aAXZAKBgiO3Wuy+ugiFmutu/l03rq++uVtL8JgBNT9K2N1mqbmE1I2wTgNxu3u3Rw928zY21uT9azQjbCKDVjKUFAUgEIJHnD7TmACQCkAhAIgCJACQCkMh/3fCbUdV0jg4AAAAASUVORK5CYII=" /><!-- --></p>
</div>
<div id="unequal-variances-and-unequal-expectations" class="section level3">
<h3>Unequal variances and unequal expectations:</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m) {
  x &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(n1 <span class="op">*</span><span class="st"> </span>p, <span class="dt">mean =</span> <span class="fl">0.4</span>, <span class="dt">sd =</span> <span class="dv">1</span>), <span class="dt">ncol =</span> p)<span class="co">#unequal variances and unequal expectations</span>
  y &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(n2 <span class="op">*</span><span class="st"> </span>p, <span class="dt">mean =</span> <span class="dv">0</span>, <span class="dt">sd =</span> <span class="fl">1.4</span>), <span class="dt">ncol =</span> p)
  z &lt;-<span class="st"> </span><span class="kw">rbind</span>(x, y)
  p.values[i, <span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">eqdist.nn</span>(z, N, k)<span class="op">$</span>p.value
  p.values[i, <span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">eqdist.etest</span>(z, <span class="dt">sizes =</span> N, <span class="dt">R =</span> R)<span class="op">$</span>p.value
  p.values[i, <span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw">bd.test</span>(<span class="dt">x =</span> x,  <span class="dt">y =</span> y,  <span class="dt">R =</span> <span class="dv">999</span>,  <span class="dt">seed =</span> i)<span class="op">$</span>p.value
}
alpha &lt;-<span class="st"> </span><span class="fl">0.1</span>
pow &lt;-<span class="st"> </span><span class="kw">apply</span>(p.values<span class="op">&lt;</span>alpha,<span class="dv">2</span>,mean)
<span class="kw">print</span>(pow)</code></pre></div>
<pre><code>## [1] 0.3666667 0.7666667 0.9666667</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">power &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">methods =</span> <span class="kw">c</span>(<span class="st">'NN'</span>,<span class="st">'energy'</span>,<span class="st">'Ball'</span>),pow)
<span class="kw">ggplot</span>(power,<span class="kw">aes</span>(methods,pow))<span class="op">+</span>
<span class="st">  </span><span class="kw">geom_col</span>(<span class="dt">fill =</span> <span class="st">'lightslategrey'</span>)<span class="op">+</span>
<span class="st">  </span><span class="kw">coord_flip</span>()</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAA3lBMVEUAAAAAADoAAGYAOpAAZmYAZrYzMzM6AAA6ADo6AGY6Ojo6OpA6kNtNTU1NTW5NTY5Nbo5NbqtNjqtNjshmAABmADpmAGZmtttmtv9uTU1uTW5uTY5ubo5ubqtujshuq+R3iJmOTU2OTW6OTY6Obk2Obm6ObquOjo6OyP+QOgCQOjqQZgCQkGaQ2/+rbk2rbm6rbo6rjk2ryKur5Mir5P+2ZgC2///Ijk3Ijm7IyI7I///bkDrb/7bb///kq27k///r6+v/tmb/yI7/25D/5Kv//7b//8j//9v//+T////5eRoRAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAGzElEQVR4nO3cD1/aRgDGcaqjfzbRdnUbbeeYa9fBVtuuQ9ighYEK3Pt/Q7tLjoQEyUMwoiy/5/OpJlzOS77eXdIIqRiSmcpd78B9D0AiAIkAJAKQCEAixQD9G2VhMZFVr99pQVYFgEQFgEQFgEQFgEQFgEQFgEQFgEQFgEQFgEQFgESF2wL6uaAABBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAZQb6PJF03191YsWAEoBHfU8kF8AKAn06p/nIdB8AaAUUK/bDIH8gn3xoU28SVFAeXbrVpMXaPJ2GAKFC76AHhQBmUEjBAoXAFoCmn34FAIFCwAtAQWnsmgBoGUgM/BAbgGg9QIQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAhQFtsi93VgCQKABIFAAkCgASBQCJAoBEwS5fBxUdgEQAEgFIBCARgESKAbo6rl8dV/Y6AK0AOq+a/l6nXwXoeiDbgaatqhnl70LlAbo6PgBoJdC0dTB6cOYGGkDXz0EXjypVc77/BSBO8wClcnMgewE0D5P0qh7UdzT2XEYPWnmad984zQO0IdB8iB3k9SkNkBm5OTr/FFQeoE0DEEBh+m6I5Z+CSgPEJJ0NxGkeoJsBMcQUEJO0Ato0AAEU5OIR94OygNzffDZLSYBSt8omp7Wjnrk8+bNWa0ZrP7456tnFp7+23YMUT4alApq2EkDdhhk8N5cvGmb8bBitNYOC8WF7bL82zG09aLLobDg0UnNQ4hJx8kvbTF4Hz3Nzz5WM1+w3M/vQnrwd/t025elBSzft7UCq1Q7bc6CFNTuwLNDsw19vhyUCWorrKCZ8NKADitd8DzKD3xrzbUsClPq/WDeYfDzQwpqfg8zly3a5geyocgweKF5zi9/aHjT72CsVUD+ag9a4GgrYfopWSwG09p8MZ+/ddG0Gh9EIKwvQ5ikNkB1m9T5vf1n9HsX9z+G78ABa/Ra8OrdcAdoUyPTdEOOetPrbPPekOc0DtJRigLgnnQ3EPWkBtMnbN0sFlLonDVAaaJNLxHIBMUlnA01bG1wjlgmISVr2IIAygczFEybpLKD53w6ZpPm/GEDpACQCkAhAIvcGaFWbWftyZwUAiQKARAFAogAgUQCQKABIFAAkCrhQzB+ARAASAUgEIBGARAASAUgEIBGARAASAUgEIBGARAASAUgEIBGARAASAUgEIBGARAASAUgEIBGARAASAUgEIBGARG4AdPmiVgueqOhXXwWPxAMoBnIc8XPLALoWyD39znWlJkAretDJ0D1yMtQJgXbjSZzrZumw885Bz4JnS4bP5KQHLfeg2Xv3LNda7Qig64FMtzk5bSaGGECLQHYCCh6h+LIN0BLQ/DpoUKs9fdMEKE8AAgigMACJACQCkAhAIgCJACQCkAhAIgCJACQCkAhAIgCJACQCkAhAIgCJACQCkAhAIgCJACQCkAhAIrcOtLCYyKrXS/eILoAAAgigdSoAJCoAJCoAJCoAJCoAJCoAJCoUCxTnod7khhW2UWOhAkCiAkCiAkCiQtFA/7sAJAKQCEAiAIkUAjQ59Z9lnS/FL6gK4Ueo3WfQake9tWr4TVUT8QaugmtEtmHmH4BbPIoigNwHWQfPF5biF1SF4CPUL9um21y9dbKG31Q1kdpgbA9YtWHG3i9xFEUATV73PL5fil9QFcbuELpN93H8NZvwm6omkhu434Nso3v4Kdw+cRRFAF2eDINdiJbiF1QFv/e2My889SKzht9UNZHcwHUl2UY0xBJHUQSQ67/hvvil+AVVwbgO3QhGWeZvOK7hN1VNJNoIvss2IqDEUdx5D5qcNvyrWXNE6id2m7l60DiazrPnodvqQTeYg+xZLNrlrJ1P/cRuM9cc1J3/DtYDKnwOcoNkfhZrhPN/Q5zF5uXex/2KZx8zDjeu4TdVTSzuQziwZBsRUOIoCrwOcj8+13WQrbBwjXKYOT3ETfhN17wOCnYqHGmyjWDj9FFwJS0CkAhAIgCJACQCkAhAIgCJACQCkMj9B7p4/K5S2f9izLRVqVTt1wNjRpW6MefVbTS/A0CP9jrTlpMJ//Ut1vnXB3a5vo3mdwGo7rrR2WivY7vOXufiSWf6+7v9L/b7NprfAaDHZ8ZcHddHbpjZFdtzLr75/H3H9aQtZPeA7Nwzqk5bP7i5aAvZAaBgiD3pjB6cBUPMjPb/qJv+V9+dbaX5XQBKTtK2N1mqUWU7I2wXgNxp3p3Rw9O8zbm1uTrezgjbCaDtjKUVAUgEIJH7D3THAUgEIBGARAASAUgEIJH/AF71jZvekBddAAAAAElFTkSuQmCC" /><!-- --></p>
</div>
<div id="non-normal-distributions-t-distribution-with-1-df-and-bimodel-distribution" class="section level3">
<h3>Non-normal distributions: t distribution with 1 df and bimodel distribution</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m) {
  x &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rt</span>(n1 <span class="op">*</span><span class="st"> </span>p,<span class="dt">df =</span> <span class="dv">1</span>), <span class="dt">ncol =</span> p)<span class="co"># t distribution</span>
  y &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(n2 <span class="op">*</span><span class="st"> </span>p,<span class="dt">sd =</span> <span class="kw">sample</span>(<span class="kw">c</span>(<span class="dv">1</span>,<span class="fl">1.3</span>),<span class="dt">size =</span> n2<span class="op">*</span>p, <span class="dt">prob =</span> <span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">0.5</span>),<span class="dt">replace =</span> T)), <span class="dt">ncol =</span> p)<span class="co">#bimodel distribution</span>
  z &lt;-<span class="st"> </span><span class="kw">rbind</span>(x, y)
  p.values[i, <span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">eqdist.nn</span>(z, N, k)<span class="op">$</span>p.value
  p.values[i, <span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">eqdist.etest</span>(z, <span class="dt">sizes =</span> N, <span class="dt">R =</span> R)<span class="op">$</span>p.value
  p.values[i, <span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw">bd.test</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y, <span class="dt">R =</span> <span class="dv">999</span>, <span class="dt">seed =</span> i)<span class="op">$</span>p.value
}
alpha &lt;-<span class="st"> </span><span class="fl">0.01</span>
pow &lt;-<span class="st"> </span><span class="kw">apply</span>(p.values<span class="op">&lt;</span>alpha,<span class="dv">2</span>,mean)
<span class="kw">print</span>(pow)</code></pre></div>
<pre><code>## [1] 0.5333333 0.9666667 0.0000000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">power &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">methods =</span> <span class="kw">c</span>(<span class="st">'NN'</span>,<span class="st">'energy'</span>,<span class="st">'Ball'</span>),pow)
<span class="kw">ggplot</span>(power,<span class="kw">aes</span>(methods,pow))<span class="op">+</span>
<span class="st">  </span><span class="kw">geom_col</span>(<span class="dt">fill =</span> <span class="st">'mediumpurple'</span>)<span class="op">+</span>
<span class="st">  </span><span class="kw">coord_flip</span>()</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAA3lBMVEUAAAAAADoAAGYAOpAAZmYAZrYzMzM6AAA6ADo6AGY6Ojo6OpA6kNtNTU1NTW5NTY5Nbo5NbqtNjqtNjshmAABmADpmAGZmtttmtv9uTU1uTW5uTY5ubo5ubqtujshuq+SOTU2OTW6OTY6Obk2Obm6ObquOjo6OyP+QOgCQOjqQZgCQkGaQ2/+TcNurbk2rbm6rbo6rjk2ryKur5Mir5P+2ZgC2///Ijk3Ijm7IyI7I///bkDrb/7bb///kq27k///r6+v/tmb/yI7/25D/5Kv//7b//8j//9v//+T///+QXuuaAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAG4UlEQVR4nO3cAV/aRhjHcapjbTfRdnUb6hyuXYdbXdt1CBtUGKjgvf83tLvkSAiY/AnFUJrf8/lUE548XvLl7pKiScUQmVHZ9A587gGQCIBEACQCIBEAiVgP0H9RzCwmIu31jSayCgASBQCJAoBEAUCiACBRAJAoAEgUACQKABIFAImChwL6s7AACCCAAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAmgubo6a7utJN1oAaA7ooOuB/AJASaCTf1+EQNMFgOaAup1mCOQX7Itf24g3KQ4oz45/QuQFGr8ehEDhgk/QgyIg02+EQOECQAtAd+8+hEDBAkALQMGpLFoAaBHI9D2QWwBouQAIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAgigtQGtsi8bSwAkEgCJBEAiAZBIACQSAInEl30d9DABkAiARAAkAiARAInIBro9rN8eVnbaAKUAXVZNb6fdqwJ0P5DtQJPzqhnm70LlAbo93AMoFWhyvjd8dOEGGkD3z0HXjytVc7l7BRCn+bQASEQ6kL0AmgaTdFoP6jkaey6jB6We5t03TvMApcdSQ2wvr09pgMzQzdH5p6DyAK0aAAEURs8NsfxTUGmAmKQ5zasASARDTASTtAhO8yIAEpENdP2Yz4OygNzvfFaLkgDNfVQ2PqsddM3N6V+1WjNa++nVQdcuPvu15R6keDooFdDkPAHUaZj+C3Nz1DCj54NorRkkRvutkf3aMJt70OTDxMLgSMxBiUvE8S8tM34ZPM/NPVcyXrPfzN271vj14J9WiYbYwof2diDVavutKdDMmh1YFuju3d+vByUCWgjXUUz4aEAHFK/5HmT6vzWm25YEaO7/Yp1g8vFAM2t+DjI3x61yA9lR5Rg8ULzmFr+zPejufbdUQL1oDlriaihg+zlaLQXQ0r8yvHvrpmvT349GWFmAVo/SANlhVu/x5y/pf6O4+zH8KzyA0v8Er85HrgBlRPYQ67khxmfS6nfzfCbNaT49ABKRDcRn0nwmrUKd5gHK7kEAZc9BK1wilguISVre7bNS/ykNEJM0k7QMMQc9ZZLOHmLc7ZPdg1YOgAACKAyARAAk4sGBZhYTkfZ66R7RBRBAAAG0TAFAogAgUQCQKABIFAAkCgASBQCJAoBEAUCiACBRAJAoAEgUACQKABIFAIkCgEQBQKIAIFEAkCgASBQAJAoAEgUAiQKARAFAogAgUQCQKABIFAAkCgASBQCJAoBEAUCiID/QzVGtFjxR0a+eBI/EAygGchzxc8sAuhfIPf3OdaUmQCk96HTgHjkZ6oRAySdxfnGRdw56HjxbMnwmJz0oCRQMsbfuWa612gFA9wOZTnN81kwMMYBmgewEFDxC8bgF0ALQ9DqoX6s9e9UEKE982r5sLAGQSAAkEgCJBEAiAZBIACQSAIkEQCIBkEgAJBIAiQRAIgGQSAAkEgCJBEAiAZBIACQSAIkEQCIBkEgAJBIAiQRAIgGQSAAkEgCJBEAiAZBIACQSAIkEQCIBkEgAJBIAiQRAIgGQSBQHFEfu2zby3+dRQMVMAUCiACBRAJAoWDfQFxcAiQBIBEAiABKxFqDxmb+XdboUv6AKwluo3T1otYPuUhV+U9VEvIErcI3INsz0BrjZo1gHkLuRtf9iZil+QRUEt1Aft0ynmb51ssJvqpqY22BkD1i1YUbeL3EU6wAav+x6fL8Uv6AKRu4QOk13O/6STfhNVRPJDdz7INvo7H8It08cxTqAbk4HwS5ES/ELqsDvve3MM0+9yKzwm6omkhu4riTbiIZY4ijWAeT6b7gvfil+QRUY16EbwSjLfIfjCr+paiLRRvBdthEBJY5i4z1ofNbwr2bNEXM/sdPM1YNG0XSePQ89VA/6hDnInsWiXc7a+bmf2GnmmoM60/dgOaC1z0FukEzPYo1w/m+Is9g0733cW3z3PuNw4wq/qWpidh/CgSXbiIASR7HG6yD343NdB9mCmWuU/czpIW7Cb7rkdVCwU+FIk20EG88fBVfSIgASAZAIgEQAJAIgEQCJAEgEQCIAEvH5A10/eVOp7F4ZMzmvVKr2654xw0rdmMtqEc1vAdDjnfbk3MmE/3oW6/KbPbtcL6L5bQCqu250Mdxp266z075+2p78/mb3yn4vovktAHpyYcztYX3ohpldsT3n+tuPP7RdTyogtg/Izj3D6uT8RzcXFRBbABQMsaft4aOLYIiZ4e4fddP76vuLQprfBqDkJG17k6UaVooZYdsA5E7z7owenuZtXFqb28NiRthWABUzllICIBEAifj8gTYcAIkASARAIgASAZAIgET8D60nUTZpcrjuAAAAAElFTkSuQmCC" /><!-- --></p>
</div>
<div id="unbalanced-samples" class="section level3">
<h3>Unbalanced samples</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n1 &lt;-<span class="st"> </span><span class="dv">50</span>
n2 &lt;-<span class="st"> </span><span class="dv">5</span>
n &lt;-<span class="st"> </span>n1<span class="op">+</span>n2
N =<span class="st"> </span><span class="kw">c</span>(n1,n2)
<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>m) {
  x &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(n1<span class="op">*</span>p,<span class="dt">mean =</span> <span class="dv">1</span>), <span class="dt">ncol =</span> p)<span class="co">#100 samples</span>
  y &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(n2<span class="op">*</span>p,<span class="dt">mean =</span> <span class="dv">2</span>), <span class="dt">ncol =</span> <span class="dv">2</span>)<span class="co">#10 samples</span>
  z &lt;-<span class="st"> </span><span class="kw">rbind</span>(x, y)
  p.values[i, <span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">eqdist.nn</span>(z, N, k)<span class="op">$</span>p.value
  p.values[i, <span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">eqdist.etest</span>(z, <span class="dt">sizes =</span> N, <span class="dt">R =</span> R)<span class="op">$</span>p.value
  p.values[i, <span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw">bd.test</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y, <span class="dt">R =</span> <span class="dv">999</span>, <span class="dt">seed =</span> i)<span class="op">$</span>p.value
}
alpha &lt;-<span class="st"> </span><span class="fl">0.1</span>
pow &lt;-<span class="st"> </span><span class="kw">apply</span>(p.values<span class="op">&lt;</span>alpha,<span class="dv">2</span>,mean)
<span class="kw">print</span>(pow)</code></pre></div>
<pre><code>## [1] 0.4 0.9 0.7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">power &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">methods =</span> <span class="kw">c</span>(<span class="st">'NN'</span>,<span class="st">'energy'</span>,<span class="st">'Ball'</span>),pow)
<span class="kw">ggplot</span>(power,<span class="kw">aes</span>(methods,pow))<span class="op">+</span>
<span class="st">  </span><span class="kw">geom_col</span>(<span class="dt">fill =</span> <span class="st">'sienna2'</span>)<span class="op">+</span>
<span class="st">  </span><span class="kw">coord_flip</span>()</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAA3lBMVEUAAAAAADoAAGYAOpAAZmYAZrYzMzM6AAA6ADo6AGY6Ojo6OpA6kNtNTU1NTW5NTY5Nbo5NbqtNjqtNjshmAABmADpmAGZmtttmtv9uTU1uTW5uTY5ubo5ubqtujshuq+SOTU2OTW6OTY6Obk2Obm6ObquOjo6OyP+QOgCQOjqQZgCQkGaQ2/+rbk2rbm6rbo6rjk2ryKur5Mir5P+2ZgC2///Ijk3Ijm7IyI7I///bkDrb/7bb///kq27k///r6+vueUL/tmb/yI7/25D/5Kv//7b//8j//9v//+T///+tHvRPAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAGqElEQVR4nO3cjX/aRADGcdbKtmppN1uVbVbq5qS6blMpKGyg9AV6//8/5F1eCC9NnguGtJjf8/msvXAJl365u2Rpk5ohmand9w489AAkApAIQCIAiQAkUgzQ37PMFZeSXrNe1YbfDyBRA5CoAUjUACRqABI1AIkagEQNQKIGIFEDkKjZFNA/mwhAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQD65ftF2X1/1ZwWAloAO+xFQVABoEejVX0chUFwAaAmo32uHQFHBvvjYJlllI0B59rHo5AWavB2FQGEhqqAHzYDMsBUChQWAVoBuP/weAgUFgFaAgkPZrADQKpAZRkCuAJBfAAIIIIAAAggggAACCCCAAAIIIIAAAggggAACCCCAAAIIIIAAAqgwIO8dKPoHKvr9ABI1AIkagEQNQKIGIFEDkKjZFNBGzoNKDEAiAIkAJAKQCEAi2UA3x82b49pOF6AUoIu6Gex0B3WA7gayHWh6Vjfj/F2oOkA3x/sApQJNz/bHj87dQAPo7jno6kmtbi52PwPEYT4tAImkA9kToDhM0mk9aOBo7LGMHpR6mHffOMwDlB6vIbaf16cyQGbs5uj8U1B1gNYNQACFGbghln8KqgwQkzSHeRWARBhiIkzSIhzmRQASyQa6esL1oCwg9zuf9VIRoKVLZZPTxmHfXJ/81mi0Z0vfvzns2+KznzruQYono0oBTc8WgHotMzwy1y9a5vL5aLbUDiouDzqX9mvLlPKgyRKzMjgW5qCFU8TJjx0zeR08z809VzJZst/M7YfO5O3oz06FhtjKRXs7kBqNg04MNLdkB5YFuv3wx9tRhYBW4jqKCR8N6ICSpagHmeHPrXjdigAt/V+sF0w+EdDcUjQHmeuXnWoD2VHlGCKgZMkVv7Y96PZjv1JAg9kc5HE2FLD9MFusBJD3rwxv37vp2gwPZiOsKkDrpzJAdpg1B/z5SyrQxe6n8K/wAEr/E7wml1wBykj2EBu4IcY16YxJesw1aQ7z2QFIJBuIa9Jck1ZRh3mAsnsQQNlz0BqniNUCYpKWd/us1X8qA8QkzSQtI+agPSbp7CHG3T7ZPWjtAAQQQGEAEgFIZONAPKILIIDCRYCyawASNQCJGoBEDUCiZlNA932id2cAEgFIBCARgEQAEgFIBCARgEQAEgFIBCARgEQAEgFIBCARgEQAEgFIBCARgEQAEgFIBCARgEQAEgFIBCARgEQAEikH6PpFoxE8UTFafBU8Eg+gBMhxJM8tA+hOIPf0O9eV2gCl9KCTkXvkZKgTAm3Bkzi9f8Y7kncOeh48WzJ8Jic9aBEoGGLv3bNcG41DgO4GMr325LS9MMQAmgeyE1DwCMWXHYBWgOLzoGGj8exNG6A8AQgggAACKDUAiQAkApAIQCIAiQAkApAIQCIAiQAkApAIQCIAiQAkApAIQCIAiQAkApAIQCIAiQAk8pCAvHfgP1dt+P0AEjUAiRqARA1AogYgUQOQqAFI1AAkagASNQCJmmKBkjzWq9znRvm3AkgEIBGARIoG+t8FIBGARAASAUikEKDJaXQva1xKXvDZKLyN2t2H1jjs+24Ure3TUrKS28i15dNUlCKA3I2sw6O5UvKCz0bBbdQvO6bXzt5iYaNobZ+Wlla6tFYeTcUpAmjyuh/d0RqVkhd8Nrp0+95ru1vy/VuK1vZpaXEl93H4NBWnCKDrk1HQ7qyUvOCzkYst2XEw9+QLtVG0tk9Liyu5ruTTVJwigFynDXcgKiUv+Gxk3CBoBaNMfbTJRtHaPi0tNBV892kqzoPoQZPTVvSqmByW3rjXzt2DLmdTuuc89ADmIHsUm+2r2OulN+61c89BvfijKBPIDZD4KNYKj2Itj6NYvE7k4z7b24/ZP2uyUbS2T0vzuxMOLJ+m4hR4HuQ+pNznQXajuZOTAzUvJC1Fa+c4Dwr2LxxpPk1F4UxaBCARgEQAEgFIBCARgEQAEgFIBCCRhw909fRdrbb72ZjpWa1Wt1/3jRnXmsZc1MtofguAnux0p2dOJvw3sFgXX+7bcrOM5rcBqOm60fl4p2u7zk73aq87/eXd7mf7vYzmtwDo6bkxN8fNsRtmdsH2nKuvPn3bdT2phGwfkJ17xvXp2XduLiohWwAUDLG97vjReTDEzHj316YZfPHNeSnNbwPQ4iRte5OlGtfKGWHbAOQO8+6IHh7mbS6szc1xOSNsK4DKGUspAUgEIJGHD3TPAUgEIBGARAASAUgEIJF/AYo8qL33NKaDAAAAAElFTkSuQmCC" /><!-- --></p>
<div id="explains-1" class="section level4">
<h4>Explains</h4>
<p>The comments are similar in every chunk, so I only put them in the first chunk.<br />
A function is defined to implement NN method. In every question, a loop is run to compute p values of each method. In my code, confidence level <span class="math inline">\(\alpha\)</span> is defined as 0.1, and the number of p values which are less than 0.1 are obtained by the funcion apply. At last, I use ggplot2 to generate a figure to show the difference of power in three methods.<br />
I have tried a lot of parameters to make the gap of power as large as possible, making the difference of power clear enough.</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
</div>
<div id="exercise-9.3" class="section level2">
<h2>Exercise 9.3</h2>
<p><font color="#000000"></p>
<p>Use the <span class="math inline">\(Metropolis-Hastings\)</span> sampler to generate random variables from a standard Cauchy distribution. Discard the first 1000 of the chain, and compare the deciles of the generated observations with the deciles of the standard <span class="math inline">\(Cauchy\)</span> distribution(see <span class="math inline">\(qcauchy\)</span> or <span class="math inline">\(qt\)</span> with df=1). Recall that a <span class="math inline">\(Cauchy(\theta,\eta)\)</span> distribution has density function <span class="math display">\[f(x)=\frac1{\theta\pi(1+[(x-\eta)/\theta]^2)},-\infty&lt;x&lt;\infty,\theta&gt;0.\]</span> The standard <span class="math inline">\(Cauchy\)</span> has the <span class="math inline">\(Cauchy(\theta =1,\eta= 0)\)</span> density. (Note that the standard <span class="math inline">\(Cauchy\)</span> density is equal to the Student <span class="math inline">\(t\)</span> density with one degree of freedom.)</p>
<hr />
<p><font color="#FF7F50"></p>
<div id="answer-exercise-9.3" class="section level3">
<h3>Answer Exercise 9.3</h3>
<p><font color="#000000"></p>
<div id="ideas-5" class="section level4">
<h4>Ideas</h4>
<p>Use the Metropolis-Hastings sampler to generate random variables from a <span class="math inline">\(Cauchy\)</span> distribution. Implementation of a Metropolis-Hastings sampler for this example is as follows. Note that the base of the array in <span class="math inline">\(R\)</span> is 1, so we initialize the chain at <span class="math inline">\(X_0\)</span> in <span class="math inline">\(x[1]\)</span>.</p>
<ul>
<li>1.Set <span class="math inline">\(g(??|X)\)</span> to the density of <span class="math inline">\(qt(X)\)</span></li>
<li>2.Generate <span class="math inline">\(X_0\)</span> from distribution <span class="math inline">\(qt(1)\)</span> and store in <span class="math inline">\(x[1]\)</span></li>
<li><ol start="3" style="list-style-type: decimal">
<li>Repeat for <span class="math inline">\(i = 2,\cdots,N:\)</span></li>
</ol></li>
<li><ol style="list-style-type: lower-alpha">
<li>Generate <span class="math inline">\(Y\)</span> from <span class="math inline">\(qt(df = X_t) =qt(df= x[i-1])\)</span></li>
</ol></li>
<li><ol start="2" style="list-style-type: lower-alpha">
<li>Generate <span class="math inline">\(U\)</span> from <span class="math inline">\(Uniform(0, 1)\)</span></li>
</ol></li>
<li><ol start="3" style="list-style-type: lower-alpha">
<li>With <span class="math inline">\(X_t = x[i-1]\)</span>, compute<span class="math display">\[r(X_t,Y)=\frac{f(Y)g(X_t|Y)}{f(X_t)g(Y|X_t)},\]</span> where <span class="math inline">\(f\)</span> is the <span class="math inline">\(Cauchy\)</span> density with parameter <span class="math inline">\(\theta=1.\eta=0\)</span>, <span class="math inline">\(g(Y|X_t)\)</span> is the <span class="math inline">\(qt(df = X_t)\)</span> density evaluated at <span class="math inline">\(Y\)</span> , and <span class="math inline">\(g(X_t|Y)\)</span> is the <span class="math inline">\(qt(df = Y)\)</span> density evaluated at <span class="math inline">\(X_t.\)</span> If <span class="math display">\[U \leq r(X_t, Y)\]</span> accept <span class="math inline">\(Y\)</span> and set <span class="math inline">\(X_{t+1} = Y\)</span> ; otherwise set <span class="math inline">\(X_t+1 =X_t\)</span>. Store <span class="math inline">\(X_{t+1}\)</span> in <span class="math inline">\(x[i]\)</span></li>
</ol></li>
<li>(d)Increment t.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">N &lt;-<span class="st"> </span><span class="dv">5000</span>
my_vector &lt;-<span class="st"> </span><span class="kw">numeric</span>(N)
my_vector[<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">rnorm</span>(<span class="dv">1</span>)
counter &lt;-<span class="st"> </span><span class="dv">0</span>
<span class="co">#generates random deviates</span>
u &lt;-<span class="st"> </span><span class="kw">runif</span>(N)

<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span>N) 
{
    xt &lt;-<span class="st"> </span>my_vector[i<span class="op">-</span><span class="dv">1</span>]
    y &lt;-<span class="st"> </span><span class="kw">rnorm</span>(<span class="dv">1</span>, <span class="dt">mean =</span> xt)
    <span class="co">#Density Cauchy distribution</span>
    numerator &lt;-<span class="st"> </span><span class="kw">dcauchy</span>(y)<span class="op">*</span><span class="kw">dnorm</span>(xt, <span class="dt">mean =</span> y)
    denominator &lt;-<span class="st"> </span><span class="kw">dcauchy</span>(xt)<span class="op">*</span><span class="kw">dnorm</span>(y, <span class="dt">mean =</span> xt)
    
    <span class="cf">if</span>(u[i] <span class="op">&lt;=</span><span class="st"> </span>numerator<span class="op">/</span>denominator) 
    {
        my_vector[i] &lt;-<span class="st"> </span>y
    }
    <span class="cf">else</span> 
    {
        my_vector[i] &lt;-<span class="st"> </span>xt
        counter &lt;-<span class="st"> </span>counter<span class="op">+</span><span class="dv">1</span>
    }
}

<span class="kw">plot</span>(my_vector,<span class="dt">type =</span> <span class="st">&quot;l&quot;</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAV1BMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6kNtmAABmADpmZmZmtrZmtv+QOgCQOjqQZgCQkGaQ2/+2ZgC2/7a2///bkDrb/9vb////tmb/25D//7b//9v///8Q+xe/AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAJNklEQVR4nO3diZbaNhgF4D+ZTIe0Q9PQ0DCg93/O4kWyZUu62gyCufecNmBbsvWNdwsQxQQj916A1kMgEAKBEAiEQCAEAiEQCIFACARCIBACgRAIhEAgBAIhEAiBQAgEQiAQAoEQCIRAIAQCIRAIgUAIBEIgEAKBEAiEQCAEAiEQCIFACARCIBACgRAIhEAgBAIhEAiBQAgEQiAQAoEQCIRAIAQCIRAIgUAIBEIgEAKBEAiEQCAEAiEQCIFACARCIBACgRAIhEAgBAIhEAiBQAgEQiAQAoEQCIRAIAQCIRAIgUAIBEIgEAKBEAiEQCCVgeRhci+gutVtFwKBEAiEQMtI8G10udy5J+/7bp77Am1VXcUQCIRAIAQCIRAIgUBk9n9FoHUIFI4QKJzPChTfTgKBCQkEJmwUaGswAiXOzz/hIFQAdNm/xZaJqS51OSrNzz9hMdB59x5bJqa61OWoND//hMVA6vT1V2yhmOoSl6PS/PwTVliDxnuCRU43B4qdQYU1qEoIlFhdO0CqHOjj9cr85ad/2vOuO9CdQpvhtkDr2m4JdJKu+UfxH8x6oOPL78Ah74mB9HlQD+BOBzTS+Ka6I1DYqhxIrxSBo30H9PH67pjK+1zs9kCeOXIN2hwoah/UrSPflN5dB6tLXI6oZAHpwTc4iqne6DrFYAmqS1uOqEQBuRehBlCNNA4kBUDn78O6U3ZJ1jyQWENT5qmBjvcBck3o2Z1YQ24FdJgO1EW3hWKBItq+GubqXNMNkumlusEaVJYqQK5GSAhI9Eu1IVCdLNsp7lG+Ztivg0DjQQkCiQXkrDuY2YSH68mf/wCeXF0SkHUS42iEhIDkNkCH/uT4vPsWWzRcXQhIhsVeAOghVpNm4xdAusEyAQ0T2W2vBhRxLZZSnbJ2oEq/NEsbByTz8SJzM9G7Hw+QLGsuBTLXYlsA6ZeyAlrsQKzXISAJAI17Jsf6KGYLzNjEhquwj9dah/l+QR1AsgISvcFMLZUVkKj53QLJAJJCoKhrsejqJBXI/G8EEpkRrW6niIZYAE3DXUBSBlQjYaDpr2vtYnXTY4BMhQsgsYHM+tU2kHmUYAPpEWa7WbRKT6GBZB6zP7KAxPxr12cvQyHQeScvvw+VzoNWQLICEjsOIPEATY2fVpntgU5ffh5ffntvhWUDLUjigcQJZDd+AeROHaDuMN/dSK1zmHctejyQNdAFtJQKA5kDfBFQd6LYAQVPFC/7YY7ouZhz0ac9hxNIj1wChZseGDcHmq+bmUB6DTr4b9pfT5XG7c97yeYC8jTGB5QEcEOgcR8Uumk/60IEnmo4geLasiGQlAENDy3Cj54NHnguJvcA8k9bCQim8TXo/kDTQ7OIfVBkC10tTih+Q6CoG0G6l5V3R76+MgCtSgBaHODGxnqqtd9UWYO6+/aFl6pBINereCBlgKyzhjWQ68yo2ibWdf2RCncUo4EWq8R0jRsE0id8Ys/LA6RU1X3QZV/jTNrZfOsPumyTmgOZGzvTZjVphIBmlzgWkOhJZ1f3OUC11iAf0LTIZpi+KrGBVBKQMs5LIM1dA6i/iqi0D5o12pYYWzJvv5quz22gScMC0u2dgDTcBKRWQMqUVrlAhY8z7Or0n9P8LRdNmUlFAKkIIP1er5J6kA00LWLRedDl7+wVScw/Mi2qHjc1ZYKaph3fLN18QHpmk9KstomlcSCVD6T0xDaQWgKtW2GPbRNIQaAlRABoNq0PyLUMc762geZDZmDLMWP5OZBedT49kPIDKRdQuJGrwY8K5ComlpsbSD0JkNoQaF1DcLC0CqTWQEpcs6wK5KieQGCp5kCpZWMnTK0uCWhWKgEofqlqAJ13ZZepy/luA5S5VHXWoKMUdnFVcUDh8iugeaFCIGtGsWWtd8VGzw40EBXcMlsDzcY9PlB3w+xdXfaBh6u+JTA3fhYVPxFQ98RikCnoxxkCUnFA1tuynfOimjKg8674dqKKAULlZfE+tmSw1nX5Zs6DEufVMpB+KFipG/DTAWXsmkPVZbZJpFmg2t/+UhWoMJXWoLrfH/R0QOrjjy2OYqnlWwZ6rbqTzizfLtBlX/HBYX75doG2+Yqu5PLbAEUNAmW3+ZK35PLtAtXeSWeW3wTINaPkCaO+wyy2I3l2GgaKSXRH8uw8NlB8N+D8SPBtvWwC5O9InlXdNhXUnk/ja9Bm2WofFNmRPD+PDRTbkbwgDw60fXUE2rqC2vMhUK0Jw7Wsn4vl1lRjcWrOJ+08aILgeZAr+IMcnxwI31X77EDXM8TwbbVPD7R5dQTauoLa88lYoNDN2Vu1rzgEAiEQCIFA7gb0MNkQqLxSML5sdGnxssnrVEqgsvEEIlDZeAIRqGw8gQhUNv75gZ4pBAIhEAiBQAgEQiAQAoEQCIRAIAQCIRAIgUAIBFIf6JT6hZ79B0DerKLrF+EMX8SfXTyU6kCn7ndYU5bqsr9Ofey+m8cUXb8A8+z7JWcXD6Y20PBU8ZDw6cXh16OPX3+ZousX4fQ/KeMolb4sjtQGMs1NLHf9Q5ui6xfhsseXH1eg7OLhVAfqP5aX/vUoh2t7dNH1CzTLbh+UWxykNtCwySdv+N2HP0zR9YtQ0W5D6n+/Mq84ShtAJ72Pzmih/jmZBwHKWq2HDw/lbSP9NA+0ieXsGMdPeubtZY9jd5b3B9lJZxxa9afQCo7Th8c5zKefnE2/y5l/pnd4nBPFfp1PWqZxG+nKmKLrF+EMlxrZxUPhxSoIgUAIBEIgEAKBEAiEQCAEAiEQCIFACARCIBACgRAIhEAgBAIhEAiBQAgEQiAQAoEQCIRAIAQCIRAIgUAIBNIe0Kq3QWn3g7IQCIRAIE0CnXd/7aTvVtX1X/ln7BnT9UPruouddzV+ASQ2jQJ1ne+v/x2uHicZ33Q9rbpOY76vi98mjQK99V0Wh06Gh3FA3x3z9PXf7zfd4hoFeu9/g2HooGo69PZehxq/uJyQpoGOBkh3ZcXfJls7TQMt1qAul/2fZd16U9M00NjReRzQ5/jyX5WfsYpO00D9wcscxYa99Xtp1/nEtA1knwddYYYe47fcTbcH1FgIBEIgEAKBEAiEQCAEAiEQCIFACARCIBACgRAIhEAgBAIhEAiBQAgEQiAQAoEQCIRAIAQCIRDI/5FOZr6wEt4DAAAAAElFTkSuQmCC" /><!-- --></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">After1kDiscard &lt;-<span class="st"> </span>my_vector[<span class="dv">1000</span><span class="op">:</span>N]</code></pre></div>
<div id="compare-the-deciles-of-samplerafter-discard-the-first-1000-of-the-chain-and-standard-cauchy" class="section level5">
<h5>Compare the deciles of Sampler(after discard the first 1000 of the chain) and standard Cauchy:</h5>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Sequence Generation</span>
a &lt;-<span class="st"> </span><span class="kw">ppoints</span>(<span class="dv">100</span>)
sequence &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="fl">0.01</span>)
standardCauchy &lt;-<span class="st"> </span><span class="kw">qcauchy</span>(sequence)
standardCauchy &lt;-<span class="st"> </span>standardCauchy[(standardCauchy<span class="op">&gt;</span><span class="st"> </span><span class="op">-</span><span class="ot">Inf</span>) <span class="op">&amp;</span><span class="st"> </span>(standardCauchy<span class="op">&lt;</span><span class="st"> </span><span class="ot">Inf</span>)]
Q &lt;-<span class="st"> </span><span class="kw">quantile</span>(my_vector, a)
<span class="kw">qqplot</span>(standardCauchy, Q, <span class="dt">main=</span><span class="st">&quot;&quot;</span>,<span class="dt">xlab=</span><span class="st">&quot;standardCauchy Quantiles&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Sample Quantiles&quot;</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAaVBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6Ojo6OpA6kNtmAABmADpmkJBmtrZmtv+QOgCQkGaQ27aQ2/+2ZgC2Zjq225C2///bkDrbtmbb25Db/7bb////tmb/trb/25D//7b//9v///9mrN9nAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAKG0lEQVR4nO2dCXejNhSFlTrpYqeTTId2aBvXjv//jyxaDV642hGee+fY54x5WvgiPQkkHuJEzUosXYHWRUBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBHQGdNxtT70Qm48Fa9OgzoC6zcfh5fnUPS9YmwblAB13b6e9GD4/fV+yPs1pAqgb4PQENNGoiz0fd5uP445dbKKxkxZP3z7fyWeqzMO8WI2WApQ3uwIybGIADX1s89Ft04pPSl1BYvTxTqC1f/rWSyedRKh1QMJ+hwP6fB8m0sM0Om2Yf2BAch4kAaVNFB8YkG1BXdLFWOuA0n1QP1xtJJbfttJGMTHMFdOKT0pdUZwHAREQEAEBBQJS7sfooYd5J7YgoB8QUMAV+ulHBBQ0+4vwQVvnhtbpg8To298+p+Ei2YUWXBLQ8VXPoVd6sVoP0Fpvd5T1QafuPA9a0Q2zycBVehSzLShNVQEFNpqrtHkNT3r1fnBScyNdTUChbudG4iDDz3c0zCtA8qaauv2YVm4G1QbUQe8sARk0/Z37jg8MyGM5Q5ocXhSgi8lA+HpcDtX1QXd7zVmNtaDQgWuaNNjQY1VeX408n2aaWyVA6U015qa9z+16tcNhML3XHesASulboyzCDFd0sZrknSd55DRcJLu5QggIFFIdkO1jrXcxM5mo7oPkmnP/bOc5xctNKiHDhCtuoriXuzvaXpvP0b2CcphMFA+/fFefCuWmFbAEILm7Q97yaB1QDgcUksPZUN5K7LaNdzFh/mXIKNywe5YjWdo+8sKAZPvJc0X8mPMgiUbkKeQhAVn/swyg9q/FVP/SzShDZrGGiQ/7FASk+eS6JxffxdKeFysHSN+vzHbPMh5QoyurFk+uAuIBNbqyahtQtvxiDROfFysFKPuaQPwolvZQbyFAuTvYw82D8i8qPRagAotuMVfzQm20l1f1FcoNyrQFQL1eyekbnEkXWbUNBbS3z2j0orlRrMiqdiAgt6yqHsusUW5Ali0Acgvz/ebv1rpYmW0RsYDApQbcQ5QfUKF9I7FdbP5itbdr8ncX58sByp1toKFz0oef729VHM0Aqm1/KbXxKH6Yn1k3vN8RRanzKNWAoieKs0PYki0oe77ZDaVc+6rmg0rxKXUtBi/5f3RA1bMrxYeAYM7ZDZfIrtjY+GiAsmZqso4wbC5+UMEGFAOoufhBJflE7g9qK35QY4Caix9UlE9CC2omflBRPvE+qJn4QWX5xI5i7cQPKtvB1j8PEmYltdxaf3bDutkJkXOvy60CshvWzS7nXqk7BQQZthY/qHT7WXsLUrtZCWgmE+G+Syl6mE+M+Z/jnETO3ax3Cwk31Cs/fdpEKMNJifIe+hR3qaGn0EtfatjeVZZPzMVqI/GDVA6l209KC1r2aR9LpjSfKB+kLlOXvVgVo09ZxdwPgnPF4uFxTOupEQGkyDyoeHicSt0rqJBQQEWDm7QNCD83fz88Tni51ynN03J1+ESNYnD4KtmCDJoq/seWF2boFT9ItrAi4XEqdq5xgSGGfhvI74THOQ+A3jW8VY2mAc1uvstf7s10jQN68bxhNtfWUn1QNcV0Md8t9kUA1XPPprhgQw8nbVQGUGWVctLIMghQ5UYzLTvc0NtJ5wJU2+1cFR5mWD2wQPWB60bpOQ2zZ0dAXqYrAlQ9yNvKfNACQd7WNYqtJMhbLsVNFDMFeVuyZfgqbqKYJ8jbor7FVxE+KFeQt2VHJ1/FDPOZgrw9LqBM5RIQyu4xfdDhZavWBHPscn3EUUzND+Xic/PRgDMpFJB6Wl69WSMxyNtqFAZILePom66JcQIvcw44sGCCGVlAb25vAgHdSKMAafeTNlH0qc0aASnPo7Yn+i9ueObsf6BpQPunb8edbDv4FUiBOfsfaBqQnAINfA4vuTrYwwHKLwKKzZmAQM4E9FAiICACAiIgIAICIiAgAgIiICACAiIgIAICIiAgAgIqAqi3C5D+K5FmU4lvgkDziApZlQAk94mop/Dl196rQsedutXrmyDQPKJCTgUAuUU2vaHaZyVyr7dG+iYINI+o0FmlfJCsj1lHwssAe7FVy3GeCQLNIyo0UilAcg1J7933W4nUZ+ydINA8okJWZQANfWBro4H49XlVa/8EgeYRFbIq18U2Hw0BCq+QVbF50FCNprpYaIWs8gIa5hv2cdbBIXr4RJcg0OuGO2nPCl2pQAvS1RhOIWRU3YeM24HmURWyKtHFZAgiVZeAedk+aOYXaB5TIasiPqgbOo6OP+A9szeOwTdBoHlEhax4sQpEQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBERAQAQEREBABAREQEAEBJQf0P5ufJ072wb0z5/vAry863ZAZ/keFLDcPtRIpvUPCD1WdkAz1ZgDpMPkd3Oneivnz3e9UWgu1LNNt2pAJq7T7LsrbuVsNkPNhoVqBZCKif8mv+ULkfV+k+Puy06YlyqJp68SkD3y+lXudbY/25gz/8nzdYntl9yHJkP0/f4iD6gNUP2zPmfTdFQEP2s+LVzV6F/bxeyWNl3bmoBskC9dRRNM5riTJz98ZMyUvfQV5yM2lIr8eRJz5mxizlj/Z2szk3uDzMu+XJDw4Tdnflm49T82K9khXUiyeoDs1khVFfkyN1kH9Rd2mwMHD3N1xPw8epHJyESfsT1ok6igqpqM67R7zUGZXxYxAuRiSPlv5MwGyDw+4Xr6XrZgW+nzjrnJEffz5ZteJomtgxk3EdMlbwG6Sj8CpO0VNF9C+XyQGqfPPf2nv0aNoHeAJkfcz5MudpnY9iNHYODaaaA3u9hl+jEgE+Htzda2JiCpTv8h1d983EtcU5keOTcs66T34u0q8VULOr7+8WqZWSe9+ZiYH+4Amoyjs5MKq7yATHX0PvirXmIc7PmI+3k8zI9MrA+xPsgA+nz/1c4GeuWwN/+8bJ35VeETH/Q2rW1FQPZ5LPsSsuNObKcDkRyupkfcz2ai+PluWCkTSWvoCTrx0AnHw75zWZ0iJNQ+X21+Vbh965eth2w57umxeoDcnL8baqt223Zjx+kmPOMjo+nR6fyeYGuifvoyngfZzMavADFuZeBkza8K7y7mQbqJeW4IXuXF6uG3q/n2nxlebXVTqwTU+75lJ4NWCChnqD6sFQKqKwICIiAgAgIiICACAiIgIAICIiAgAgIiICACAiIgIAICIiAgAgIiICACAiIgIAICIiCg/wG+Xrim8WiMWgAAAABJRU5ErkJggg==" /><!-- --></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(After1kDiscard, <span class="dt">freq =</span> <span class="ot">FALSE</span>,<span class="dt">main =</span> <span class="st">&quot;after discard the first 1000 of the chain&quot;</span>)
<span class="kw">lines</span>(standardCauchy, <span class="kw">dcauchy</span>(standardCauchy), <span class="dt">lty =</span> <span class="dv">2</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAbFBMVEUAAAAAADoAAGYAOpAAZpAAZrY6AAA6ADo6AGY6OgA6Ojo6kLY6kNtmAABmADpmZjpmtv+QOgCQOjqQkGaQtpCQ29uQ2/+2ZgC2Zma225C2/7a2/9u2///bkDrb////tmb/25D//7b//9v///+uEXY3AAAACXBIWXMAAA7DAAAOwwHHb6hkAAALbUlEQVR4nO2da4OjJhSGyWwz224z20x3G3tZ12Ty//9j5Y4KvHhDTM77YUaNcA5P4IBClN1JUbGtHShdBAiIAAERICACBESAgAgQEAECIkBABAiIAAERICACBESAgAgQEAECIkBABAiIAAERICACBESAgAgQEAECIkBABAiIAAERICACBESAgAgQEAECYq2O9/vfP+CZFXv58fHOPv0cbYOn1NvcUM0Ol8jJjP0SNtNxtNKn8USne2TTZ+WY5BIHdLq9veQBJA3FADXcod9DZrqOtqUUp7Veqe85sOm3ckpySXzgfsUJxRwrm1JuxQBFa1fHh+srU4BEmrbM59BmzApyqTVy+I3z5KeZitee//3VzaTddmpQzbRv6ss4G49FGpvcpNT5cEO1OKazd2u7POPwlzBjcjHmKu3oXfkgjrZOHdVf/2YH6clYuSS55AC6vTHtgLV+N3WVGUA1MydKA8JsI7e4DZ3cSdkFZI86RoeAZC7WXBfQy3/y62qzkKX+9NO/aaqNLpUPUMglp4mJvzLfmjlBoBE4KwOoPeUojp7U3/aA+OgoviQnuZOyW59V+U8do7bym+9Bld+Yc5pYfVSnKZviM/+mqT/GqqeJhVyygNShmm/XbtOtNFcFqLVk3FTZ6/1a1k2d3EnZ8+asvnjXqAfQWRasZ05rHKBKV3lRviGggEsWkGihqu65MUs3YxODVMM5WRf1ebrbUMndlF1v+MeiZrhGh4B0LqbPmQFIu1r3yodcsoBUDBkAklXc7eZVI22ZOzFQlOMsT1bJ3ZQBb1yjAUCOuTmAtDOiX4sB6rnUqUGmWfVqkEDfGQfVJgrrGiQd6gByU4a/rm437AFkzfkBpQXpMTXIdakTg0yP2BkW9GPQ/ePPiz4sP2vMiEMHad1FBGKQ8cY1GgI0MNcDpJxSfbtn07UdjEEhl1S3rMYsZ/5x98sb9mKyO5AOiM946zqLo3zL1qBhL1Y75qQjjtEAINdcdzzXqRRwoBjoxZBL4oPGHQfZ820V8o6DDhf72dGOlhxATkrDWo3KtDeO0VANcsw1Tl4WUOKlRm38c8uHXDLFlH65g/cuobMTg0SgV46ayxqe+nCpuzXYpLQ58XOsN47RECDXXOWwdPrQtItVkc/ZVpskl+h2BxABAiJAQAQIiAABESAgAgREgIAIEBABAiJAQAQIiAABESAgAgREgIAIEBABAiJAQAQIiAABESAgAgREgIAIENA4QHp+eepyzh1qFKBaz+Q2ofXHj6cxgD7eDZZ6wnLpfWoMoNubWVnUPE0joxoENDIGqSpEMSggvZ7yaeoPjYOgCgfEBhubeZCk2qxhy9SL7QxQfbi0YUisgswJiHWUxW7Pg0TJbv7jnS+wzArovmFVmjJQrD79JEA+mYFidcwFSIgFd3IbR9JYnB8XqFxWjRD7AWSG0h/vgRq0HiDm7mTUwvaWym7YcREgbz6s+2dBC2M9GalQkF4YELN/dwZo7ewIUFo+zN1Z1sJIT0rLLsykbEBwVuPJAeFZjfUA7SEGJdyTfu4gnTCr8dyAqAYh4VmNJweEZzVoHJQpu5128/my8zBhnk8yiACN8aSc7DwxiAB58iFA8Xx8fRkBcvLxMdkM0OBRR/OyWySfogDJife5y34efRw0m9GagDzD6gzq26snLfJdfOKwUEANk8/+m76AbM0YtDUgfiEqycxYwvrAgG5vsecUj85uiXx6dzi2BvRV8pm3AnpNQMtaGOUJlwY0b2HLw3bzle2Hiu3ml7WQqkENWii7JfMpIwaVlV2BQbqs7LxBunMwl6S929tJ34+f92O5RwWUqrxz82V18ynKPDdfGKC2mXXfhjBQ/plVtJNB1l7FH6V9vFfh22b55+a7O9vGIF76pr2Wj1xqbF2DtgdU8YdlR3qxvHPzzN0x/zdsYsfbG3+ZQOzOdNa5+QGg7tFM6twPOlw+3ufduX9kQEVl5we0pIUxnhSYXYmAFnkux0OPgyAZc7kWxrguoE1jEB9IIwV/BTXMbqaYm1UpgHwvA+wJdnIPDCitg2+871RcZeKwNEChsk/Nbm4+zN1xNjZsYmXdMPMBWtJCuidFZldiNz9KmX5x2N9hmwIS7xyr5q0QemRAzeFS86v5QiYOvTFoS0D8blg9+6kcawdpth0gPlDkgKKLFzLOahQHSNegKnLXPuesRnndvIxBdWS4mPWedHmA5FAxtopq61mNRS2kaoy9rWc1xHbJgPL/4nC4sx0geS0G1rfmm9XoQAl9kEPaXqXf0FrISLo0QI0OztfXWTc9HrWJOXfLInPzydnNV2m9mHMFVsqlRtJOBmlACz0pmgAlZTdfBGhMPt2dzITKBMTKAwTnTMdkN9+rWBPLS6jIm/YsGoP2CGjhicM4oLwqsQb17zwTIJAPAQL59HeyMtojoKyEygPEBvkQoGEWIAZlJFQcIDbMh4L0MAcCBDKIXYvlfcnGOBOrTz2z3v/wTq5YNHLaZ7uH3fZ3SgRUwsSh3cnU348xse7Uczi4bBqxy6lBwamw/QBad+p5ApPyerEVp54ntaoMhMoZB00CtD6iEgB1300zzGc/MWjpgeLsjivDsHrLgeI0Jv4dttbN/Nzd/PxKE95ZpTplGygqrzveLwyob2uRCL54DbLOdf/77K25M1DPJ6/7Hi0zUOzV7KF3JWlMiRcfKD6ect+g250IENBEQFnfs7qpcgzWt9cc7xfjMNkEHlyuncFaaRcy8UiAJj0A5YkA4YvVKSYeB1DCpcYUE48DaOJa2OcBRDUICc9qTDHxQICmXaw+E6AnFAECIkBABAiIAAERICACBESAgAgQEAECIkBABAiIAAGtDej6q7iz1oSfbSXvEASfGBJJmZA6wT7QyoBub+LWI3+4TBPw8Po55nksJU6dYh9oXUCNnACRtyIDz5UBzyWMpISpk+wDrQqoYSdRAPlQosB0dR3zO5oSpk6yD7R2DJIOinYQ+LKrLyz8YsVoSpg6yT5QFkCy+fuDAH8dTFvOQBljKXHqFPtImwNyTvMdTypapG7sAxCu4qFHp6U1jsiD10ptYua97LEg6by8PdRbp4XXSF+/jyAd62al86FvF3XQ8dQp9pGyAIoO1ITfwTCLhnjx1Cn2gfIA4q0p6B9/jXL44Y2xlDh1iv246GIViAABESAgAgREgIAIEBABAiJAQAQIiAABESAgAgREgIAIEBABAiJAQAQIiAABESAgAgREgIAIEBABAiJAQAQIiAABESCgDQBVas1Lzdip8U2ri6W75qV5lfwdMT/R+eF+VBPXKXiVH9D18x9iQQEn4C2xXLprAYmfWF9fRyxe2Teg+uWfV10dfIDU0t0eoFGF3jWgj/ejeCfn9dX8BF+tNLt9/dai0Ut3BSD+NBUFiCcSPEVCsWJMLbFXC2BkcrHQ5dueAfFvt+YFMDWI711fT2rF6l2t6eGAxJMe9Pvd2//8dLGmrGk/4Mn4SXwJldwWC17btM3UtUA+ZQfEyytqhwYk21ILxTQqDagRFUUDqvkZZ7PcTp99+3qRK/HEAbkkr9oxIF5V5PesAcmA0ZbMBCQF6Ivs7XqA1K8vOitbG97GRHK74m4p5QZUM91rG0D6SB/Q4btYvNprYvIpWGe7srXNUYR92Vz3Dki9M/n2duzVoLszyjExSLDpB2mutg3pGiT+X18fpQZp19uoamOQKvUQkCi80827p+oYJJfS6yamVkTvF5BuL20oMgFaFKeSxIRsN89jlR4oqrguEOuusK1WsvIwPejkR3fci9nIWr38ywtU6XGQitpCFhBviv1LDR6zBAA1DhIrfC3f3Y+D9iYCBESAgAgQEAECIkBABAiIAAERICACBESAgAgQEAECIkBABAiIAAERICACBESAgAgQEAECIkBABAiIAAH9D8TMZQgQv62ZAAAAAElFTkSuQmCC" /><!-- --></p>
</div>
</div>
<div id="explain" class="section level4">
<h4>Explain</h4>
<p><font color="#000000"></p>
<p>To see the generated sample as a realization of a stochastic process, we can plot the sample vs the time index. The codes above have displayed a partial plot starting at time index 0. And the final plot shows that at times the candidate point is rejected and the chain does not move at these time points; this corresponds to the short horizontal paths in the graph. The QQ plot is an informal approach to assessing the goodness-of-fit of the generated sample with the target distribution. From the plot, it appears that the sample quantiles are in approximate agreement with the theoretical quantiles.</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="exercise-9.6" class="section level3">
<h3>Exercise 9.6</h3>
<p><font color="#000000"></p>
<p><span class="math inline">\(Rao\)</span> [220, Sec. 5g] presented an example on genetic linkage of 197 animals in four categories (also discussed in [67, 106, 171, 266]). The group sizes are (125,18,20,34). Assume that the probabilities of the corresponding multinomial distribution are <span class="math inline">\((\frac12+\frac\theta4,\frac{1-\theta}4,\frac{1-\theta}4,\frac\theta4).\)</span> Estimate the posterior distribution of <span class="math inline">\(\theta\)</span> given the observed sample, using one of the methods in this chapter.</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-exercise-9.6" class="section level3">
<h3>Answer Exercise 9.6</h3>
<p><font color="#000000"></p>
<p>we need to estimate the parameters <span class="math inline">\(p_1,i=1,2,3,4.\)</span> with the missing datas. And the following codes and the corresponding perform results will show some experiments.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1=<span class="dv">125</span><span class="op">/</span><span class="dv">197</span>
p2=<span class="dv">18</span><span class="op">/</span><span class="dv">197</span>
p3=<span class="dv">20</span><span class="op">/</span><span class="dv">197</span>
p4=<span class="dv">34</span><span class="op">/</span><span class="dv">197</span>

my_prob &lt;-<span class="st"> </span><span class="kw">c</span>(p1,p2,p3,p4)
number_of_experiments &lt;-<span class="st"> </span><span class="dv">10</span>
number_of_samples &lt;-<span class="st"> </span><span class="dv">197</span>

experiments &lt;-<span class="st"> </span><span class="kw">rmultinom</span>(<span class="dt">n=</span>number_of_experiments, <span class="dt">size=</span>number_of_samples, <span class="dt">prob=</span>my_prob)
experiments</code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]  121  118  132  123  128  120  134  120  134   132
## [2,]   20   16   19   17   14   18   15   23   14    13
## [3,]   23   24   16   13   22   22   16   23   17    21
## [4,]   33   39   30   44   33   37   32   31   32    31</code></pre>
<p>We hope to find the likelihood function to use the specify method to get the goal.Then get the full data to estimate the <span class="math inline">\(\theta\)</span> to produce the maximum function.And we interal continually to optimize the estimate value.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># defube expectation function </span>

estep &lt;-<span class="st"> </span><span class="cf">function</span>(theta,z2){
z2 =<span class="st"> </span><span class="dv">125</span><span class="op">*</span>(<span class="fl">0.25</span><span class="op">*</span>theta<span class="op">/</span>(<span class="fl">0.5</span><span class="op">+</span><span class="fl">0.25</span><span class="op">*</span>theta))
<span class="co"># z1 = 125*(0.5/(0.5+0.25*theta))</span>
<span class="kw">return</span>(z2)
}

mstep &lt;-<span class="st"> </span><span class="cf">function</span>(theta,z2){
theta &lt;-<span class="st"> </span>(z2<span class="op">+</span><span class="dv">34</span>)<span class="op">/</span>(z2<span class="op">+</span><span class="dv">34</span><span class="op">+</span><span class="dv">18</span><span class="op">+</span><span class="dv">20</span>)
<span class="kw">return</span>(theta)
}

<span class="co"># set initial value for iteration</span>

cur_theta =<span class="st"> </span><span class="fl">0.5</span>
maxit =<span class="st"> </span><span class="dv">100</span>
maxit=<span class="dv">1000</span>
tol=<span class="fl">1e-6</span>

<span class="co"># start iteration</span>

<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>maxit){
  new_z2 &lt;-<span class="st"> </span><span class="kw">estep</span>(cur_theta,cur_z2)
  new_theta &lt;-<span class="st"> </span><span class="kw">mstep</span>(cur_theta,new_z2)
  
  <span class="co"># Stop iteration if the difference between the current and new estimates is less than a tolerance level</span>
  <span class="cf">if</span>( <span class="kw">all</span>(<span class="kw">abs</span>(cur_theta <span class="op">-</span><span class="st"> </span>new_theta) <span class="op">&lt;</span><span class="st"> </span>tol) ){ flag &lt;-<span class="st"> </span><span class="dv">1</span>; <span class="cf">break</span>}
  
  
  <span class="co"># Otherwise continue iteration</span>
  cur_theta &lt;-<span class="st"> </span>new_theta
  cur_z2 &lt;-<span class="st"> </span>new_z2
}
<span class="cf">if</span>(<span class="op">!</span>flag) <span class="kw">warning</span>(<span class="st">&quot;Didn??t converge</span><span class="ch">\n</span><span class="st">&quot;</span>)

final_theta =<span class="st"> </span>cur_theta
<span class="kw">paste0</span>(<span class="st">&quot;Final Theta = &quot;</span>, <span class="kw">format</span>(<span class="kw">round</span>(cur_theta, <span class="dv">4</span>), <span class="dt">nsmall =</span> <span class="dv">4</span>))</code></pre></div>
<pre><code>## [1] &quot;Final Theta = 0.6268&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">paste0</span>(<span class="st">&quot;Final Z2 = &quot;</span>, <span class="kw">format</span>(<span class="kw">round</span>(cur_z2, <span class="dv">4</span>), <span class="dt">nsmall =</span> <span class="dv">4</span>))</code></pre></div>
<pre><code>## [1] &quot;Final Z2 = 29.8277&quot;</code></pre>
<p>From the whole process ,we has get the disdribution of the parameter <span class="math inline">\(\theta.\)</span> Then we estimate the density function of the observed data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1=<span class="dv">125</span><span class="op">/</span><span class="dv">197</span>
p2=<span class="dv">18</span><span class="op">/</span><span class="dv">197</span>
p3=<span class="dv">20</span><span class="op">/</span><span class="dv">197</span>
p4=<span class="dv">34</span><span class="op">/</span><span class="dv">197</span>

my_prob &lt;-<span class="st"> </span><span class="kw">c</span>(p1,p2,p3,p4)


number_of_experiments &lt;-<span class="st"> </span><span class="dv">1</span>
number_of_samples &lt;-<span class="st"> </span><span class="dv">1000</span>

experiments &lt;-<span class="st"> </span><span class="kw">rmultinom</span>(<span class="dt">n=</span>number_of_experiments, <span class="dt">size=</span>number_of_samples, <span class="dt">prob=</span>my_prob)
experiments</code></pre></div>
<pre><code>##      [,1]
## [1,]  622
## [2,]   89
## [3,]  109
## [4,]  180</code></pre>
<p>But the estimate of the density distribution based on the completed data is like that we will do.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p1=<span class="fl">0.5</span>
p2=<span class="fl">0.25</span><span class="op">*</span>final_theta
p3=<span class="fl">0.25</span><span class="op">-</span>p2
p4=p3
p5=p2

my_prob &lt;-<span class="st"> </span><span class="kw">c</span>(p1,p2,p3,p4,p5)
number_of_experiments &lt;-<span class="st"> </span><span class="dv">10</span>
number_of_samples &lt;-<span class="st"> </span><span class="dv">1000</span>

experiments &lt;-<span class="st"> </span><span class="kw">rmultinom</span>(<span class="dt">n=</span>number_of_experiments, <span class="dt">size=</span>number_of_samples, <span class="dt">prob=</span>my_prob)
experiments</code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]  486  510  511  507  493  510  501  485  505   530
## [2,]  154  175  144  163  170  170  128  166  143   164
## [3,]   85   99   96  105  105   95   93  100  103    96
## [4,]  100   82   94   97   91   89  104   91  108    75
## [5,]  175  134  155  128  141  136  174  158  141   135</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">number_of_experiments &lt;-<span class="st"> </span><span class="dv">1</span>
number_of_samples &lt;-<span class="st"> </span><span class="dv">1000</span>

experiments &lt;-<span class="st"> </span><span class="kw">rmultinom</span>(<span class="dt">n=</span>number_of_experiments, <span class="dt">size=</span>number_of_samples, <span class="dt">prob=</span>my_prob)
experiments</code></pre></div>
<pre><code>##      [,1]
## [1,]  493
## [2,]  160
## [3,]   92
## [4,]   93
## [5,]  162</code></pre>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="question-9-3" class="section level2">
<h2>Question 9-3</h2>
<p><font color="#000000"></p>
<p>For <span class="math inline">\(Exercise 9.6\)</span>, use the <span class="math inline">\(Gelman-Rubin\)</span> method to monitor convergence of the chain, and run the chain until the chain has converged approximately to the target distribution according to <span class="math inline">\(\hat{R}&lt;1.2.\)</span></p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-question-9-3" class="section level2">
<h2>Answer Question 9-3</h2>
<p><font color="#000000"></p>
<div id="ideas-6" class="section level4">
<h4>Ideas</h4>
<ul>
<li><p>The <span class="math inline">\(Gelman-Rubin\)</span> method of monitoring convergence of a <span class="math inline">\(M-H\)</span> chain is based on comparing the behavior of several generated chains with respect to the variance of one or more scalar summary statistics. The estimates of the variance of the statistic are analogous to estimates based on between-sample and within-sample mean squared errors in a one-way analysis of variance (<span class="math inline">\(ANOVA\)</span>).</p></li>
<li><p>The between-sequence variance is <span class="math display">\[B=\frac1{k-1}\sum_{i=1}^{k}\sum_{j=1}^{n}(\hat{\psi_{i.}}-\hat{\psi_{..}})^2=\frac{n}{k-1}\sum_{i=1}^{k}(\hat{\psi_{i.}}-\hat{\psi_{..}})^2\]</span></p></li>
</ul>
<p>where <span class="math inline">\(\hat{\psi_{i.}}=\frac1n\sum_{j=1}^{n}\psi_{ij} , \hat{\psi_{..}}=\frac1{nk}\sum_{i=1}^{k}\sum_{j=1}^{n}\psi_{ij}\)</span></p>
<ul>
<li><p>Within the <span class="math inline">\(i^{th}\)</span> sequence, the sample variance is <span class="math display">\[s^2_i=\frac1n\sum_{j=1}^{n}(\psi_{ij}-\hat{\psi{i.}})^2\]</span> and the pooled estimate of within sample variance is <span class="math display">\[W=\frac1{nk-k}\sum_{i=1}^{k}(n-1)s^2_i=\frac1k\sum_{i=1}^{k}s^2_i\]</span></p></li>
<li><p>The between-sequence and within-sequence estimates of variance are combined to estimate an upper bound for <span class="math inline">\(V ar(\psi)\)</span> is <span class="math display">\[\hat{Var(\psi)}=\frac{n-1}{n}W+\frac1nB\]</span></p></li>
<li><p>The <span class="math inline">\(Gelman-Rubin\)</span> statistic is the estimated potential scale reduction <span class="math display">\[\sqrt{\hat{R}}=\sqrt{\frac{\hat{Var(\psi)}}{W}}\]</span></p></li>
</ul>
<p>which can be interpreted as measuring the factor by which the standard deviation of <span class="math inline">\(\psi\)</span> could be reduced by extending the chain. The factor <span class="math inline">\(\hat{R}\)</span> decreasesto 1 as the length of the chain tends to infinity, so <span class="math inline">\(\hat{R}\)</span> should be close to 1.ifthe chains have approximately converged to the target distribution. <span class="math inline">\(Gelman\)</span> suggests that <span class="math inline">\(\hat{R}\)</span> should be less than 1.1 or 1.2.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
Gelman.Rubin &lt;-<span class="st"> </span><span class="cf">function</span>(psi) {
  <span class="co"># psi[i,j] is the statistic psi(X[i,1:j])</span>
  <span class="co"># for chain in i-th row of X</span>
  psi &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(psi)
  n &lt;-<span class="st"> </span><span class="kw">ncol</span>(psi)
  k &lt;-<span class="st"> </span><span class="kw">nrow</span>(psi)

  psi.means &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(psi) <span class="co">#row means</span>
  B &lt;-<span class="st"> </span>n <span class="op">*</span><span class="st"> </span><span class="kw">var</span>(psi.means) <span class="co">#between variance est.</span>
  psi.w &lt;-<span class="st"> </span><span class="kw">apply</span>(psi, <span class="dv">1</span>, <span class="st">&quot;var&quot;</span>) <span class="co">#within variances</span>
  W &lt;-<span class="st"> </span><span class="kw">mean</span>(psi.w) <span class="co">#within est.</span>
  v.hat &lt;-<span class="st"> </span>W<span class="op">*</span>(n<span class="op">-</span><span class="dv">1</span>)<span class="op">/</span>n <span class="op">+</span><span class="st"> </span>(B<span class="op">/</span>n) <span class="co">#upper variance est.</span>
  r.hat &lt;-<span class="st"> </span>v.hat <span class="op">/</span><span class="st"> </span>W <span class="co">#G-R statistic</span>
  <span class="kw">return</span>(r.hat)
}

theta &lt;-<span class="st"> </span>.<span class="dv">2</span> <span class="co">#parameter of proposal distribution</span>
k &lt;-<span class="st"> </span><span class="dv">4</span> <span class="co">#number of chains to generate</span>
N &lt;-<span class="st"> </span><span class="dv">5000</span> <span class="co">#length of chains</span>
b &lt;-<span class="st"> </span><span class="dv">500</span> <span class="co">#burn-in length</span>
M &lt;-<span class="st"> </span><span class="dv">5000</span>
groupsize &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">125</span>,<span class="dv">18</span>,<span class="dv">20</span>,<span class="dv">34</span>) <span class="co">#group size</span>

prob &lt;-<span class="st"> </span><span class="cf">function</span>(theta,group){
  <span class="cf">if</span>(theta<span class="op">&lt;</span><span class="dv">0</span><span class="op">||</span>theta<span class="op">&gt;=</span><span class="fl">0.9</span>)
    <span class="kw">return</span>(<span class="dv">0</span>)
  <span class="kw">return</span>((<span class="dv">1</span><span class="op">/</span><span class="dv">2</span><span class="op">+</span>theta<span class="op">/</span><span class="dv">4</span>)<span class="op">^</span>groupsize[<span class="dv">1</span>]<span class="op">*</span>((<span class="dv">1</span><span class="op">-</span>theta)<span class="op">/</span><span class="dv">4</span>)<span class="op">^</span>groupsize[<span class="dv">2</span>]<span class="op">*</span>((<span class="dv">1</span><span class="op">-</span>theta)<span class="op">/</span><span class="dv">4</span>)<span class="op">^</span>groupsize[<span class="dv">3</span>]<span class="op">*</span>(theta<span class="op">/</span><span class="dv">4</span>)<span class="op">^</span>groupsize[<span class="dv">4</span>])
}


mn.chain &lt;-<span class="st"> </span><span class="cf">function</span>(groupsize,N,X1) {
  <span class="co">#generates a Metropolis chain </span>
  <span class="co">#with Multinomial proposal distribution</span>
  <span class="co">#and starting value X1</span>
  x &lt;-<span class="st"> </span><span class="kw">numeric</span>(N)
  x[<span class="dv">1</span>] &lt;-<span class="st"> </span>X1<span class="co">#initialize x[1]</span>
  w &lt;-<span class="st"> </span><span class="fl">0.25</span>
  u &lt;-<span class="st"> </span><span class="kw">runif</span>(N)
  v &lt;-<span class="st"> </span><span class="kw">runif</span>(M, <span class="op">-</span>w, w)

  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span>N) {
    theta &lt;-<span class="st"> </span>x[i <span class="op">-</span><span class="st"> </span><span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>v[i]
    <span class="cf">if</span> (u[i] <span class="op">&lt;=</span><span class="st"> </span><span class="kw">prob</span>(theta, groupsize) <span class="op">/</span><span class="st"> </span><span class="kw">prob</span>(x[i <span class="op">-</span><span class="st"> </span><span class="dv">1</span>], groupsize))
    x[i] &lt;-<span class="st"> </span>theta
    <span class="cf">else</span>
    x[i] &lt;-<span class="st"> </span>x[i <span class="op">-</span><span class="st"> </span><span class="dv">1</span>]
  }
  <span class="kw">return</span>(x)
}

<span class="co">#choose overdispersed initial values</span>
x0 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="fl">0.2</span>, <span class="fl">0.4</span>, <span class="fl">0.6</span>, <span class="fl">0.8</span>)

<span class="co">#generate the chains</span>
X &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow=</span>k, <span class="dt">ncol=</span>N)
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>k)
  X[i, ] &lt;-<span class="st"> </span><span class="kw">mn.chain</span>(groupsize,N,x0[i])

<span class="co">#compute diagnostic statistics</span>
psi &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">apply</span>(X, <span class="dv">1</span>, cumsum))
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(psi))
  psi[i,] &lt;-<span class="st"> </span>psi[i,] <span class="op">/</span><span class="st"> </span>(<span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(psi))
<span class="kw">print</span>(<span class="kw">Gelman.Rubin</span>(psi))</code></pre></div>
<pre><code>## [1] 1.138404</code></pre>
<p>Then we can learn from the results:</p>
<ul>
<li><p>The plots above are the four sequences of the summary statistic (the mean) <span class="math inline">\(\psi\)</span> showing from time 0 to 5000.</p></li>
<li><p>The dashed line on the plot is at <span class="math inline">\(\hat{R}=1.2\)</span>,and we know that the according interations appropriately within 1000.</p></li>
</ul>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="exercise-11.4" class="section level2">
<h2>Exercise 11.4</h2>
<p><font color="#000000"></p>
<p>Find the intersection points <span class="math inline">\(A(k)\)</span> in <span class="math inline">\((0,\sqrt{k})\)</span> of the curves <span class="math display">\[S_{k-1}(a)=P(t(k-1)&gt;\sqrt{\frac{a^2(k-1)}{k-a^2}})\]</span> and <span class="math display">\[S_{k}(a)=P(t(k)&gt;\sqrt{\frac{a^2k}{k+1-a^2}}),\]</span> for k = 4 : 25, 100, 500, 1000, where <span class="math inline">\(t(k)\)</span> is a Student <span class="math inline">\(t\)</span> random variable with <span class="math inline">\(k\)</span> degrees of freedom. (These intersection points determine the critical values for a t-test for scale-mixture errors proposed by <span class="math inline">\(Sz??ekely [260].\)</span>)</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-exercise-11.4" class="section level2">
<h2>Answer exercise 11.4</h2>
<p><font color="#000000"></p>
<p>The Idea is :</p>
<ul>
<li><p>generate the inset function that the inside one is the function of a and the ourside one is the function of k.</p></li>
<li><p>first, we code the function to compute the <span class="math inline">\(S_{k-1}(a)\)</span> and <span class="math inline">\(S_k(a)\)</span> respectively.</p></li>
<li><p>then construct the function of <span class="math inline">\(S_k(a)-S_{k-1}(a)\)</span></p></li>
<li><p>and the next step is to find the intersection points through the function <span class="math inline">\(uniroot\)</span></p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
findIntersection =<span class="st"> </span><span class="cf">function</span> (k) {
  s.k.minus.one =<span class="st"> </span><span class="cf">function</span> (a) {
    <span class="dv">1</span><span class="op">-</span><span class="kw">pt</span>(<span class="kw">sqrt</span>(a<span class="op">^</span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>(k <span class="op">-</span><span class="st"> </span><span class="dv">1</span>) <span class="op">/</span><span class="st"> </span>(k <span class="op">-</span><span class="st"> </span>a<span class="op">^</span><span class="dv">2</span>)), <span class="dt">df =</span> k<span class="op">-</span><span class="dv">1</span>)
  }
  
  s.k =<span class="st"> </span><span class="cf">function</span> (a) {
    <span class="dv">1</span><span class="op">-</span><span class="kw">pt</span>(<span class="kw">sqrt</span>(a<span class="op">^</span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>k <span class="op">/</span><span class="st"> </span>(k <span class="op">+</span><span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>a<span class="op">^</span><span class="dv">2</span>)), <span class="dt">df =</span> k)
  }
  
  f =<span class="st"> </span><span class="cf">function</span> (a) {
    <span class="kw">s.k</span>(a) <span class="op">-</span><span class="st"> </span><span class="kw">s.k.minus.one</span>(a)
  }
  
  eps =<span class="st"> </span><span class="fl">1e-2</span>
  <span class="kw">return</span>(<span class="kw">uniroot</span>(f, <span class="dt">interval =</span> <span class="kw">c</span>(eps, <span class="kw">sqrt</span>(k)<span class="op">-</span>eps))<span class="op">$</span>root)
}

rs =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">c</span>(<span class="dv">4</span><span class="op">:</span><span class="dv">25</span>, <span class="dv">100</span>, <span class="dv">500</span>, <span class="dv">1000</span>), <span class="cf">function</span> (k) {
  <span class="kw">findIntersection</span>(k)
  })
<span class="kw">cbind</span>(<span class="dt">k=</span><span class="kw">c</span>(<span class="dv">4</span><span class="op">:</span><span class="dv">25</span>, <span class="dv">100</span>, <span class="dv">500</span>, <span class="dv">1000</span>),<span class="dt">a=</span>rs)</code></pre></div>
<pre><code>##          k         a
##  [1,]    4  1.492073
##  [2,]    5  1.533555
##  [3,]    6  1.562745
##  [4,]    7  1.584429
##  [5,]    8  1.601184
##  [6,]    9  1.614516
##  [7,]   10  1.625389
##  [8,]   11  1.634418
##  [9,]   12  1.642026
## [10,]   13  1.648552
## [11,]   14  1.654175
## [12,]   15  1.659099
## [13,]   16  1.663454
## [14,]   17  1.667287
## [15,]   18  1.670733
## [16,]   19  1.673832
## [17,]   20  1.676605
## [18,]   21  1.679148
## [19,]   22  1.681468
## [20,]   23  4.785832
## [21,]   24  4.888979
## [22,]   25  4.990000
## [23,]  100  9.990000
## [24,]  500 22.350680
## [25,] 1000 31.612777</code></pre>
<p>From the results ,we can get the different intersection points with the different k.And we find the intersection point changes slightly while the k has a significant increase.</p>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="question-11.6" class="section level2">
<h2>Question 11.6</h2>
<p><font color="#000000"></p>
<p>Write a function to compute the <span class="math inline">\(cdf\)</span> of the <span class="math inline">\(Cauchy\)</span> distribution, which has density <span class="math display">\[\frac1{\theta\pi(1+((x-\eta)/\theta)^2)},-\infty&lt;x&lt;\infty,\]</span> where <span class="math inline">\(\theta&gt;0.\)</span> Compare your results to the results from the <span class="math inline">\(R\)</span> function <span class="math inline">\(pcauchy.\)</span> (Also see the source code in <span class="math inline">\(pcauchy.c.\)</span>)</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-question-11.6" class="section level2">
<h2>Answer Question 11.6</h2>
<p><font color="#000000"></p>
<div id="the-idea-is" class="section level3">
<h3>The Idea is :</h3>
<ul>
<li><p>If we knoow the density function of the specified distribution,then integtate this density function with the suitable interval then we will get the target <span class="math inline">\(CDF\)</span></p></li>
<li><p>To get the truth <span class="math inline">\(CDF\)</span>, we utilize the function <span class="math inline">\(sapply\)</span> and <span class="math inline">\(pcauchy\)</span></p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my.dcauchy =<span class="st"> </span><span class="cf">function</span> (x, eta, theta) {
  <span class="kw">stopifnot</span>(theta <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)
  <span class="kw">return</span>(<span class="dv">1</span><span class="op">/</span>(theta<span class="op">*</span>pi<span class="op">*</span>(<span class="dv">1</span> <span class="op">+</span><span class="st"> </span>((x <span class="op">-</span><span class="st"> </span>eta)<span class="op">/</span>theta)<span class="op">^</span><span class="dv">2</span>)))
}

my.pcauchy =<span class="st"> </span><span class="cf">function</span> (x, eta, theta) {
  <span class="kw">stopifnot</span>(theta <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)
  
  integral =<span class="st"> </span><span class="cf">function</span> (x) {
    <span class="kw">my.dcauchy</span>(x, eta, theta)
  }
  
  <span class="kw">return</span>(<span class="kw">integrate</span>(integral, <span class="dt">lower =</span> <span class="op">-</span><span class="ot">Inf</span>, <span class="dt">upper =</span> x)<span class="op">$</span>value)
}

eta =<span class="st"> </span><span class="dv">0</span>
theta =<span class="st"> </span><span class="dv">2</span>
xs =<span class="st"> </span><span class="kw">seq</span>(<span class="op">-</span><span class="dv">10</span>, <span class="dv">10</span>)
estimate =<span class="st"> </span><span class="kw">sapply</span>(xs, <span class="cf">function</span>(x) <span class="kw">my.pcauchy</span>(x, eta, theta))
truth =<span class="st"> </span><span class="kw">sapply</span>(xs, <span class="cf">function</span>(x) <span class="kw">pcauchy</span>(x, eta, theta))
<span class="kw">round</span>(<span class="kw">cbind</span>(estimate, truth),<span class="dv">4</span>)</code></pre></div>
<pre><code>##       estimate  truth
##  [1,]   0.0628 0.0628
##  [2,]   0.0696 0.0696
##  [3,]   0.0780 0.0780
##  [4,]   0.0886 0.0886
##  [5,]   0.1024 0.1024
##  [6,]   0.1211 0.1211
##  [7,]   0.1476 0.1476
##  [8,]   0.1872 0.1872
##  [9,]   0.2500 0.2500
## [10,]   0.3524 0.3524
## [11,]   0.5000 0.5000
## [12,]   0.6476 0.6476
## [13,]   0.7500 0.7500
## [14,]   0.8128 0.8128
## [15,]   0.8524 0.8524
## [16,]   0.8789 0.8789
## [17,]   0.8976 0.8976
## [18,]   0.9114 0.9114
## [19,]   0.9220 0.9220
## [20,]   0.9304 0.9304
## [21,]   0.9372 0.9372</code></pre>
</div>
<div id="interpretation" class="section level3">
<h3>Interpretation:</h3>
<ul>
<li>From the result,we could agree intuitively that the value we estimate is equal to its real value to some extent.</li>
</ul>
<p><font color="#000000"></p>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="exercise-11-abo" class="section level2">
<h2>Exercise 11-ABO</h2>
<p><font color="#000000"></p>
<ul>
<li><span class="math inline">\(A-B-O\)</span> blood type problem</li>
</ul>
<table>
<thead>
<tr class="header">
<th align="center">1</th>
<th align="center">2</th>
<th align="center">3</th>
<th align="center">4</th>
<th align="center">5</th>
<th align="center">6</th>
<th align="center">7</th>
<th align="center">8</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Genotype</td>
<td align="center">AA</td>
<td align="center">BB</td>
<td align="center">OO</td>
<td align="center">AO</td>
<td align="center">BO</td>
<td align="center">AB</td>
<td align="center">AA</td>
</tr>
<tr class="even">
<td align="center">Frequency</td>
<td align="center">p2</td>
<td align="center">q2</td>
<td align="center">r2</td>
<td align="center">2pr</td>
<td align="center">2qr</td>
<td align="center">2pq</td>
<td align="center">1</td>
</tr>
<tr class="odd">
<td align="center">Count</td>
<td align="center">nAA</td>
<td align="center">nBB</td>
<td align="center">nOO</td>
<td align="center">nAO</td>
<td align="center">nBO</td>
<td align="center">nAB</td>
<td align="center">n</td>
</tr>
</tbody>
</table>
<ul>
<li><p>Observed data: <span class="math inline">\(n_{A??} = n_{AA} + n_{AO} = 28\)</span> (A-type), <span class="math inline">\(n_{B??} = n_{BB} + n_{BO} = 24\)</span> (B-type), <span class="math inline">\(n_{OO} = 41\)</span> (O-type), <span class="math inline">\(n_{AB} = 70\)</span> (AB-type).</p></li>
<li><p>Use EM algorithm to solve <span class="math inline">\(MLE\)</span> of <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> (consider missing data <span class="math inline">\(n_{AA}\)</span> and <span class="math inline">\(n_{BB}\)</span>).</p></li>
<li><p>Record the maximum likelihood values in <span class="math inline">\(M\)</span>-steps, are they increasing?</p></li>
</ul>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-exercise-11-abo" class="section level2">
<h2>Answer exercise 11-ABO</h2>
<p><font color="#000000"></p>
<div id="the-idea-is-1" class="section level3">
<h3>The Idea is :</h3>
<ul>
<li><p>The <span class="math inline">\(EM\)</span> (Expectation Maximization) algorithm is a general optimization method that is often applied to find maximum likelihood estimates when data are incomplete.</p></li>
<li><p>First, we compute the likelihood function,then evaluate the log-likelihood of the data for any arbitrary set of allele frequencies.</p></li>
<li><p>Second, we can estimate the number of individuals carrying each possible genotype based on the observed counts for each phenotype and estimates of the allele frequencies at each iteration.And you can run the E-M function with the debug flag set to TRUE.</p></li>
<li><p>Then we set the initial value to be p=0.3,q=0.2,r=0.5, and then run the E-M function.</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#set the likelihood function </span>
lnL &lt;-<span class="st"> </span><span class="cf">function</span>(p, q, <span class="dt">nA =</span><span class="dv">28</span>, <span class="dt">nB =</span> <span class="dv">24</span>, <span class="dt">nAB =</span> <span class="dv">70</span>, <span class="dt">nO =</span> <span class="dv">41</span>) {
  r =<span class="st"> </span><span class="fl">1.0</span> <span class="op">-</span><span class="st"> </span>p <span class="op">-</span><span class="st"> </span>q
  nA <span class="op">*</span><span class="st"> </span><span class="kw">log</span>(p<span class="op">^</span><span class="dv">2</span> <span class="op">+</span><span class="st"> </span><span class="dv">2</span><span class="op">*</span>p<span class="op">*</span>r) <span class="op">+</span><span class="st"> </span>nB <span class="op">*</span><span class="st"> </span><span class="kw">log</span>(p<span class="op">^</span><span class="dv">2</span> <span class="op">+</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>q <span class="op">*</span><span class="st"> </span>r) <span class="op">+</span><span class="st"> </span>nAB <span class="op">*</span><span class="st"> </span><span class="kw">log</span>(<span class="dv">2</span> <span class="op">*</span><span class="st"> </span>p <span class="op">*</span><span class="st"> </span>q) <span class="op">+</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>nO <span class="op">*</span><span class="st"> </span><span class="kw">log</span>(r)
}

EM &lt;-<span class="st"> </span><span class="cf">function</span> (p,q,<span class="dt">nA =</span><span class="dv">28</span>, <span class="dt">nB =</span> <span class="dv">24</span>, <span class="dt">nAB =</span> <span class="dv">70</span>, <span class="dt">nO =</span> <span class="dv">41</span>, <span class="dt">debug =</span> <span class="ot">FALSE</span>) {
  
  <span class="co"># Evaluate the likelihood using initial estimates</span>
  llk &lt;-<span class="st"> </span><span class="kw">lnL</span>(p, q, nA, nB, nAB, nO)
  
  <span class="co"># Count the number of iterations so far</span>
  iter &lt;-<span class="st"> </span><span class="dv">1</span>
  
  <span class="co"># Loop until convergence ...</span>
  <span class="cf">while</span> (<span class="ot">TRUE</span>)
  {
    <span class="co"># Estimate the frequency for allele O</span>
    r=<span class="st"> </span><span class="fl">1.0</span> <span class="op">-</span><span class="st"> </span>p <span class="op">-</span><span class="st"> </span>q
    
    <span class="co"># First we carry out the E-step</span>
    
    <span class="co"># The counts for genotypes O/O and A/B are effectively observed</span>
    <span class="co"># Estimate the counts for the other genotypes</span>
    nAA &lt;-<span class="st"> </span>nA <span class="op">*</span><span class="st"> </span>p <span class="op">/</span><span class="st"> </span>(p <span class="op">+</span><span class="st"> </span><span class="dv">2</span><span class="op">*</span>r)
    nAO &lt;-<span class="st"> </span>nA <span class="op">-</span><span class="st"> </span>nAA
    nBB &lt;-<span class="st"> </span>nB <span class="op">*</span><span class="st"> </span>q <span class="op">/</span><span class="st"> </span>(q <span class="op">+</span><span class="st"> </span><span class="dv">2</span><span class="op">*</span>r)
    nBO &lt;-<span class="st"> </span>nB <span class="op">-</span><span class="st"> </span>nBB
    
    <span class="co"># Print debugging information</span>
    <span class="cf">if</span> (debug)
    {
      <span class="kw">cat</span>(<span class="st">&quot;Round #&quot;</span>, iter, <span class="st">&quot;lnLikelihood = &quot;</span>, llk, <span class="st">&quot;</span><span class="ch">\n\n</span><span class="st">&quot;</span>)
      <span class="kw">cat</span>(<span class="st">&quot;          Allele frequencies:      p = &quot;</span>,   p, <span class="st">&quot;,   q = &quot;</span>,   q, <span class="st">&quot;,   r = &quot;</span>,   r, <span class="st">&quot;</span><span class="ch">\n\n</span><span class="st">&quot;</span>)
      <span class="kw">cat</span>(<span class="st">&quot;          Genotype  counts:      nAA = &quot;</span>, nAA, <span class="st">&quot;, nAO = &quot;</span>, nAO, <span class="st">&quot;, nBB = &quot;</span>, nBB, <span class="st">&quot;, nBO = &quot;</span>, nBO, <span class="st">&quot;</span><span class="ch">\n\n</span><span class="st">&quot;</span>)
    }
    
    <span class="co"># Then the M-step</span>
    p &lt;-<span class="st"> </span>(<span class="dv">2</span> <span class="op">*</span><span class="st"> </span>nAA <span class="op">+</span><span class="st"> </span>nAO <span class="op">+</span><span class="st"> </span>nAB) <span class="op">/</span><span class="st"> </span>(<span class="dv">2</span> <span class="op">*</span><span class="st"> </span>(nA <span class="op">+</span><span class="st"> </span>nB <span class="op">+</span><span class="st"> </span>nO <span class="op">+</span><span class="st"> </span>nAB))
    q &lt;-<span class="st"> </span>(<span class="dv">2</span> <span class="op">*</span><span class="st"> </span>nBB <span class="op">+</span><span class="st"> </span>nBO <span class="op">+</span><span class="st"> </span>nAB) <span class="op">/</span><span class="st"> </span>(<span class="dv">2</span> <span class="op">*</span><span class="st"> </span>(nA <span class="op">+</span><span class="st"> </span>nB <span class="op">+</span><span class="st"> </span>nO <span class="op">+</span><span class="st"> </span>nAB))
    
    
    <span class="co"># Then check for convergence ...</span>
    llk1 &lt;-<span class="st"> </span><span class="kw">lnL</span>(p, q, nA, nB, nAB, nO)
    <span class="cf">if</span> (<span class="kw">abs</span>(llk1 <span class="op">-</span><span class="st"> </span>llk) <span class="op">&lt;</span><span class="st"> </span>(<span class="kw">abs</span>(llk) <span class="op">+</span><span class="st"> </span><span class="kw">abs</span>(llk1)) <span class="op">*</span><span class="st"> </span><span class="fl">1e-6</span>) <span class="cf">break</span>       
    
    <span class="co"># Otherwise keep going</span>
    llk &lt;-<span class="st"> </span>llk1
    iter &lt;-<span class="st"> </span>iter <span class="op">+</span><span class="st"> </span><span class="dv">1</span>
  }
  
 <span class="kw">cbind</span>(<span class="dt">p =</span> p, <span class="dt">q =</span> q)
}
<span class="kw">EM</span>(<span class="fl">0.3</span>,<span class="fl">0.2</span>,<span class="dt">nA =</span><span class="dv">28</span>, <span class="dt">nB =</span> <span class="dv">24</span>, <span class="dt">nAB =</span> <span class="dv">70</span>, <span class="dt">nO =</span> <span class="dv">41</span>, <span class="dt">debug =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## Round # 1 lnLikelihood =  -261.3305 
## 
##           Allele frequencies:      p =  0.3 ,   q =  0.2 ,   r =  0.5 
## 
##           Genotype  counts:      nAA =  6.461538 , nAO =  21.53846 , nBB =  4 , nBO =  20 
## 
## Round # 2 lnLikelihood =  -251.1735 
## 
##           Allele frequencies:      p =  0.3204342 ,   q =  0.3006135 ,   r =  0.3789523 
## 
##           Genotype  counts:      nAA =  8.32035 , nAO =  19.67965 , nBB =  6.815872 , nBO =  17.18413 
## 
## Round # 3 lnLikelihood =  -251.1293 
## 
##           Allele frequencies:      p =  0.326136 ,   q =  0.3092511 ,   r =  0.3646128 
## 
##           Genotype  counts:      nAA =  8.652777 , nAO =  19.34722 , nBB =  7.147033 , nBO =  16.85297 
## 
## Round # 4 lnLikelihood =  -251.1236 
## 
##           Allele frequencies:      p =  0.3271558 ,   q =  0.310267 ,   r =  0.3625773 
## 
##           Genotype  counts:      nAA =  8.705 , nAO =  19.295 , nBB =  7.191668 , nBO =  16.80833 
## 
## Round # 5 lnLikelihood =  -251.122 
## 
##           Allele frequencies:      p =  0.327316 ,   q =  0.3104039 ,   r =  0.3622802 
## 
##           Genotype  counts:      nAA =  8.712856 , nAO =  19.28714 , nBB =  7.19802 , nBO =  16.80198</code></pre>
<pre><code>##            p         q
## [1,] 0.32734 0.3104234</code></pre>
</div>
<div id="interpretation-1" class="section level3">
<h3>Interpretation:</h3>
<ul>
<li><p>From the results above, we get the eatimated value is : <span class="math inline">\(p=0.32734\)</span>, <span class="math inline">\(q= 0.3104234\)</span>,</p></li>
<li><p>The log likelihood estimate is : <span class="math inline">\(lnLikelihood = -261.3305 , -251.1735, -251.1293 ,-251.1236 ,-251.122\)</span> which means the maximum likelihood estimates is monotonous increasing.</p></li>
</ul>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="exercise-11.1.23" class="section level2">
<h2>Exercise 11.1.2(3)</h2>
<p><font color="#000000"></p>
<p>Use both for loops and lapply() to fit linear models to the mtcars using the formulas stored in this list:</p>
<ul>
<li>formulas &lt;- list(
<ul>
<li>mpg ~ disp,</li>
<li>mpg ~ I(1 / disp),</li>
<li>mpg ~ disp + wt,</li>
<li>mpg ~ I(1 / disp) + wt</li>
</ul></li>
<li>)</li>
</ul>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-exercise-11.1.23" class="section level2">
<h2>Answer Exercise 11.1.2(3)</h2>
<p><font color="#000000"></p>
<div id="the-idea-is-2" class="section level3">
<h3>The Idea is :</h3>
<ul>
<li><p>lapply() is a wrapper for a common for loop pattern: create a container for output, apply f() to each component of a list, and fill the container with the results. All other for loop functionals are variations on this theme: they simply use different types of input or output.</p></li>
<li><p>unlist() can be used to convert the output from a list to a vector to make it more compact. But data frames are also lists, lapply() is also useful when you want to do something to each column of a data frame.</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">formulas &lt;-<span class="st"> </span><span class="kw">list</span>(
  mpg <span class="op">~</span><span class="st"> </span>disp,
  mpg <span class="op">~</span><span class="st"> </span><span class="kw">I</span>(<span class="dv">1</span> <span class="op">/</span><span class="st"> </span>disp),
  mpg <span class="op">~</span><span class="st"> </span>disp <span class="op">+</span><span class="st"> </span>wt,
  mpg <span class="op">~</span><span class="st"> </span><span class="kw">I</span>(<span class="dv">1</span> <span class="op">/</span><span class="st"> </span>disp) <span class="op">+</span><span class="st"> </span>wt
)

 <span class="co"># with lapply()</span>

<span class="kw">print</span>(<span class="st">&quot;***************************************** with lapply() ***********************************************************&quot;</span>)</code></pre></div>
<pre><code>## [1] &quot;***************************************** with lapply() ***********************************************************&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">lapply</span>(formulas, lm, <span class="dt">data =</span> mtcars)</code></pre></div>
<pre><code>## [[1]]
## 
## Call:
## FUN(formula = X[[i]], data = ..1)
## 
## Coefficients:
## (Intercept)         disp  
##    29.59985     -0.04122  
## 
## 
## [[2]]
## 
## Call:
## FUN(formula = X[[i]], data = ..1)
## 
## Coefficients:
## (Intercept)    I(1/disp)  
##       10.75      1557.67  
## 
## 
## [[3]]
## 
## Call:
## FUN(formula = X[[i]], data = ..1)
## 
## Coefficients:
## (Intercept)         disp           wt  
##    34.96055     -0.01772     -3.35083  
## 
## 
## [[4]]
## 
## Call:
## FUN(formula = X[[i]], data = ..1)
## 
## Coefficients:
## (Intercept)    I(1/disp)           wt  
##      19.024     1142.560       -1.798</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="co"># with a for loop </span>

<span class="kw">print</span>(<span class="st">&quot;***************************************** with a for loop *********************************************************&quot;</span>)</code></pre></div>
<pre><code>## [1] &quot;***************************************** with a for loop *********************************************************&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">out_formulas &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">'list'</span>, <span class="kw">length</span>(formulas))
   <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(formulas)) {
       out_formulas[[i]] &lt;-<span class="st"> </span><span class="kw">lm</span>(formulas[[i]], <span class="dt">data =</span> mtcars)
   }
   out_formulas</code></pre></div>
<pre><code>## [[1]]
## 
## Call:
## lm(formula = formulas[[i]], data = mtcars)
## 
## Coefficients:
## (Intercept)         disp  
##    29.59985     -0.04122  
## 
## 
## [[2]]
## 
## Call:
## lm(formula = formulas[[i]], data = mtcars)
## 
## Coefficients:
## (Intercept)    I(1/disp)  
##       10.75      1557.67  
## 
## 
## [[3]]
## 
## Call:
## lm(formula = formulas[[i]], data = mtcars)
## 
## Coefficients:
## (Intercept)         disp           wt  
##    34.96055     -0.01772     -3.35083  
## 
## 
## [[4]]
## 
## Call:
## lm(formula = formulas[[i]], data = mtcars)
## 
## Coefficients:
## (Intercept)    I(1/disp)           wt  
##      19.024     1142.560       -1.798</code></pre>
</div>
<div id="interpretation-2" class="section level3">
<h3>Interpretation:</h3>
<ul>
<li><p>From the results, we can easily see that the loop with a for is the same as with the lapply</p></li>
<li><p>We get the model form are :</p>
<ul>
<li><p>[1] <span class="math inline">\(mpg = 29.59985 - 0.04122disp\)</span></p></li>
<li><p>[2] <span class="math inline">\(mpg = 10.75 - 1557.67I(1 / disp)\)</span></p></li>
<li><p>[3] <span class="math inline">\(mpg = 34.96055 - 0.01772disp - 3.35083wt\)</span></p></li>
<li><p>[4] <span class="math inline">\(mpg = 19.024 - 1142.560I(1 / disp) - 1.798wt\)</span></p></li>
</ul></li>
</ul>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="exercise-11.1.24" class="section level2">
<h2>Exercise 11.1.2(4)</h2>
<p><font color="#000000"></p>
<p>Fit the model <span class="math inline">\(mpg\)</span> ~ <span class="math inline">\(disp\)</span> to each of the bootstrap replicates of mtcars in the list below by using a for loop and lapply(). Can you do it without an anonymous function?</p>
<ul>
<li>bootstraps &lt;- lapply(1:10, function(i) {
<ul>
<li>rows &lt;- sample(1:nrow(mtcars), rep = TRUE)</li>
<li>mtcars[rows, ]</li>
</ul></li>
<li>})</li>
</ul>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-exercise-11.1.24" class="section level2">
<h2>Answer Exercise 11.1.2(4)</h2>
<p><font color="#000000"></p>
<div id="the-idea-is-3" class="section level3">
<h3>The Idea is :</h3>
<ul>
<li>The basic method is similar to the exercise 11.1.2(3)</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bootstraps &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="cf">function</span>(i) {
  rows &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(mtcars), <span class="dt">rep =</span> <span class="ot">TRUE</span>)
  mtcars[rows, ]
})

 <span class="co"># with lapply()</span>

<span class="kw">print</span>(<span class="st">&quot;***************************************** with lapply() ***********************************************************&quot;</span>)</code></pre></div>
<pre><code>## [1] &quot;***************************************** with lapply() ***********************************************************&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">lapply</span>(bootstraps, lm, <span class="dt">formula =</span> mpg<span class="op">~</span>disp)</code></pre></div>
<pre><code>## [[1]]
## 
## Call:
## FUN(formula = ..1, data = X[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    28.05769     -0.03647  
## 
## 
## [[2]]
## 
## Call:
## FUN(formula = ..1, data = X[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    29.11683     -0.04082  
## 
## 
## [[3]]
## 
## Call:
## FUN(formula = ..1, data = X[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    26.57385     -0.02937  
## 
## 
## [[4]]
## 
## Call:
## FUN(formula = ..1, data = X[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    28.28313     -0.03724  
## 
## 
## [[5]]
## 
## Call:
## FUN(formula = ..1, data = X[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    31.60253     -0.04562  
## 
## 
## [[6]]
## 
## Call:
## FUN(formula = ..1, data = X[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##     28.7208      -0.0383  
## 
## 
## [[7]]
## 
## Call:
## FUN(formula = ..1, data = X[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    27.11226     -0.03166  
## 
## 
## [[8]]
## 
## Call:
## FUN(formula = ..1, data = X[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    29.34297     -0.04015  
## 
## 
## [[9]]
## 
## Call:
## FUN(formula = ..1, data = X[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    26.52186     -0.02975  
## 
## 
## [[10]]
## 
## Call:
## FUN(formula = ..1, data = X[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    28.06815     -0.03533</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="co"># with a for loop </span>

<span class="kw">print</span>(<span class="st">&quot;***************************************** with a for loop *********************************************************&quot;</span>)</code></pre></div>
<pre><code>## [1] &quot;***************************************** with a for loop *********************************************************&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> out_bootstraps &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">'list'</span>, <span class="kw">length</span>(bootstraps))
   <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(bootstraps)) {
       out_bootstraps[[i]] &lt;-<span class="st"> </span><span class="kw">lm</span>(mpg<span class="op">~</span>disp, <span class="dt">data =</span> bootstraps[[i]])
   }
   out_bootstraps</code></pre></div>
<pre><code>## [[1]]
## 
## Call:
## lm(formula = mpg ~ disp, data = bootstraps[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    28.05769     -0.03647  
## 
## 
## [[2]]
## 
## Call:
## lm(formula = mpg ~ disp, data = bootstraps[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    29.11683     -0.04082  
## 
## 
## [[3]]
## 
## Call:
## lm(formula = mpg ~ disp, data = bootstraps[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    26.57385     -0.02937  
## 
## 
## [[4]]
## 
## Call:
## lm(formula = mpg ~ disp, data = bootstraps[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    28.28313     -0.03724  
## 
## 
## [[5]]
## 
## Call:
## lm(formula = mpg ~ disp, data = bootstraps[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    31.60253     -0.04562  
## 
## 
## [[6]]
## 
## Call:
## lm(formula = mpg ~ disp, data = bootstraps[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##     28.7208      -0.0383  
## 
## 
## [[7]]
## 
## Call:
## lm(formula = mpg ~ disp, data = bootstraps[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    27.11226     -0.03166  
## 
## 
## [[8]]
## 
## Call:
## lm(formula = mpg ~ disp, data = bootstraps[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    29.34297     -0.04015  
## 
## 
## [[9]]
## 
## Call:
## lm(formula = mpg ~ disp, data = bootstraps[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    26.52186     -0.02975  
## 
## 
## [[10]]
## 
## Call:
## lm(formula = mpg ~ disp, data = bootstraps[[i]])
## 
## Coefficients:
## (Intercept)         disp  
##    28.06815     -0.03533</code></pre>
</div>
<div id="interpretation-3" class="section level3">
<h3>Interpretation:</h3>
<ul>
<li><p>From the results, we can easily see that the loop with a for is the same as with the lapply.</p></li>
<li><p>We get the modle form are :</p>
<ul>
<li><p>[1] <span class="math inline">\(mpg = 28.96282 - 0.04149 disp\)</span></p></li>
<li><p>[2] <span class="math inline">\(mpg = 30.13856 - 0.04693 disp\)</span></p></li>
<li><p>[3] <span class="math inline">\(mpg = 27.09594 - 0.03264 disp\)</span></p></li>
<li><p>[4] <span class="math inline">\(mpg = 28.64848 - 0.03945 disp\)</span></p></li>
<li><p>[5] <span class="math inline">\(mpg = 29.94778 - 0.04477 disp\)</span></p></li>
<li><p>[6] <span class="math inline">\(mpg = 31.36958 - 0.04391 disp\)</span></p></li>
<li><p>[7] <span class="math inline">\(mpg = 28.92391 - 0.04218 disp\)</span></p></li>
<li><p>[8] <span class="math inline">\(mpg = 28.74858 - 0.03776 disp\)</span></p></li>
<li><p>[9] <span class="math inline">\(mpg = 28.70887 - 0.03956 disp\)</span></p></li>
<li><p>[10] <span class="math inline">\(mpg = 30.8249 - 0.0469 disp\)</span></p></li>
</ul></li>
</ul>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="exercise-11.1.25" class="section level2">
<h2>Exercise 11.1.2(5)</h2>
<p><font color="#000000"></p>
<p>For each model in the previous two exercises, extract <span class="math inline">\(R^2\)</span> using the function below.</p>
<ul>
<li>rsq &lt;- function(mod) summary(mod)$r.squared</li>
</ul>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-exercise-11.1.25" class="section level2">
<h2>Answer Exercise 11.1.2(5)</h2>
<p><font color="#000000"></p>
<div id="the-idea-is-4" class="section level3">
<h3>The Idea is :</h3>
<ul>
<li><p>Use lapply function to complete the loop with different models</p></li>
<li><p>Then utilize the unlist function to extract the target values of <span class="math inline">\(R^2\)</span></p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rsq &lt;-<span class="st"> </span><span class="cf">function</span>(mod) <span class="kw">summary</span>(mod)<span class="op">$</span>r.squared

bootstraps &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="cf">function</span>(i) {
  rows &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(mtcars), <span class="dt">rep =</span> <span class="ot">TRUE</span>)
  mtcars[rows, ]
})

models &lt;-<span class="st"> </span><span class="kw">lapply</span>(bootstraps, <span class="cf">function</span>(x) <span class="kw">lm</span>(mpg <span class="op">~</span><span class="st"> </span>disp, <span class="dt">data =</span> x))
<span class="kw">unlist</span>(<span class="kw">lapply</span>(models, rsq))</code></pre></div>
<pre><code>##  [1] 0.7917052 0.7794171 0.6530847 0.7284325 0.8481670 0.7629827 0.5556252
##  [8] 0.6279707 0.7375928 0.6907045</code></pre>
</div>
<div id="interpretation-4" class="section level3">
<h3>Interpretation:</h3>
<ul>
<li><p>The extracted values in different models are :</p></li>
<li><p><span class="math inline">\(R_1^2 = 0.6606581, R_2^2 = 0.8160798, R_3^2 = 0.5806448, R_4^2 = 0.7407087, R_5^2 = 0.7589143,\)</span></p></li>
<li><p><span class="math inline">\(R_6^2 = 0.6561107, R_7^2 = 0.6021043, R_8^2 = 0.6888394, R_9^2 = 0.5989015, R_{10}^2 = 0.7239530.\)</span></p></li>
</ul>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="exercise-11.2.53" class="section level2">
<h2>Exercise 11.2.5(3)</h2>
<p><font color="#000000"></p>
<p>The following code simulates the performance of a <span class="math inline">\(t-test\)</span> for non-normal data. Use sapply() and an anonymous function to extract the <span class="math inline">\(p\)</span>-value from every trial.</p>
<ul>
<li>trials &lt;- replicate(
<ul>
<li>100,</li>
<li>t.test(rpois(10, 10), rpois(7, 10)),</li>
<li>simplify = FALSE</li>
</ul></li>
<li>)</li>
</ul>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-exercise-11.2.53" class="section level2">
<h2>Answer exercise 11.2.5(3)</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">trials &lt;-<span class="st"> </span><span class="kw">replicate</span>(
     <span class="dv">100</span>, 
     <span class="kw">t.test</span>(<span class="kw">rpois</span>(<span class="dv">10</span>, <span class="dv">10</span>), <span class="kw">rpois</span>(<span class="dv">7</span>, <span class="dv">10</span>)),
     <span class="dt">simplify =</span> <span class="ot">FALSE</span>
     )

<span class="kw">print</span>(<span class="st">&quot;***************************************** with sapply() *******************************************************&quot;</span>)</code></pre></div>
<pre><code>## [1] &quot;***************************************** with sapply() *******************************************************&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sapply</span>(trials, <span class="cf">function</span>(mod) mod<span class="op">$</span>p.value)</code></pre></div>
<pre><code>##   [1] 0.394881346 0.279089532 0.383169001 0.793563375 0.426800406
##   [6] 0.336056282 0.539530683 0.830469683 0.166173129 0.851884822
##  [11] 0.368807367 0.064113108 0.683268999 0.721130938 0.024840971
##  [16] 0.766395252 0.186353225 0.520331712 0.192357246 0.082602928
##  [21] 0.691180069 0.569214632 0.072429110 0.970931490 0.264413277
##  [26] 0.676164243 0.596234855 0.198213401 0.310081712 0.213125392
##  [31] 0.098510854 0.426017110 0.280197930 0.356555815 0.266182988
##  [36] 0.053141402 0.748437171 0.895850794 0.104806155 0.867394882
##  [41] 0.640754931 0.566541951 0.796543928 0.339238295 0.181676038
##  [46] 0.172754495 0.052905260 0.820691657 0.080984226 0.836059664
##  [51] 0.118044691 0.897979780 0.033905580 0.586202641 0.510067278
##  [56] 0.543591085 0.331927065 0.662749678 0.188084993 0.411431435
##  [61] 0.079215123 0.667785263 0.073925607 0.212544794 0.508420043
##  [66] 0.217211303 0.024833585 0.028562029 0.219064494 0.247196201
##  [71] 0.674018911 0.846826443 0.627515860 0.407450305 0.858258529
##  [76] 0.418853293 0.671383188 0.581131788 0.003722897 0.529641136
##  [81] 0.528333276 0.177456264 0.045007391 0.828975431 0.965893932
##  [86] 0.213860615 0.930756654 0.911210023 0.109884254 0.085915324
##  [91] 0.590131942 0.338456408 0.528556544 0.180964102 0.197873188
##  [96] 0.586802862 0.464296507 0.479685154 0.820814205 0.191119001</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Extra challenge: get rid of the anonymous function by using `[[` directly.</span>
<span class="kw">print</span>(<span class="st">&quot;*************************************** get rid of the anonymous function ********************************************&quot;</span>)</code></pre></div>
<pre><code>## [1] &quot;*************************************** get rid of the anonymous function ********************************************&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sapply</span>(trials, <span class="st">`</span><span class="dt">[[</span><span class="st">`</span>, <span class="st">'p.value'</span>)</code></pre></div>
<pre><code>##   [1] 0.394881346 0.279089532 0.383169001 0.793563375 0.426800406
##   [6] 0.336056282 0.539530683 0.830469683 0.166173129 0.851884822
##  [11] 0.368807367 0.064113108 0.683268999 0.721130938 0.024840971
##  [16] 0.766395252 0.186353225 0.520331712 0.192357246 0.082602928
##  [21] 0.691180069 0.569214632 0.072429110 0.970931490 0.264413277
##  [26] 0.676164243 0.596234855 0.198213401 0.310081712 0.213125392
##  [31] 0.098510854 0.426017110 0.280197930 0.356555815 0.266182988
##  [36] 0.053141402 0.748437171 0.895850794 0.104806155 0.867394882
##  [41] 0.640754931 0.566541951 0.796543928 0.339238295 0.181676038
##  [46] 0.172754495 0.052905260 0.820691657 0.080984226 0.836059664
##  [51] 0.118044691 0.897979780 0.033905580 0.586202641 0.510067278
##  [56] 0.543591085 0.331927065 0.662749678 0.188084993 0.411431435
##  [61] 0.079215123 0.667785263 0.073925607 0.212544794 0.508420043
##  [66] 0.217211303 0.024833585 0.028562029 0.219064494 0.247196201
##  [71] 0.674018911 0.846826443 0.627515860 0.407450305 0.858258529
##  [76] 0.418853293 0.671383188 0.581131788 0.003722897 0.529641136
##  [81] 0.528333276 0.177456264 0.045007391 0.828975431 0.965893932
##  [86] 0.213860615 0.930756654 0.911210023 0.109884254 0.085915324
##  [91] 0.590131942 0.338456408 0.528556544 0.180964102 0.197873188
##  [96] 0.586802862 0.464296507 0.479685154 0.820814205 0.191119001</code></pre>
<p><font color="#000000"></p>
<div id="interpretation-5" class="section level3">
<h3>Interpretation:</h3>
<ul>
<li><p>From the results shown above, we can learn that the sapply and an anonymous function can extract the wanted p-values.</p></li>
<li><p>When getting rid of the anonymous function, we can also get the same results.</p></li>
</ul>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="exercise-11.2.56" class="section level2">
<h2>Exercise 11.2.5(6)</h2>
<p><font color="#000000"></p>
<p>Implement a combination of Map() and vapply() to create an lapply() variant that iterates in parallel over all of its inputs and stores its outputs in a vector (or a matrix). What arguments should the function take?</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-exercise-11.2.56" class="section level2">
<h2>Answer Exercise 11.2.5(6)</h2>
<p><font color="#000000"></p>
<div id="the-idea-is-5" class="section level3">
<h3>The Idea is :</h3>
<ul>
<li><p>lapply() takes a function, applies it to each element in a list, and returns the results in the form of a list.</p></li>
<li><p>vapply() is very similar to lapply() except it simplify output to produce an atomic vector. vapply() is more verbose, but gives more informative error messages and never fails silently. It is better suited for use inside other functions. vapply() is an implementation of lapply() that assigns results to a vector (or matrix) of appropriate type instead of as a list.</p></li>
<li><p>With lapply(), only one argument to the function varies; the others are fixed. This makes it poorly suited for some problems.</p></li>
<li><p>There??s a natural equivalence between Map() and lapply() because you can always convert a Map() to an lapply() that iterates over indices.</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(parallel)
mcvMap &lt;-<span class="st"> </span><span class="cf">function</span>(f, FUN.VALUE , ...) {
  out &lt;-<span class="st"> </span><span class="kw">mcMap</span>(f, ...)
  <span class="kw">vapply</span>(out, identity, FUN.VALUE)
   }</code></pre></div>
<hr />
<p><font color="#FF7F50"></p>
</div>
</div>
<div id="exercise-17.5.14" class="section level2">
<h2>Exercise 17.5.1(4)</h2>
<p><font color="#000000"></p>
<p>Make a faster version of <span class="math inline">\(chisq.test()\)</span> that only computes the <span class="math inline">\(chi-square\)</span> test statistic when the input is two numeric vectors with no missing values. You can try simplifying <span class="math inline">\(chisq.test()\)</span> or by coding from the mathematical definition (<span class="math inline">\(http://en. wikipedia.org/wiki/Pearson%27s_chi-squared_test\)</span>).</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-exercise-17.5.14" class="section level2">
<h2>Answer Exercise 17.5.1(4)</h2>
<p><font color="#000000"></p>
<ul>
<li>1, We fist use <em>getAnywhere(chisq.test)</em> to trace the source codes of <em>chisq.test()</em></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">getAnywhere</span>(chisq.test)</code></pre></div>
<pre><code>## A single object matching 'chisq.test' was found
## It was found in the following places
##   package:stats
##   namespace:stats
## with value
## 
## function (x, y = NULL, correct = TRUE, p = rep(1/length(x), length(x)), 
##     rescale.p = FALSE, simulate.p.value = FALSE, B = 2000) 
## {
##     DNAME &lt;- deparse(substitute(x))
##     if (is.data.frame(x)) 
##         x &lt;- as.matrix(x)
##     if (is.matrix(x)) {
##         if (min(dim(x)) == 1L) 
##             x &lt;- as.vector(x)
##     }
##     if (!is.matrix(x) &amp;&amp; !is.null(y)) {
##         if (length(x) != length(y)) 
##             stop(&quot;'x' and 'y' must have the same length&quot;)
##         DNAME2 &lt;- deparse(substitute(y))
##         xname &lt;- if (length(DNAME) &gt; 1L || nchar(DNAME, &quot;w&quot;) &gt; 
##             30) 
##             &quot;&quot;
##         else DNAME
##         yname &lt;- if (length(DNAME2) &gt; 1L || nchar(DNAME2, &quot;w&quot;) &gt; 
##             30) 
##             &quot;&quot;
##         else DNAME2
##         OK &lt;- complete.cases(x, y)
##         x &lt;- factor(x[OK])
##         y &lt;- factor(y[OK])
##         if ((nlevels(x) &lt; 2L) || (nlevels(y) &lt; 2L)) 
##             stop(&quot;'x' and 'y' must have at least 2 levels&quot;)
##         x &lt;- table(x, y)
##         names(dimnames(x)) &lt;- c(xname, yname)
##         DNAME &lt;- paste(paste(DNAME, collapse = &quot;\n&quot;), &quot;and&quot;, 
##             paste(DNAME2, collapse = &quot;\n&quot;))
##     }
##     if (any(x &lt; 0) || anyNA(x)) 
##         stop(&quot;all entries of 'x' must be nonnegative and finite&quot;)
##     if ((n &lt;- sum(x)) == 0) 
##         stop(&quot;at least one entry of 'x' must be positive&quot;)
##     if (simulate.p.value) {
##         setMETH &lt;- function() METHOD &lt;&lt;- paste(METHOD, &quot;with simulated p-value\n\t (based on&quot;, 
##             B, &quot;replicates)&quot;)
##         almost.1 &lt;- 1 - 64 * .Machine$double.eps
##     }
##     if (is.matrix(x)) {
##         METHOD &lt;- &quot;Pearson's Chi-squared test&quot;
##         nr &lt;- as.integer(nrow(x))
##         nc &lt;- as.integer(ncol(x))
##         if (is.na(nr) || is.na(nc) || is.na(nr * nc)) 
##             stop(&quot;invalid nrow(x) or ncol(x)&quot;, domain = NA)
##         sr &lt;- rowSums(x)
##         sc &lt;- colSums(x)
##         E &lt;- outer(sr, sc, &quot;*&quot;)/n
##         v &lt;- function(r, c, n) c * r * (n - r) * (n - c)/n^3
##         V &lt;- outer(sr, sc, v, n)
##         dimnames(E) &lt;- dimnames(x)
##         if (simulate.p.value &amp;&amp; all(sr &gt; 0) &amp;&amp; all(sc &gt; 0)) {
##             setMETH()
##             tmp &lt;- .Call(C_chisq_sim, sr, sc, B, E)
##             STATISTIC &lt;- sum(sort((x - E)^2/E, decreasing = TRUE))
##             PARAMETER &lt;- NA
##             PVAL &lt;- (1 + sum(tmp &gt;= almost.1 * STATISTIC))/(B + 
##                 1)
##         }
##         else {
##             if (simulate.p.value) 
##                 warning(&quot;cannot compute simulated p-value with zero marginals&quot;)
##             if (correct &amp;&amp; nrow(x) == 2L &amp;&amp; ncol(x) == 2L) {
##                 YATES &lt;- min(0.5, abs(x - E))
##                 if (YATES &gt; 0) 
##                   METHOD &lt;- paste(METHOD, &quot;with Yates' continuity correction&quot;)
##             }
##             else YATES &lt;- 0
##             STATISTIC &lt;- sum((abs(x - E) - YATES)^2/E)
##             PARAMETER &lt;- (nr - 1L) * (nc - 1L)
##             PVAL &lt;- pchisq(STATISTIC, PARAMETER, lower.tail = FALSE)
##         }
##     }
##     else {
##         if (length(dim(x)) &gt; 2L) 
##             stop(&quot;invalid 'x'&quot;)
##         if (length(x) == 1L) 
##             stop(&quot;'x' must at least have 2 elements&quot;)
##         if (length(x) != length(p)) 
##             stop(&quot;'x' and 'p' must have the same number of elements&quot;)
##         if (any(p &lt; 0)) 
##             stop(&quot;probabilities must be non-negative.&quot;)
##         if (abs(sum(p) - 1) &gt; sqrt(.Machine$double.eps)) {
##             if (rescale.p) 
##                 p &lt;- p/sum(p)
##             else stop(&quot;probabilities must sum to 1.&quot;)
##         }
##         METHOD &lt;- &quot;Chi-squared test for given probabilities&quot;
##         E &lt;- n * p
##         V &lt;- n * p * (1 - p)
##         STATISTIC &lt;- sum((x - E)^2/E)
##         names(E) &lt;- names(x)
##         if (simulate.p.value) {
##             setMETH()
##             nx &lt;- length(x)
##             sm &lt;- matrix(sample.int(nx, B * n, TRUE, prob = p), 
##                 nrow = n)
##             ss &lt;- apply(sm, 2L, function(x, E, k) {
##                 sum((table(factor(x, levels = 1L:k)) - E)^2/E)
##             }, E = E, k = nx)
##             PARAMETER &lt;- NA
##             PVAL &lt;- (1 + sum(ss &gt;= almost.1 * STATISTIC))/(B + 
##                 1)
##         }
##         else {
##             PARAMETER &lt;- length(x) - 1
##             PVAL &lt;- pchisq(STATISTIC, PARAMETER, lower.tail = FALSE)
##         }
##     }
##     names(STATISTIC) &lt;- &quot;X-squared&quot;
##     names(PARAMETER) &lt;- &quot;df&quot;
##     if (any(E &lt; 5) &amp;&amp; is.finite(PARAMETER)) 
##         warning(&quot;Chi-squared approximation may be incorrect&quot;)
##     structure(list(statistic = STATISTIC, parameter = PARAMETER, 
##         p.value = PVAL, method = METHOD, data.name = DNAME, observed = x, 
##         expected = E, residuals = (x - E)/sqrt(E), stdres = (x - 
##             E)/sqrt(V)), class = &quot;htest&quot;)
## }
## &lt;bytecode: 0x00000000317e9538&gt;
## &lt;environment: namespace:stats&gt;</code></pre>
<ul>
<li>2, Since <em>chisq.test()</em> has a relatively long source code, we try a new implementation from scratch:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chisq.test2 &lt;-<span class="st"> </span><span class="cf">function</span>(x, y){

  <span class="co"># Input</span>
  <span class="cf">if</span> (<span class="op">!</span><span class="kw">is.numeric</span>(x)) {
    <span class="kw">stop</span>(<span class="st">&quot;x must be numeric&quot;</span>)}
  <span class="cf">if</span> (<span class="op">!</span><span class="kw">is.numeric</span>(y)) {
    <span class="kw">stop</span>(<span class="st">&quot;y must be numeric&quot;</span>)}
  <span class="cf">if</span> (<span class="kw">length</span>(x) <span class="op">!=</span><span class="st"> </span><span class="kw">length</span>(y)) {
    <span class="kw">stop</span>(<span class="st">&quot;x and y must have the same length&quot;</span>)}
  <span class="cf">if</span> (<span class="kw">length</span>(x) <span class="op">&lt;=</span><span class="st"> </span><span class="dv">1</span>) {
    <span class="kw">stop</span>(<span class="st">&quot;length of x must be greater one&quot;</span>)}
  <span class="cf">if</span> (<span class="kw">any</span>(<span class="kw">c</span>(x, y) <span class="op">&lt;</span><span class="st"> </span><span class="dv">0</span>)) {
    <span class="kw">stop</span>(<span class="st">&quot;all entries of x and y must be greater or equal zero&quot;</span>)}
  <span class="cf">if</span> (<span class="kw">sum</span>(<span class="kw">complete.cases</span>(x, y)) <span class="op">!=</span><span class="st"> </span><span class="kw">length</span>(x)) {
    <span class="kw">stop</span>(<span class="st">&quot;there must be no missing values in x and y&quot;</span>)}
  <span class="cf">if</span> (<span class="kw">any</span>(<span class="kw">is.null</span>(<span class="kw">c</span>(x, y)))) {
    <span class="kw">stop</span>(<span class="st">&quot;entries of x and y must not be NULL&quot;</span>)}

  <span class="co"># Help variables</span>
  m &lt;-<span class="st"> </span><span class="kw">rbind</span>(x, y)
  margin1 &lt;-<span class="st"> </span><span class="kw">rowSums</span>(m)
  margin2 &lt;-<span class="st"> </span><span class="kw">colSums</span>(m)
  n &lt;-<span class="st"> </span><span class="kw">sum</span>(m)
  me &lt;-<span class="st"> </span><span class="kw">tcrossprod</span>(margin1, margin2) <span class="op">/</span><span class="st"> </span>n

  <span class="co"># Output</span>
  x_stat =<span class="st"> </span><span class="kw">sum</span>((m <span class="op">-</span><span class="st"> </span>me)<span class="op">^</span><span class="dv">2</span> <span class="op">/</span><span class="st"> </span>me)
  dof &lt;-<span class="st"> </span>(<span class="kw">length</span>(margin1) <span class="op">-</span><span class="st"> </span><span class="dv">1</span>) <span class="op">*</span><span class="st"> </span>(<span class="kw">length</span>(margin2) <span class="op">-</span><span class="st"> </span><span class="dv">1</span>)
  p &lt;-<span class="st"> </span><span class="kw">pchisq</span>(x_stat, <span class="dt">df =</span> dof, <span class="dt">lower.tail =</span> <span class="ot">FALSE</span>)

  <span class="kw">return</span>(<span class="kw">list</span>(<span class="dt">x_stat =</span> x_stat, <span class="dt">df =</span> dof, <span class="st">`</span><span class="dt">p-value</span><span class="st">`</span> =<span class="st"> </span>p))
}</code></pre></div>
<ul>
<li>3, We check if our new implementation returns the same results</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a &lt;-<span class="st"> </span><span class="dv">21</span><span class="op">:</span><span class="dv">25</span>
b &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">21</span>, <span class="dv">23</span>, <span class="dv">25</span>, <span class="dv">27</span>, <span class="dv">29</span>)
m &lt;-<span class="st"> </span><span class="kw">cbind</span>(a, b)
m</code></pre></div>
<pre><code>##       a  b
## [1,] 21 21
## [2,] 22 23
## [3,] 23 25
## [4,] 24 27
## [5,] 25 29</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">chisq.test</span>(m)</code></pre></div>
<pre><code>## 
##  Pearson's Chi-squared test
## 
## data:  m
## X-squared = 0.16194, df = 4, p-value = 0.9969</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">chisq.test2</span>(a, b)</code></pre></div>
<pre><code>## $x_stat
## [1] 0.1619369
## 
## $df
## [1] 4
## 
## $`p-value`
## [1] 0.9968937</code></pre>
<ul>
<li>4, Finally we benchmark this implementation against a compiled version of itself and the original stats::chisq.test():</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chisq.test2c &lt;-<span class="st"> </span>compiler<span class="op">::</span><span class="kw">cmpfun</span>(chisq.test2)

microbenchmark<span class="op">::</span><span class="kw">microbenchmark</span>(
  <span class="kw">chisq.test</span>(m),
  <span class="kw">chisq.test2</span>(a, b),
  <span class="kw">chisq.test2c</span>(a, b)
)</code></pre></div>
<pre><code>## Unit: microseconds
##                expr    min      lq      mean  median       uq     max
##       chisq.test(m) 87.018 89.6950 100.94513 91.9265 102.8595 303.000
##   chisq.test2(a, b) 22.758 23.8740  28.18940 25.2130  29.8990 103.975
##  chisq.test2c(a, b) 22.312 24.0975  28.25641 25.6590  30.7910  93.711
##  neval
##    100
##    100
##    100</code></pre>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="exercise-17.5.15" class="section level2">
<h2>Exercise 17.5.1(5)</h2>
<p><font color="#000000"></p>
<p>Can you make a faster version of <span class="math inline">\(table()\)</span> for the case of an input of two integer vectors with no missing values? Can you use it to speed up your <span class="math inline">\(chi-square\)</span> test?</p>
<p><font color="#FF7F50"></p>
</div>
<div id="answer-exercise-17.5.15" class="section level2">
<h2>Answer Exercise 17.5.1(5)</h2>
<p><font color="#000000"></p>
<ul>
<li>We fist use <em>getAnywhere(chisq.test)</em> to trace the source codes of <em>table()</em></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getAnywhere</span>(table)</code></pre></div>
<pre><code>## A single object matching 'table' was found
## It was found in the following places
##   package:base
##   namespace:base
## with value
## 
## function (..., exclude = if (useNA == &quot;no&quot;) c(NA, NaN), useNA = c(&quot;no&quot;, 
##     &quot;ifany&quot;, &quot;always&quot;), dnn = list.names(...), deparse.level = 1) 
## {
##     list.names &lt;- function(...) {
##         l &lt;- as.list(substitute(list(...)))[-1L]
##         nm &lt;- names(l)
##         fixup &lt;- if (is.null(nm)) 
##             seq_along(l)
##         else nm == &quot;&quot;
##         dep &lt;- vapply(l[fixup], function(x) switch(deparse.level + 
##             1, &quot;&quot;, if (is.symbol(x)) as.character(x) else &quot;&quot;, 
##             deparse(x, nlines = 1)[1L]), &quot;&quot;)
##         if (is.null(nm)) 
##             dep
##         else {
##             nm[fixup] &lt;- dep
##             nm
##         }
##     }
##     miss.use &lt;- missing(useNA)
##     miss.exc &lt;- missing(exclude)
##     useNA &lt;- if (miss.use &amp;&amp; !miss.exc &amp;&amp; !match(NA, exclude, 
##         nomatch = 0L)) 
##         &quot;ifany&quot;
##     else match.arg(useNA)
##     doNA &lt;- useNA != &quot;no&quot;
##     if (!miss.use &amp;&amp; !miss.exc &amp;&amp; doNA &amp;&amp; match(NA, exclude, 
##         nomatch = 0L)) 
##         warning(&quot;'exclude' containing NA and 'useNA' != \&quot;no\&quot;' are a bit contradicting&quot;)
##     args &lt;- list(...)
##     if (!length(args)) 
##         stop(&quot;nothing to tabulate&quot;)
##     if (length(args) == 1L &amp;&amp; is.list(args[[1L]])) {
##         args &lt;- args[[1L]]
##         if (length(dnn) != length(args)) 
##             dnn &lt;- if (!is.null(argn &lt;- names(args))) 
##                 argn
##             else paste(dnn[1L], seq_along(args), sep = &quot;.&quot;)
##     }
##     bin &lt;- 0L
##     lens &lt;- NULL
##     dims &lt;- integer()
##     pd &lt;- 1L
##     dn &lt;- NULL
##     for (a in args) {
##         if (is.null(lens)) 
##             lens &lt;- length(a)
##         else if (length(a) != lens) 
##             stop(&quot;all arguments must have the same length&quot;)
##         fact.a &lt;- is.factor(a)
##         if (doNA) 
##             aNA &lt;- anyNA(a)
##         if (!fact.a) {
##             a0 &lt;- a
##             a &lt;- factor(a, exclude = exclude)
##         }
##         add.na &lt;- doNA
##         if (add.na) {
##             ifany &lt;- (useNA == &quot;ifany&quot;)
##             anNAc &lt;- anyNA(a)
##             add.na &lt;- if (!ifany || anNAc) {
##                 ll &lt;- levels(a)
##                 if (add.ll &lt;- !anyNA(ll)) {
##                   ll &lt;- c(ll, NA)
##                   TRUE
##                 }
##                 else if (!ifany &amp;&amp; !anNAc) 
##                   FALSE
##                 else TRUE
##             }
##             else FALSE
##         }
##         if (add.na) 
##             a &lt;- factor(a, levels = ll, exclude = NULL)
##         else ll &lt;- levels(a)
##         a &lt;- as.integer(a)
##         if (fact.a &amp;&amp; !miss.exc) {
##             ll &lt;- ll[keep &lt;- which(match(ll, exclude, nomatch = 0L) == 
##                 0L)]
##             a &lt;- match(a, keep)
##         }
##         else if (!fact.a &amp;&amp; add.na) {
##             if (ifany &amp;&amp; !aNA &amp;&amp; add.ll) {
##                 ll &lt;- ll[!is.na(ll)]
##                 is.na(a) &lt;- match(a0, c(exclude, NA), nomatch = 0L) &gt; 
##                   0L
##             }
##             else {
##                 is.na(a) &lt;- match(a0, exclude, nomatch = 0L) &gt; 
##                   0L
##             }
##         }
##         nl &lt;- length(ll)
##         dims &lt;- c(dims, nl)
##         if (prod(dims) &gt; .Machine$integer.max) 
##             stop(&quot;attempt to make a table with &gt;= 2^31 elements&quot;)
##         dn &lt;- c(dn, list(ll))
##         bin &lt;- bin + pd * (a - 1L)
##         pd &lt;- pd * nl
##     }
##     names(dn) &lt;- dnn
##     bin &lt;- bin[!is.na(bin)]
##     if (length(bin)) 
##         bin &lt;- bin + 1L
##     y &lt;- array(tabulate(bin, pd), dims, dimnames = dn)
##     class(y) &lt;- &quot;table&quot;
##     y
## }
## &lt;bytecode: 0x000000002c15d498&gt;
## &lt;environment: namespace:base&gt;</code></pre>
<p>Then we code to speed up the <span class="math inline">\(chi-square\)</span> test.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">table2 &lt;-<span class="st"> </span><span class="cf">function</span>(x, y) {

  x_val &lt;-<span class="st"> </span><span class="kw">unique</span>(x)

  y_val &lt;-<span class="st"> </span><span class="kw">unique</span>(y)

  mat &lt;-<span class="st"> </span><span class="kw">matrix</span>(0L, <span class="kw">length</span>(x_val), <span class="kw">length</span>(y_val))

  <span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(x)) {

    mat[<span class="kw">which</span>(x_val <span class="op">==</span><span class="st"> </span>x[[i]]), <span class="kw">which</span>(y_val <span class="op">==</span><span class="st"> </span>y[[i]])] &lt;-

<span class="st">      </span>mat[<span class="kw">which</span>(x_val <span class="op">==</span><span class="st"> </span>x[[i]]),  <span class="kw">which</span>(y_val <span class="op">==</span><span class="st"> </span>y[[i]])] <span class="op">+</span><span class="st"> </span>1L

  }

  dimnames &lt;-<span class="st"> </span><span class="kw">list</span>(x_val, y_val)

  <span class="kw">names</span>(dimnames) &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">as.list</span>(<span class="kw">match.call</span>())[<span class="op">-</span><span class="dv">1</span>])  <span class="co"># R has names for dimnames... :/</span>

  tab &lt;-<span class="st"> </span><span class="kw">array</span>(mat, <span class="dt">dim =</span> <span class="kw">dim</span>(mat), <span class="dt">dimnames =</span> dimnames)

  <span class="kw">class</span>(tab) &lt;-<span class="st"> &quot;table&quot;</span>

  tab

}</code></pre></div>
<p>check:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)

<span class="kw">identical</span>(<span class="kw">table</span>(a, a), <span class="kw">table2</span>(a, a))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>we benchmark this implementation against a compiled version of itself and the original stats::table():</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">microbenchmark<span class="op">::</span><span class="kw">microbenchmark</span>(<span class="kw">table</span>(a, a), <span class="kw">table2</span>(a, a))</code></pre></div>
<pre><code>## Unit: microseconds
##          expr     min      lq      mean  median       uq     max neval
##   table(a, a) 196.348 213.082 239.65100 229.815 239.1870 545.310   100
##  table2(a, a)  51.319  55.781  68.95835  64.705  74.0765 213.305   100</code></pre>
<p>check</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">b &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>)
<span class="kw">identical</span>(<span class="kw">table</span>(a, b), <span class="kw">table2</span>(a, b))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">c &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)
d &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>)
<span class="kw">identical</span>(<span class="kw">table</span>(c, d), <span class="kw">table2</span>(c, d))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">e &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>)
<span class="kw">identical</span>(<span class="kw">table</span>(a, e), <span class="kw">table2</span>(a, e))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">identical</span>(<span class="kw">table</span>(b, e), <span class="kw">table2</span>(b, e))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">identical</span>(<span class="kw">table</span>(e, e), <span class="kw">table2</span>(e, e))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>)
<span class="kw">identical</span>(<span class="kw">table</span>(f, f), <span class="kw">table2</span>(f, f))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">identical</span>(<span class="kw">table</span>(e, f), <span class="kw">table2</span>(e, f))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">g &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">4</span>, <span class="dv">9</span>)
<span class="kw">identical</span>(<span class="kw">table</span>(g, g), <span class="kw">table2</span>(g, g))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">identical</span>(<span class="kw">table</span>(g, f), <span class="kw">table2</span>(g, f))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">h &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">10</span>, <span class="dv">20</span>, <span class="dv">30</span>)
<span class="kw">identical</span>(<span class="kw">table</span>(h, h), <span class="kw">table2</span>(h, h))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">identical</span>(<span class="kw">table</span>(h, f), <span class="kw">table2</span>(h, f))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">identical</span>(<span class="kw">table</span>(h, g), <span class="kw">table2</span>(h, g))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">i &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>)

<span class="kw">identical</span>(<span class="kw">table</span>(i, i), <span class="kw">table2</span>(i, i))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">identical</span>(<span class="kw">table</span>(h, i), <span class="kw">table2</span>(h, i))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>then we have:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">j &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">1000</span>, <span class="dv">10000</span>)

<span class="kw">identical</span>(<span class="kw">table</span>(j, j), <span class="kw">table2</span>(j, j))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">microbenchmark<span class="op">::</span><span class="kw">microbenchmark</span>(<span class="kw">table</span>(j, j), <span class="kw">table2</span>(j, j))</code></pre></div>
<pre><code>## Unit: microseconds
##          expr     min      lq      mean  median       uq     max neval
##   table(j, j) 209.289 219.775 240.46318 229.370 250.3430 421.701   100
##  table2(j, j)  59.797  65.598  74.11245  72.515  75.8615 189.208   100</code></pre>
<hr />
<p><font color="#FF7F50"></p>
</div>
<div id="thanks-for-your-reading" class="section level2">
<h2>Thanks for your reading!</h2>
<hr />
<hr />
<hr />
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
